<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data</title>
  <meta name="description" content="Analysis of single cell RNA-seq data">
  <meta name="generator" content="bookdown 0.6 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data" />
  
  
  

<meta name="author" content="Vladimir Kiselev (wikiselev), Tallulah Andrews, Davis McCarthy (davisjmcc), Maren Büttner (marenbuettner) and Martin Hemberg (m_hemberg)">


<meta name="date" content="2018-02-02">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="data-types.html">
<link rel="next" href="tabula-muris.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#video"><i class="fa fa-check"></i><b>1.1</b> Video</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#r-based"><i class="fa fa-check"></i><b>1.4</b> R-based</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#docker-image-rstudio"><i class="fa fa-check"></i><b>1.5</b> Docker image (RStudio)</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#manual-installation"><i class="fa fa-check"></i><b>1.6</b> Manual installation</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.7</b> License</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.8</b> Prerequisites</a></li>
<li class="chapter" data-level="1.9" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.9</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html"><i class="fa fa-check"></i><b>3</b> Processing Raw scRNA-seq Data</a><ul>
<li class="chapter" data-level="3.1" data-path="processing-raw-scrna-seq-data.html"><a href="processing-raw-scrna-seq-data.html#references"><i class="fa fa-check"></i><b>3.1</b> References</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="solution-and-downloading-the-report.html"><a href="solution-and-downloading-the-report.html"><i class="fa fa-check"></i><b>4</b> Solution and Downloading the Report</a></li>
<li class="chapter" data-level="5" data-path="trimming-reads.html"><a href="trimming-reads.html"><i class="fa fa-check"></i><b>5</b> Trimming Reads</a></li>
<li class="chapter" data-level="6" data-path="solution.html"><a href="solution.html"><i class="fa fa-check"></i><b>6</b> Solution</a></li>
<li class="chapter" data-level="7" data-path="file-formats.html"><a href="file-formats.html"><i class="fa fa-check"></i><b>7</b> File formats</a><ul>
<li class="chapter" data-level="7.1" data-path="file-formats.html"><a href="file-formats.html#fastq"><i class="fa fa-check"></i><b>7.1</b> FastQ</a></li>
<li class="chapter" data-level="7.2" data-path="file-formats.html"><a href="file-formats.html#bam"><i class="fa fa-check"></i><b>7.2</b> BAM</a></li>
<li class="chapter" data-level="7.3" data-path="file-formats.html"><a href="file-formats.html#cram"><i class="fa fa-check"></i><b>7.3</b> CRAM</a></li>
<li class="chapter" data-level="7.4" data-path="file-formats.html"><a href="file-formats.html#mannually-inspecting-file"><i class="fa fa-check"></i><b>7.4</b> Mannually Inspecting file</a></li>
<li class="chapter" data-level="7.5" data-path="file-formats.html"><a href="file-formats.html#genome-fasta-gtf"><i class="fa fa-check"></i><b>7.5</b> Genome (FASTA, GTF)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="demultiplexing.html"><a href="demultiplexing.html"><i class="fa fa-check"></i><b>8</b> Demultiplexing</a><ul>
<li class="chapter" data-level="8.1" data-path="demultiplexing.html"><a href="demultiplexing.html#identifying-cell-containing-dropletsmicrowells"><i class="fa fa-check"></i><b>8.1</b> Identifying cell-containing droplets/microwells</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="using-star-to-align-reads.html"><a href="using-star-to-align-reads.html"><i class="fa fa-check"></i><b>9</b> Using STAR to Align Reads</a><ul>
<li class="chapter" data-level="9.1" data-path="using-star-to-align-reads.html"><a href="using-star-to-align-reads.html#references-1"><i class="fa fa-check"></i><b>9.1</b> References</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="solution-for-star-alignment.html"><a href="solution-for-star-alignment.html"><i class="fa fa-check"></i><b>10</b> Solution for STAR Alignment</a></li>
<li class="chapter" data-level="11" data-path="kallisto-and-pseudo-alignment.html"><a href="kallisto-and-pseudo-alignment.html"><i class="fa fa-check"></i><b>11</b> Kallisto and Pseudo-Alignment</a><ul>
<li class="chapter" data-level="11.1" data-path="kallisto-and-pseudo-alignment.html"><a href="kallisto-and-pseudo-alignment.html#what-is-a-k-mer"><i class="fa fa-check"></i><b>11.1</b> What is a k-mer?</a></li>
<li class="chapter" data-level="11.2" data-path="kallisto-and-pseudo-alignment.html"><a href="kallisto-and-pseudo-alignment.html#why-map-k-mers-rather-than-reads"><i class="fa fa-check"></i><b>11.2</b> Why map k-mers rather than reads?</a></li>
<li class="chapter" data-level="11.3" data-path="kallisto-and-pseudo-alignment.html"><a href="kallisto-and-pseudo-alignment.html#kallistos-pseudo-mode"><i class="fa fa-check"></i><b>11.3</b> Kallisto’s pseudo mode</a></li>
<li class="chapter" data-level="11.4" data-path="kallisto-and-pseudo-alignment.html"><a href="kallisto-and-pseudo-alignment.html#references-2"><i class="fa fa-check"></i><b>11.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="solution-to-kallisto-pseudo-alignment.html"><a href="solution-to-kallisto-pseudo-alignment.html"><i class="fa fa-check"></i><b>12</b> Solution to Kallisto Pseudo-Alignment</a><ul>
<li class="chapter" data-level="12.1" data-path="solution-to-kallisto-pseudo-alignment.html"><a href="solution-to-kallisto-pseudo-alignment.html#understanding-the-output-of-kallisto-pseudo-alignment"><i class="fa fa-check"></i><b>12.1</b> Understanding the Output of Kallisto Pseudo-Alignment</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>13</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="13.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-qc"><i class="fa fa-check"></i><b>13.1</b> Reads QC</a></li>
<li class="chapter" data-level="13.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>13.2</b> Reads alignment</a></li>
<li class="chapter" data-level="13.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-example"><i class="fa fa-check"></i><b>13.3</b> Alignment example</a></li>
<li class="chapter" data-level="13.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>13.4</b> Mapping QC</a></li>
<li class="chapter" data-level="13.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>13.5</b> Reads quantification</a></li>
<li class="chapter" data-level="13.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>13.6</b> Unique Molecular Identifiers (UMIs)</a><ul>
<li class="chapter" data-level="13.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>13.6.1</b> Introduction</a></li>
<li class="chapter" data-level="13.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-barcodes"><i class="fa fa-check"></i><b>13.6.2</b> Mapping Barcodes</a></li>
<li class="chapter" data-level="13.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#counting-barcodes"><i class="fa fa-check"></i><b>13.6.3</b> Counting Barcodes</a></li>
<li class="chapter" data-level="13.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correcting-for-errors"><i class="fa fa-check"></i><b>13.6.4</b> Correcting for Errors</a></li>
<li class="chapter" data-level="13.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downstream-analysis"><i class="fa fa-check"></i><b>13.6.5</b> Downstream Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html"><i class="fa fa-check"></i><b>14</b> Introduction to R/Bioconductor</a><ul>
<li class="chapter" data-level="14.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>14.1</b> Installing packages</a><ul>
<li class="chapter" data-level="14.1.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#cran"><i class="fa fa-check"></i><b>14.1.1</b> CRAN</a></li>
<li class="chapter" data-level="14.1.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#github-1"><i class="fa fa-check"></i><b>14.1.2</b> Github</a></li>
<li class="chapter" data-level="14.1.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>14.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="14.1.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#source"><i class="fa fa-check"></i><b>14.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>14.2</b> Installation instructions:</a></li>
<li class="chapter" data-level="14.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#data-typesclasses"><i class="fa fa-check"></i><b>14.3</b> Data-types/classes</a><ul>
<li class="chapter" data-level="14.3.1" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#numeric"><i class="fa fa-check"></i><b>14.3.1</b> Numeric</a></li>
<li class="chapter" data-level="14.3.2" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#characterstring"><i class="fa fa-check"></i><b>14.3.2</b> Character/String</a></li>
<li class="chapter" data-level="14.3.3" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#logical"><i class="fa fa-check"></i><b>14.3.3</b> Logical</a></li>
<li class="chapter" data-level="14.3.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#factors"><i class="fa fa-check"></i><b>14.3.4</b> Factors</a></li>
<li class="chapter" data-level="14.3.5" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#checking-classtype"><i class="fa fa-check"></i><b>14.3.5</b> Checking class/type</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#basic-data-structures"><i class="fa fa-check"></i><b>14.4</b> Basic data structures</a></li>
<li class="chapter" data-level="14.5" data-path="introduction-to-rbioconductor.html"><a href="introduction-to-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>14.5</b> More information</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>15</b> Data Types</a><ul>
<li class="chapter" data-level="15.1" data-path="data-types.html"><a href="data-types.html#what-is-tidy-data"><i class="fa fa-check"></i><b>15.1</b> What is Tidy Data?</a></li>
<li class="chapter" data-level="15.2" data-path="data-types.html"><a href="data-types.html#what-is-rich-data"><i class="fa fa-check"></i><b>15.2</b> What is Rich Data?</a></li>
<li class="chapter" data-level="15.3" data-path="data-types.html"><a href="data-types.html#what-is-bioconductor"><i class="fa fa-check"></i><b>15.3</b> What is Bioconductor?</a></li>
<li class="chapter" data-level="15.4" data-path="data-types.html"><a href="data-types.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>15.4</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="15.5" data-path="data-types.html"><a href="data-types.html#scater-package"><i class="fa fa-check"></i><b>15.5</b> <code>scater</code> package</a></li>
<li class="chapter" data-level="15.6" data-path="data-types.html"><a href="data-types.html#references-3"><i class="fa fa-check"></i><b>15.6</b> References</a></li>
<li class="chapter" data-level="15.7" data-path="data-types.html"><a href="data-types.html#bioconductor-singlecellexperiment-and-scater"><i class="fa fa-check"></i><b>15.7</b> Bioconductor, <code>SingleCellExperiment</code> and <code>scater</code></a><ul>
<li class="chapter" data-level="15.7.1" data-path="data-types.html"><a href="data-types.html#bioconductor-1"><i class="fa fa-check"></i><b>15.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="15.7.2" data-path="data-types.html"><a href="data-types.html#singlecellexperiment-class-1"><i class="fa fa-check"></i><b>15.7.2</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="15.7.3" data-path="data-types.html"><a href="data-types.html#scater-package-1"><i class="fa fa-check"></i><b>15.7.3</b> <code>scater</code> package</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>16</b> An Introduction to ggplot2</a><ul>
<li class="chapter" data-level="16.1" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#what-is-ggplot2"><i class="fa fa-check"></i><b>16.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="16.2" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>16.2</b> Principles of ggplot2</a></li>
<li class="chapter" data-level="16.3" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#using-the-aes-mapping-function"><i class="fa fa-check"></i><b>16.3</b> Using the <code>aes</code> mapping function</a></li>
<li class="chapter" data-level="16.4" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#geoms"><i class="fa fa-check"></i><b>16.4</b> Geoms</a></li>
<li class="chapter" data-level="16.5" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#plotting-data-from-more-than-2-cells"><i class="fa fa-check"></i><b>16.5</b> Plotting data from more than 2 cells</a></li>
<li class="chapter" data-level="16.6" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#plotting-heatmaps"><i class="fa fa-check"></i><b>16.6</b> Plotting heatmaps</a></li>
<li class="chapter" data-level="16.7" data-path="an-introduction-to-ggplot2.html"><a href="an-introduction-to-ggplot2.html#principle-component-analysis"><i class="fa fa-check"></i><b>16.7</b> Principle Component Analysis</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="tabula-muris.html"><a href="tabula-muris.html"><i class="fa fa-check"></i><b>17</b> Tabula Muris</a><ul>
<li class="chapter" data-level="17.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-1"><i class="fa fa-check"></i><b>17.1</b> Introduction</a></li>
<li class="chapter" data-level="17.2" data-path="tabula-muris.html"><a href="tabula-muris.html#downloading-the-data"><i class="fa fa-check"></i><b>17.2</b> Downloading the data</a></li>
<li class="chapter" data-level="17.3" data-path="tabula-muris.html"><a href="tabula-muris.html#reading-the-data-smartseq2"><i class="fa fa-check"></i><b>17.3</b> Reading the data (Smartseq2)</a></li>
<li class="chapter" data-level="17.4" data-path="tabula-muris.html"><a href="tabula-muris.html#building-a-scater-object"><i class="fa fa-check"></i><b>17.4</b> Building a scater object</a></li>
<li class="chapter" data-level="17.5" data-path="tabula-muris.html"><a href="tabula-muris.html#reading-the-data-10x"><i class="fa fa-check"></i><b>17.5</b> Reading the data (10X)</a></li>
<li class="chapter" data-level="17.6" data-path="tabula-muris.html"><a href="tabula-muris.html#building-a-scater-object-1"><i class="fa fa-check"></i><b>17.6</b> Building a scater object</a></li>
<li class="chapter" data-level="17.7" data-path="tabula-muris.html"><a href="tabula-muris.html#advanced-exercise"><i class="fa fa-check"></i><b>17.7</b> Advanced Exercise</a></li>
<li class="chapter" data-level="17.8" data-path="tabula-muris.html"><a href="tabula-muris.html#exprs-qc"><i class="fa fa-check"></i><b>17.8</b> Expression QC (UMI)</a><ul>
<li class="chapter" data-level="17.8.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-2"><i class="fa fa-check"></i><b>17.8.1</b> Introduction</a></li>
<li class="chapter" data-level="17.8.2" data-path="tabula-muris.html"><a href="tabula-muris.html#tung-dataset"><i class="fa fa-check"></i><b>17.8.2</b> Tung dataset</a></li>
<li class="chapter" data-level="17.8.3" data-path="tabula-muris.html"><a href="tabula-muris.html#cell-qc"><i class="fa fa-check"></i><b>17.8.3</b> Cell QC</a></li>
<li class="chapter" data-level="17.8.4" data-path="tabula-muris.html"><a href="tabula-muris.html#cell-filtering"><i class="fa fa-check"></i><b>17.8.4</b> Cell filtering</a></li>
<li class="chapter" data-level="17.8.5" data-path="tabula-muris.html"><a href="tabula-muris.html#compare-filterings"><i class="fa fa-check"></i><b>17.8.5</b> Compare filterings</a></li>
<li class="chapter" data-level="17.8.6" data-path="tabula-muris.html"><a href="tabula-muris.html#gene-analysis"><i class="fa fa-check"></i><b>17.8.6</b> Gene analysis</a></li>
<li class="chapter" data-level="17.8.7" data-path="tabula-muris.html"><a href="tabula-muris.html#save-the-data"><i class="fa fa-check"></i><b>17.8.7</b> Save the data</a></li>
<li class="chapter" data-level="17.8.8" data-path="tabula-muris.html"><a href="tabula-muris.html#big-exercise"><i class="fa fa-check"></i><b>17.8.8</b> Big Exercise</a></li>
<li class="chapter" data-level="17.8.9" data-path="tabula-muris.html"><a href="tabula-muris.html#sessioninfo"><i class="fa fa-check"></i><b>17.8.9</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="17.9" data-path="tabula-muris.html"><a href="tabula-muris.html#expression-qc-reads"><i class="fa fa-check"></i><b>17.9</b> Expression QC (Reads)</a></li>
<li class="chapter" data-level="17.10" data-path="tabula-muris.html"><a href="tabula-muris.html#data-visualization"><i class="fa fa-check"></i><b>17.10</b> Data visualization</a><ul>
<li class="chapter" data-level="17.10.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-3"><i class="fa fa-check"></i><b>17.10.1</b> Introduction</a></li>
<li class="chapter" data-level="17.10.2" data-path="tabula-muris.html"><a href="tabula-muris.html#visual-pca"><i class="fa fa-check"></i><b>17.10.2</b> PCA plot</a></li>
<li class="chapter" data-level="17.10.3" data-path="tabula-muris.html"><a href="tabula-muris.html#visual-tsne"><i class="fa fa-check"></i><b>17.10.3</b> tSNE map</a></li>
<li class="chapter" data-level="17.10.4" data-path="tabula-muris.html"><a href="tabula-muris.html#big-exercise-1"><i class="fa fa-check"></i><b>17.10.4</b> Big Exercise</a></li>
<li class="chapter" data-level="17.10.5" data-path="tabula-muris.html"><a href="tabula-muris.html#sessioninfo-1"><i class="fa fa-check"></i><b>17.10.5</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="17.11" data-path="tabula-muris.html"><a href="tabula-muris.html#data-visualization-reads"><i class="fa fa-check"></i><b>17.11</b> Data visualization (Reads)</a></li>
<li class="chapter" data-level="17.12" data-path="tabula-muris.html"><a href="tabula-muris.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>17.12</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="17.12.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-4"><i class="fa fa-check"></i><b>17.12.1</b> Introduction</a></li>
<li class="chapter" data-level="17.12.2" data-path="tabula-muris.html"><a href="tabula-muris.html#correlations-with-pcs"><i class="fa fa-check"></i><b>17.12.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="17.12.3" data-path="tabula-muris.html"><a href="tabula-muris.html#explanatory-variables"><i class="fa fa-check"></i><b>17.12.3</b> Explanatory variables</a></li>
<li class="chapter" data-level="17.12.4" data-path="tabula-muris.html"><a href="tabula-muris.html#other-confounders"><i class="fa fa-check"></i><b>17.12.4</b> Other confounders</a></li>
<li class="chapter" data-level="17.12.5" data-path="tabula-muris.html"><a href="tabula-muris.html#exercise"><i class="fa fa-check"></i><b>17.12.5</b> Exercise</a></li>
<li class="chapter" data-level="17.12.6" data-path="tabula-muris.html"><a href="tabula-muris.html#sessioninfo-2"><i class="fa fa-check"></i><b>17.12.6</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="17.13" data-path="tabula-muris.html"><a href="tabula-muris.html#identifying-confounding-factors-reads"><i class="fa fa-check"></i><b>17.13</b> Identifying confounding factors (Reads)</a></li>
<li class="chapter" data-level="17.14" data-path="tabula-muris.html"><a href="tabula-muris.html#normalization-theory"><i class="fa fa-check"></i><b>17.14</b> Normalization theory</a><ul>
<li class="chapter" data-level="17.14.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-5"><i class="fa fa-check"></i><b>17.14.1</b> Introduction</a></li>
<li class="chapter" data-level="17.14.2" data-path="tabula-muris.html"><a href="tabula-muris.html#library-size-1"><i class="fa fa-check"></i><b>17.14.2</b> Library size</a></li>
<li class="chapter" data-level="17.14.3" data-path="tabula-muris.html"><a href="tabula-muris.html#normalisations"><i class="fa fa-check"></i><b>17.14.3</b> Normalisations</a></li>
<li class="chapter" data-level="17.14.4" data-path="tabula-muris.html"><a href="tabula-muris.html#effectiveness"><i class="fa fa-check"></i><b>17.14.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="17.15" data-path="tabula-muris.html"><a href="tabula-muris.html#normalization-practice-umi"><i class="fa fa-check"></i><b>17.15</b> Normalization practice (UMI)</a><ul>
<li class="chapter" data-level="17.15.1" data-path="tabula-muris.html"><a href="tabula-muris.html#raw"><i class="fa fa-check"></i><b>17.15.1</b> Raw</a></li>
<li class="chapter" data-level="17.15.2" data-path="tabula-muris.html"><a href="tabula-muris.html#cpm-1"><i class="fa fa-check"></i><b>17.15.2</b> CPM</a></li>
<li class="chapter" data-level="17.15.3" data-path="tabula-muris.html"><a href="tabula-muris.html#size-factor-rle"><i class="fa fa-check"></i><b>17.15.3</b> Size-factor (RLE)</a></li>
<li class="chapter" data-level="17.15.4" data-path="tabula-muris.html"><a href="tabula-muris.html#upperquantile"><i class="fa fa-check"></i><b>17.15.4</b> Upperquantile</a></li>
<li class="chapter" data-level="17.15.5" data-path="tabula-muris.html"><a href="tabula-muris.html#tmm-1"><i class="fa fa-check"></i><b>17.15.5</b> TMM</a></li>
<li class="chapter" data-level="17.15.6" data-path="tabula-muris.html"><a href="tabula-muris.html#scran-1"><i class="fa fa-check"></i><b>17.15.6</b> scran</a></li>
<li class="chapter" data-level="17.15.7" data-path="tabula-muris.html"><a href="tabula-muris.html#downsampling-1"><i class="fa fa-check"></i><b>17.15.7</b> Downsampling</a></li>
<li class="chapter" data-level="17.15.8" data-path="tabula-muris.html"><a href="tabula-muris.html#normalisation-for-genetranscript-length"><i class="fa fa-check"></i><b>17.15.8</b> Normalisation for gene/transcript length</a></li>
<li class="chapter" data-level="17.15.9" data-path="tabula-muris.html"><a href="tabula-muris.html#exercise-1"><i class="fa fa-check"></i><b>17.15.9</b> Exercise</a></li>
<li class="chapter" data-level="17.15.10" data-path="tabula-muris.html"><a href="tabula-muris.html#sessioninfo-3"><i class="fa fa-check"></i><b>17.15.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="17.16" data-path="tabula-muris.html"><a href="tabula-muris.html#normalization-practice-reads"><i class="fa fa-check"></i><b>17.16</b> Normalization practice (Reads)</a></li>
<li class="chapter" data-level="17.17" data-path="tabula-muris.html"><a href="tabula-muris.html#dealing-with-confounders"><i class="fa fa-check"></i><b>17.17</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="17.17.1" data-path="tabula-muris.html"><a href="tabula-muris.html#introduction-6"><i class="fa fa-check"></i><b>17.17.1</b> Introduction</a></li>
<li class="chapter" data-level="17.17.2" data-path="tabula-muris.html"><a href="tabula-muris.html#remove-unwanted-variation"><i class="fa fa-check"></i><b>17.17.2</b> Remove Unwanted Variation</a></li>
<li class="chapter" data-level="17.17.3" data-path="tabula-muris.html"><a href="tabula-muris.html#combat"><i class="fa fa-check"></i><b>17.17.3</b> Combat</a></li>
<li class="chapter" data-level="17.17.4" data-path="tabula-muris.html"><a href="tabula-muris.html#mnncorrect"><i class="fa fa-check"></i><b>17.17.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="17.17.5" data-path="tabula-muris.html"><a href="tabula-muris.html#glm"><i class="fa fa-check"></i><b>17.17.5</b> GLM</a></li>
<li class="chapter" data-level="17.17.6" data-path="tabula-muris.html"><a href="tabula-muris.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>17.17.6</b> How to evaluate and compare confounder removal strategies</a></li>
<li class="chapter" data-level="17.17.7" data-path="tabula-muris.html"><a href="tabula-muris.html#big-exercise-2"><i class="fa fa-check"></i><b>17.17.7</b> Big Exercise</a></li>
<li class="chapter" data-level="17.17.8" data-path="tabula-muris.html"><a href="tabula-muris.html#sessioninfo-4"><i class="fa fa-check"></i><b>17.17.8</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="17.18" data-path="tabula-muris.html"><a href="tabula-muris.html#dealing-with-confounders-reads"><i class="fa fa-check"></i><b>17.18</b> Dealing with confounders (Reads)</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>18</b> Biological Analysis</a><ul>
<li class="chapter" data-level="18.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-introduction"><i class="fa fa-check"></i><b>18.1</b> Clustering Introduction</a><ul>
<li class="chapter" data-level="18.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-7"><i class="fa fa-check"></i><b>18.1.1</b> Introduction</a></li>
<li class="chapter" data-level="18.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reductions"><i class="fa fa-check"></i><b>18.1.2</b> Dimensionality reductions</a></li>
<li class="chapter" data-level="18.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>18.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="18.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>18.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="18.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#tools-for-scrna-seq-data"><i class="fa fa-check"></i><b>18.1.5</b> Tools for scRNA-seq data</a></li>
<li class="chapter" data-level="18.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-clustering"><i class="fa fa-check"></i><b>18.1.6</b> Comparing clustering</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>18.2</b> Clustering example</a><ul>
<li class="chapter" data-level="18.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>18.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="18.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>18.2.2</b> SC3</a></li>
<li class="chapter" data-level="18.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#pcareduce-1"><i class="fa fa-check"></i><b>18.2.3</b> pcaReduce</a></li>
<li class="chapter" data-level="18.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>18.2.4</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="18.2.5" data-path="biological-analysis.html"><a href="biological-analysis.html#snn-cliq-1"><i class="fa fa-check"></i><b>18.2.5</b> SNN-Cliq</a></li>
<li class="chapter" data-level="18.2.6" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>18.2.6</b> SINCERA</a></li>
<li class="chapter" data-level="18.2.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-5"><i class="fa fa-check"></i><b>18.2.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>18.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="18.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>18.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="18.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>18.3.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="18.3.3" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-methods"><i class="fa fa-check"></i><b>18.3.3</b> Comparing Methods</a></li>
<li class="chapter" data-level="18.3.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-6"><i class="fa fa-check"></i><b>18.3.4</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="biological-analysis.html"><a href="biological-analysis.html#pseudotime-analysis"><i class="fa fa-check"></i><b>18.4</b> Pseudotime analysis</a><ul>
<li class="chapter" data-level="18.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>18.4.1</b> TSCAN</a></li>
<li class="chapter" data-level="18.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>18.4.2</b> monocle</a></li>
<li class="chapter" data-level="18.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>18.4.3</b> Diffusion maps</a></li>
<li class="chapter" data-level="18.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>18.4.4</b> SLICER</a></li>
<li class="chapter" data-level="18.4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>18.4.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="18.4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>18.4.6</b> Expression of genes through time</a></li>
<li class="chapter" data-level="18.4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-7"><i class="fa fa-check"></i><b>18.4.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>18.5</b> Imputation</a><ul>
<li class="chapter" data-level="18.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>18.5.1</b> scImpute</a></li>
<li class="chapter" data-level="18.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>18.5.2</b> MAGIC</a></li>
<li class="chapter" data-level="18.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-8"><i class="fa fa-check"></i><b>18.5.3</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="18.6" data-path="biological-analysis.html"><a href="biological-analysis.html#dechapter"><i class="fa fa-check"></i><b>18.6</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="18.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>18.6.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="18.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>18.6.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="18.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>18.6.3</b> Differences in Distribution</a></li>
<li class="chapter" data-level="18.6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rnaseq-data"><i class="fa fa-check"></i><b>18.6.4</b> Models of single-cell RNASeq data</a></li>
</ul></li>
<li class="chapter" data-level="18.7" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-a-real-dataset"><i class="fa fa-check"></i><b>18.7</b> DE in a real dataset</a><ul>
<li class="chapter" data-level="18.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-8"><i class="fa fa-check"></i><b>18.7.1</b> Introduction</a></li>
<li class="chapter" data-level="18.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>18.7.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="18.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>18.7.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="18.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>18.7.4</b> edgeR</a></li>
<li class="chapter" data-level="18.7.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>18.7.5</b> Monocle</a></li>
<li class="chapter" data-level="18.7.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>18.7.6</b> MAST</a></li>
<li class="chapter" data-level="18.7.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-methods-1h-to-run"><i class="fa fa-check"></i><b>18.7.7</b> Slow Methods (&gt;1h to run)</a></li>
<li class="chapter" data-level="18.7.8" data-path="biological-analysis.html"><a href="biological-analysis.html#bpsc"><i class="fa fa-check"></i><b>18.7.8</b> BPSC</a></li>
<li class="chapter" data-level="18.7.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>18.7.9</b> SCDE</a></li>
<li class="chapter" data-level="18.7.10" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-9"><i class="fa fa-check"></i><b>18.7.10</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html"><i class="fa fa-check"></i><b>19</b> Comparing/Combining scRNASeq datasets</a><ul>
<li class="chapter" data-level="19.1" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#introduction-9"><i class="fa fa-check"></i><b>19.1</b> Introduction</a></li>
<li class="chapter" data-level="19.2" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#datasets"><i class="fa fa-check"></i><b>19.2</b> Datasets</a></li>
<li class="chapter" data-level="19.3" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#projecting-cells-onto-annotated-cell-types-scmap"><i class="fa fa-check"></i><b>19.3</b> Projecting cells onto annotated cell-types (scmap)</a><ul>
<li class="chapter" data-level="19.3.1" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#cell-to-cell-mapping"><i class="fa fa-check"></i><b>19.3.1</b> Cell-to-Cell mapping</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#metaneighbour"><i class="fa fa-check"></i><b>19.4</b> Metaneighbour</a><ul>
<li class="chapter" data-level="19.4.1" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#prepare-data"><i class="fa fa-check"></i><b>19.4.1</b> Prepare Data</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#mnncorrect-1"><i class="fa fa-check"></i><b>19.5</b> mnnCorrect</a></li>
<li class="chapter" data-level="19.6" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#cannonical-correlation-analysis-seurat"><i class="fa fa-check"></i><b>19.6</b> Cannonical Correlation Analysis (Seurat)</a><ul>
<li class="chapter" data-level="19.6.1" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#sessioninfo-10"><i class="fa fa-check"></i><b>19.6.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="19.7" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#search-scrna-seq-data"><i class="fa fa-check"></i><b>19.7</b> Search scRNA-Seq data</a><ul>
<li class="chapter" data-level="19.7.1" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#about"><i class="fa fa-check"></i><b>19.7.1</b> About</a></li>
<li class="chapter" data-level="19.7.2" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#dataset"><i class="fa fa-check"></i><b>19.7.2</b> Dataset</a></li>
<li class="chapter" data-level="19.7.3" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#gene-index"><i class="fa fa-check"></i><b>19.7.3</b> Gene Index</a></li>
<li class="chapter" data-level="19.7.4" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#marker-genes"><i class="fa fa-check"></i><b>19.7.4</b> Marker genes</a></li>
<li class="chapter" data-level="19.7.5" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#search-cells-by-a-gene-list"><i class="fa fa-check"></i><b>19.7.5</b> Search cells by a gene list</a></li>
<li class="chapter" data-level="19.7.6" data-path="comparingcombining-scrnaseq-datasets.html"><a href="comparingcombining-scrnaseq-datasets.html#sessioninfo-11"><i class="fa fa-check"></i><b>19.7.6</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="seurat-chapter.html"><a href="seurat-chapter.html"><i class="fa fa-check"></i><b>20</b> Seurat</a><ul>
<li class="chapter" data-level="20.1" data-path="seurat-chapter.html"><a href="seurat-chapter.html#seurat-object-class"><i class="fa fa-check"></i><b>20.1</b> <code>Seurat</code> object class</a></li>
<li class="chapter" data-level="20.2" data-path="seurat-chapter.html"><a href="seurat-chapter.html#expression-qc"><i class="fa fa-check"></i><b>20.2</b> Expression QC</a></li>
<li class="chapter" data-level="20.3" data-path="seurat-chapter.html"><a href="seurat-chapter.html#normalization"><i class="fa fa-check"></i><b>20.3</b> Normalization</a></li>
<li class="chapter" data-level="20.4" data-path="seurat-chapter.html"><a href="seurat-chapter.html#highly-variable-genes-1"><i class="fa fa-check"></i><b>20.4</b> Highly variable genes</a></li>
<li class="chapter" data-level="20.5" data-path="seurat-chapter.html"><a href="seurat-chapter.html#dealing-with-confounders-1"><i class="fa fa-check"></i><b>20.5</b> Dealing with confounders</a></li>
<li class="chapter" data-level="20.6" data-path="seurat-chapter.html"><a href="seurat-chapter.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>20.6</b> Linear dimensionality reduction</a></li>
<li class="chapter" data-level="20.7" data-path="seurat-chapter.html"><a href="seurat-chapter.html#significant-pcs"><i class="fa fa-check"></i><b>20.7</b> Significant PCs</a></li>
<li class="chapter" data-level="20.8" data-path="seurat-chapter.html"><a href="seurat-chapter.html#clustering-cells"><i class="fa fa-check"></i><b>20.8</b> Clustering cells</a></li>
<li class="chapter" data-level="20.9" data-path="seurat-chapter.html"><a href="seurat-chapter.html#marker-genes-1"><i class="fa fa-check"></i><b>20.9</b> Marker genes</a></li>
<li class="chapter" data-level="20.10" data-path="seurat-chapter.html"><a href="seurat-chapter.html#sessioninfo-12"><i class="fa fa-check"></i><b>20.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><i class="fa fa-check"></i><b>21</b> “Ideal” scRNAseq pipeline (as of Oct 2017)</a><ul>
<li class="chapter" data-level="21.1" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#experimental-design"><i class="fa fa-check"></i><b>21.1</b> Experimental Design</a></li>
<li class="chapter" data-level="21.2" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#processing-reads"><i class="fa fa-check"></i><b>21.2</b> Processing Reads</a></li>
<li class="chapter" data-level="21.3" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#preparing-expression-matrix"><i class="fa fa-check"></i><b>21.3</b> Preparing Expression Matrix</a></li>
<li class="chapter" data-level="21.4" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#biological-interpretation"><i class="fa fa-check"></i><b>21.4</b> Biological Interpretation</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="advanced-exercises.html"><a href="advanced-exercises.html"><i class="fa fa-check"></i><b>22</b> Advanced exercises</a></li>
<li class="chapter" data-level="23" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>23</b> Resources</a><ul>
<li class="chapter" data-level="23.1" data-path="resources.html"><a href="resources.html#scrna-seq-protocols"><i class="fa fa-check"></i><b>23.1</b> scRNA-seq protocols</a></li>
<li class="chapter" data-level="23.2" data-path="resources.html"><a href="resources.html#external-rna-control-consortium-ercc"><i class="fa fa-check"></i><b>23.2</b> External RNA Control Consortium (ERCC)</a></li>
<li class="chapter" data-level="23.3" data-path="resources.html"><a href="resources.html#scrna-seq-analysis-tools"><i class="fa fa-check"></i><b>23.3</b> scRNA-seq analysis tools</a></li>
<li class="chapter" data-level="23.4" data-path="resources.html"><a href="resources.html#scrna-seq-public-datasets"><i class="fa fa-check"></i><b>23.4</b> scRNA-seq public datasets</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="references-4.html"><a href="references-4.html"><i class="fa fa-check"></i><b>24</b> References</a></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab, 2017</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="an-introduction-to-ggplot2" class="section level1">
<h1><span class="header-section-number">16</span> An Introduction to ggplot2</h1>
<div id="what-is-ggplot2" class="section level2">
<h2><span class="header-section-number">16.1</span> What is ggplot2?</h2>
<p>ggplot2 is an R package designed by Hadley Wickham which facilitates data plotting. In this lab, we will touch briefly on some of the features of the package. If you would like to learn more about how to use ggplot2, we would recommend reading “ggplot2 Elegant graphics for data analysis”, by Hadley Wickham.</p>
</div>
<div id="principles-of-ggplot2" class="section level2">
<h2><span class="header-section-number">16.2</span> Principles of ggplot2</h2>
<ul>
<li>Your data must be a dataframe if you want to plot it using ggplot2.</li>
<li>Use the <code>aes</code> mapping function to specify how variables in the dataframe map to features on your plot</li>
<li>Use geoms to specify how your data should be represented on your graph eg. as a scatterplot, a barplot, a boxplot etc.</li>
</ul>
</div>
<div id="using-the-aes-mapping-function" class="section level2">
<h2><span class="header-section-number">16.3</span> Using the <code>aes</code> mapping function</h2>
<p>The <code>aes</code> function specifies how variables in your dataframe map to features on your plot. To understand how this works, let’s look at an example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(tidyverse)
<span class="kw">set.seed</span>(<span class="dv">1</span>)
counts &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">100</span>, <span class="dt">lambda =</span> <span class="dv">10</span>), <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">nrow=</span><span class="dv">10</span>))
Gene_ids &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">colnames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
counts&lt;-<span class="kw">data.frame</span>(Gene_ids, counts)
counts</code></pre></div>
<pre><code>##    Gene_ids cell1 cell2 cell3 cell4 cell5 cell6 cell7 cell8 cell9 cell10
## 1     gene1     8     8     3     5     5     9    11     9    13      6
## 2     gene2    10     2    11    13    12    12     7    13    12     15
## 3     gene3     7     8    13     8     9     9     9     5    15     12
## 4     gene4    11    10     7    13    12    12    12     8    11     12
## 5     gene5    14     7     8     9    11    10    13    13     5     11
## 6     gene6    12    12    11    15     8     7    10     9    10     15
## 7     gene7    11    11    14    11    11     5     9    13    13      7
## 8     gene8     9    12     9     8     6    14     7    12    12     10
## 9     gene9    14    12    11     7    10    10     8    14     7     10
## 10   gene10    11    10     9     7    11    16     8     7     7      4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> cell1, <span class="dt">y =</span> cell2))</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>Let’s take a closer look at the final command, <code>ggplot(data = counts, mapping = aes(x = cell1, y = cell2))</code>. <code>ggplot()</code> initialises a ggplot object and takes the arguments <code>data</code> and <code>mapping</code>. We pass our dataframe of counts to <code>data</code> and use the <code>aes()</code> function to specify that we would like to use the variable cell1 as our x variable and the variable cell2 as our y variable.</p>
<p>Task 1: Modify the command above to initialise a ggplot object where cell10 is the x variable and cell8 is the y variable.</p>
<p>Clearly, the plots we have just created are not very informative because no data is displayed on them. To display data, we will need to use geoms.</p>
</div>
<div id="geoms" class="section level2">
<h2><span class="header-section-number">16.4</span> Geoms</h2>
<p>We can use geoms to specify how we would like data to be displayed on our graphs. For example, our choice of geom could specify that we would like our data to be displayed as a scatterplot, a barplot or a boxplot.</p>
<p>Let’s see how our graph would look as a scatterplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> cell1, <span class="dt">y =</span> cell2)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>Now we can see that there doesn’t seem to be any correlation between gene expression in cell1 and cell2. Given we generated <code>counts</code> randomly, this isn’t too surprising.</p>
<p>Task 2: Modify the command above to create a line plot. Hint: execute <code>?ggplot</code> and scroll down the help page. At the bottom is a link to the ggplot package index. Scroll through the index until you find the geom options.</p>
</div>
<div id="plotting-data-from-more-than-2-cells" class="section level2">
<h2><span class="header-section-number">16.5</span> Plotting data from more than 2 cells</h2>
<p>So far we’ve been considering the gene counts from 2 of the cells in our dataframe. But there are actually 10 cells in our dataframe and it would be nice to compare all of them. What if we wanted to plot data from all 10 cells at the same time?</p>
<p>At the moment we can’t do this because we are treating each individual cell as a variable and assigning that variable to either the x or the y axis. We could create a 10 dimensional graph to plot data from all 10 cells on, but this is a) not possible to do with ggplot and b) not very easy to interpret. What we could do instead is to tidy our data so that we had one variable representing cell ID and another variable representing gene counts, and plot those against each other. In code, this would look like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts&lt;-<span class="kw">gather</span>(counts, <span class="kw">colnames</span>(counts)[<span class="dv">2</span><span class="op">:</span><span class="dv">11</span>], <span class="dt">key =</span> <span class="st">&#39;Cell_ID&#39;</span>, <span class="dt">value=</span><span class="st">&#39;Counts&#39;</span>)
<span class="kw">head</span>(counts)</code></pre></div>
<pre><code>##   Gene_ids Cell_ID Counts
## 1    gene1   cell1      8
## 2    gene2   cell1     10
## 3    gene3   cell1      7
## 4    gene4   cell1     11
## 5    gene5   cell1     14
## 6    gene6   cell1     12</code></pre>
<p>Essentially, the problem before was that our data was not tidy because one variable (Cell_ID) was spread over multiple columns. Now that we’ve fixed this problem, it is much easier for us to plot data from all 10 cells on one graph.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(counts,<span class="kw">aes</span>(<span class="dt">x=</span>Cell_ID, <span class="dt">y=</span>Counts)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Task 3: Use the updated <code>counts</code> dataframe to plot a barplot with Cell_ID as the x variable and Counts as the y variable. Hint: you may find it helpful to read <code>?geom_bar</code>.</p>
<p>Task 4: Use the updated <code>counts</code> dataframe to plot a scatterplot with Gene_ids as the x variable and Counts as the y variable.</p>
</div>
<div id="plotting-heatmaps" class="section level2">
<h2><span class="header-section-number">16.6</span> Plotting heatmaps</h2>
<p>A common method for visualising gene expression data is with a heatmap. Here we will use the R package <code>pheatmap</code> to perform this analysis with some gene expression data we will name <code>test</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(pheatmap)
<span class="kw">set.seed</span>(<span class="dv">2</span>)
test =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">200</span>), <span class="dv">20</span>, <span class="dv">10</span>)
test[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">2</span>)] =<span class="st"> </span>test[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">2</span>)] <span class="op">+</span><span class="st"> </span><span class="dv">3</span>
test[<span class="dv">11</span><span class="op">:</span><span class="dv">20</span>, <span class="kw">seq</span>(<span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">2</span>)] =<span class="st"> </span>test[<span class="dv">11</span><span class="op">:</span><span class="dv">20</span>, <span class="kw">seq</span>(<span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">2</span>)] <span class="op">+</span><span class="st"> </span><span class="dv">2</span>
test[<span class="dv">15</span><span class="op">:</span><span class="dv">20</span>, <span class="kw">seq</span>(<span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">2</span>)] =<span class="st"> </span>test[<span class="dv">15</span><span class="op">:</span><span class="dv">20</span>, <span class="kw">seq</span>(<span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">2</span>)] <span class="op">+</span><span class="st"> </span><span class="dv">4</span>
<span class="kw">colnames</span>(test) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">rownames</span>(test) =<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">pheatmap</span>(test)</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Let’s take a moment to work out what this graphic is showing us. Each row represents a gene and each column represents a cell. How highly expressed each gene is in each cell is represented by the colour of the corresponding box. For example, we can tell from this plot that gene18 is highly expressed in cell10 but lowly expressed in cell1.</p>
<p>This plot also gives us information on the results of a clustering algorithm. In general, clustering algorithms aim to split datapoints (eg.cells) into groups whose members are more alike one another than they are alike the rest of the datapoints. The trees drawn on the top and left hand sides of the graph are the results of clustering algorithms and enable us to see, for example, that cells 4,8,2,6 and 10 are more alike one another than they are alike cells 7,3,5,1 and 9. The tree on the left hand side of the graph represents the results of a clustering algorithm applied to the genes in our dataset.</p>
<p>If we look closely at the trees, we can see that eventually they have the same number of branches as there are cells and genes. In other words, the total number of cell clusters is the same as the total number of cells, and the total number of gene clusters is the same as the total number of genes. Clearly, this is not very informative, and will become impractical when we are looking at more than 10 cells and 20 genes. Fortunately, we can set the number of clusters we see on the plot. Let’s try setting the number of gene clusters to 2:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pheatmap</span>(test, <span class="dt">kmeans_k =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Now we can see that the genes fall into two clusters - a cluster of 8 genes which are upregulated in cells 2, 10, 6, 4 and 8 relative to the other cells and a cluster of 12 genes which are downregulated in cells 2, 10, 6, 4 and 8 relative to the other cells.</p>
<p>Task 5: Try setting the number of clusters to 3. Which number of clusters do you think is more informative?</p>
</div>
<div id="principle-component-analysis" class="section level2">
<h2><span class="header-section-number">16.7</span> Principle Component Analysis</h2>
<p>Principal component analysis (PCA) is a statistical procedure that uses a transformation to convert a set of observations into a set of values of linearly uncorrelated variables called principal components. The transformation is carried out so that the first principle component accounts for as much of the variability in the data as possible, and each following principle component accounts for the greatest amount of variance possible under the contraint that it must be orthogonal to the previous components.</p>
<p>PCA plots are a good way to get an overview of your data, and can sometimes help identify confounders which explain a high amount of the variability in your data. We will investigate how we can use PCA plots in single-cell RNA-seq analysis in more depth in a future lab, here the aim is to give you an overview of what PCA plots are and how they are generated.</p>
<p>Let’s make a PCA plot for our <code>test</code> data. We can use the <code>ggfortify</code> package to let ggplot know how to interpret principle components.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggfortify)
Principle_Components&lt;-<span class="kw">prcomp</span>(test)
<span class="kw">autoplot</span>(Principle_Components, <span class="dt">label=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="12-L4-ggplot2-pheatmap-PCA_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Task 6: Compare your clusters to the pheatmap clusters. Are they related? (Hint: have a look at the gene tree for the first pheatmap we plotted)</p>
<p>Task 7: Produce a heatmap and PCA plot for <code>counts</code> (below):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
counts &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">100</span>, <span class="dt">lambda =</span> <span class="dv">10</span>), <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">nrow=</span><span class="dv">10</span>))
<span class="kw">rownames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">colnames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-types.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="tabula-muris.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
