<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data</title>
  <meta name="description" content="Analysis of single cell RNA-seq data">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data" />
  
  
  

<meta name="author" content="Vladimir Kiselev (wikiselev), Tallulah Andrews, Davis McCarthy (davisjmcc), Maren Büttner (marenbuettner) and Martin Hemberg (m_hemberg)">


<meta name="date" content="2017-11-18">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-single-cell-rna-seq.html">
<link rel="next" href="biological-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#video"><i class="fa fa-check"></i><b>1.1</b> Video</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#r-based"><i class="fa fa-check"></i><b>1.4</b> R-based</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#docker-image-rstudio"><i class="fa fa-check"></i><b>1.5</b> Docker image (RStudio)</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#manual-installation"><i class="fa fa-check"></i><b>1.6</b> Manual installation</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.7</b> License</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.8</b> Prerequisites</a></li>
<li class="chapter" data-level="1.9" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.9</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>3</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="3.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-qc"><i class="fa fa-check"></i><b>3.1</b> Reads QC</a></li>
<li class="chapter" data-level="3.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>3.2</b> Reads alignment</a></li>
<li class="chapter" data-level="3.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-example"><i class="fa fa-check"></i><b>3.3</b> Alignment example</a></li>
<li class="chapter" data-level="3.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>3.4</b> Mapping QC</a></li>
<li class="chapter" data-level="3.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>3.5</b> Reads quantification</a></li>
<li class="chapter" data-level="3.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>3.6</b> Unique Molecular Identifiers (UMIs)</a><ul>
<li class="chapter" data-level="3.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>3.6.1</b> Introduction</a></li>
<li class="chapter" data-level="3.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-barcodes"><i class="fa fa-check"></i><b>3.6.2</b> Mapping Barcodes</a></li>
<li class="chapter" data-level="3.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#counting-barcodes"><i class="fa fa-check"></i><b>3.6.3</b> Counting Barcodes</a></li>
<li class="chapter" data-level="3.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correcting-for-errors"><i class="fa fa-check"></i><b>3.6.4</b> Correcting for Errors</a></li>
<li class="chapter" data-level="3.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downstream-analysis"><i class="fa fa-check"></i><b>3.6.5</b> Downstream Analysis</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor-singlecellexperiment-and-scater"><i class="fa fa-check"></i><b>3.7</b> Bioconductor, <code>SingleCellExperiment</code> and <code>scater</code></a><ul>
<li class="chapter" data-level="3.7.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor"><i class="fa fa-check"></i><b>3.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="3.7.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>3.7.2</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="3.7.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scater-package"><i class="fa fa-check"></i><b>3.7.3</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>3.8</b> Expression QC (UMI)</a><ul>
<li class="chapter" data-level="3.8.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>3.8.1</b> Introduction</a></li>
<li class="chapter" data-level="3.8.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tung-dataset"><i class="fa fa-check"></i><b>3.8.2</b> Tung dataset</a></li>
<li class="chapter" data-level="3.8.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>3.8.3</b> Cell QC</a></li>
<li class="chapter" data-level="3.8.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>3.8.4</b> Cell filtering</a></li>
<li class="chapter" data-level="3.8.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>3.8.5</b> Compare filterings</a></li>
<li class="chapter" data-level="3.8.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>3.8.6</b> Gene analysis</a></li>
<li class="chapter" data-level="3.8.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#save-the-data"><i class="fa fa-check"></i><b>3.8.7</b> Save the data</a></li>
<li class="chapter" data-level="3.8.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise"><i class="fa fa-check"></i><b>3.8.8</b> Big Exercise</a></li>
<li class="chapter" data-level="3.8.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo"><i class="fa fa-check"></i><b>3.8.9</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#expression-qc-reads"><i class="fa fa-check"></i><b>3.9</b> Expression QC (Reads)</a></li>
<li class="chapter" data-level="3.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>3.10</b> Data visualization</a><ul>
<li class="chapter" data-level="3.10.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>3.10.1</b> Introduction</a></li>
<li class="chapter" data-level="3.10.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>3.10.2</b> PCA plot</a></li>
<li class="chapter" data-level="3.10.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>3.10.3</b> tSNE map</a></li>
<li class="chapter" data-level="3.10.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise-1"><i class="fa fa-check"></i><b>3.10.4</b> Big Exercise</a></li>
<li class="chapter" data-level="3.10.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-1"><i class="fa fa-check"></i><b>3.10.5</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization-reads"><i class="fa fa-check"></i><b>3.11</b> Data visualization (Reads)</a></li>
<li class="chapter" data-level="3.12" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>3.12</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="3.12.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>3.12.1</b> Introduction</a></li>
<li class="chapter" data-level="3.12.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>3.12.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="3.12.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>3.12.3</b> Explanatory variables</a></li>
<li class="chapter" data-level="3.12.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#other-confounders"><i class="fa fa-check"></i><b>3.12.4</b> Other confounders</a></li>
<li class="chapter" data-level="3.12.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise"><i class="fa fa-check"></i><b>3.12.5</b> Exercise</a></li>
<li class="chapter" data-level="3.12.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-2"><i class="fa fa-check"></i><b>3.12.6</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.13" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors-reads"><i class="fa fa-check"></i><b>3.13</b> Identifying confounding factors (Reads)</a></li>
<li class="chapter" data-level="3.14" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>3.14</b> Normalization theory</a><ul>
<li class="chapter" data-level="3.14.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>3.14.1</b> Introduction</a></li>
<li class="chapter" data-level="3.14.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>3.14.2</b> Library size</a></li>
<li class="chapter" data-level="3.14.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>3.14.3</b> Normalisations</a></li>
<li class="chapter" data-level="3.14.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness"><i class="fa fa-check"></i><b>3.14.4</b> Effectiveness</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-umi"><i class="fa fa-check"></i><b>3.15</b> Normalization practice (UMI)</a><ul>
<li class="chapter" data-level="3.15.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw"><i class="fa fa-check"></i><b>3.15.1</b> Raw</a></li>
<li class="chapter" data-level="3.15.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cpm-1"><i class="fa fa-check"></i><b>3.15.2</b> CPM</a></li>
<li class="chapter" data-level="3.15.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#size-factor-rle"><i class="fa fa-check"></i><b>3.15.3</b> Size-factor (RLE)</a></li>
<li class="chapter" data-level="3.15.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#upperquantile"><i class="fa fa-check"></i><b>3.15.4</b> Upperquantile</a></li>
<li class="chapter" data-level="3.15.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tmm-1"><i class="fa fa-check"></i><b>3.15.5</b> TMM</a></li>
<li class="chapter" data-level="3.15.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scran-1"><i class="fa fa-check"></i><b>3.15.6</b> scran</a></li>
<li class="chapter" data-level="3.15.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downsampling-1"><i class="fa fa-check"></i><b>3.15.7</b> Downsampling</a></li>
<li class="chapter" data-level="3.15.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalisation-for-genetranscript-length"><i class="fa fa-check"></i><b>3.15.8</b> Normalisation for gene/transcript length</a></li>
<li class="chapter" data-level="3.15.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise-1"><i class="fa fa-check"></i><b>3.15.9</b> Exercise</a></li>
<li class="chapter" data-level="3.15.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-3"><i class="fa fa-check"></i><b>3.15.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.16" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-reads"><i class="fa fa-check"></i><b>3.16</b> Normalization practice (Reads)</a></li>
<li class="chapter" data-level="3.17" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>3.17</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="3.17.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-5"><i class="fa fa-check"></i><b>3.17.1</b> Introduction</a></li>
<li class="chapter" data-level="3.17.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#remove-unwanted-variation"><i class="fa fa-check"></i><b>3.17.2</b> Remove Unwanted Variation</a></li>
<li class="chapter" data-level="3.17.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#combat"><i class="fa fa-check"></i><b>3.17.3</b> Combat</a></li>
<li class="chapter" data-level="3.17.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>3.17.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="3.17.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#glm"><i class="fa fa-check"></i><b>3.17.5</b> GLM</a></li>
<li class="chapter" data-level="3.17.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>3.17.6</b> How to evaluate and compare confounder removal strategies</a></li>
<li class="chapter" data-level="3.17.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise-2"><i class="fa fa-check"></i><b>3.17.7</b> Big Exercise</a></li>
<li class="chapter" data-level="3.17.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-4"><i class="fa fa-check"></i><b>3.17.8</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.18" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#dealing-with-confounders-reads"><i class="fa fa-check"></i><b>3.18</b> Dealing with confounders (Reads)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>4</b> Biological Analysis</a><ul>
<li class="chapter" data-level="4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-introduction"><i class="fa fa-check"></i><b>4.1</b> Clustering Introduction</a><ul>
<li class="chapter" data-level="4.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>4.1.1</b> Introduction</a></li>
<li class="chapter" data-level="4.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reductions"><i class="fa fa-check"></i><b>4.1.2</b> Dimensionality reductions</a></li>
<li class="chapter" data-level="4.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>4.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="4.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>4.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="4.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#tools-for-scrna-seq-data"><i class="fa fa-check"></i><b>4.1.5</b> Tools for scRNA-seq data</a></li>
<li class="chapter" data-level="4.1.6" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-clustering"><i class="fa fa-check"></i><b>4.1.6</b> Comparing clustering</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>4.2</b> Clustering example</a><ul>
<li class="chapter" data-level="4.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>4.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="4.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>4.2.2</b> SC3</a></li>
<li class="chapter" data-level="4.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#pcareduce-1"><i class="fa fa-check"></i><b>4.2.3</b> pcaReduce</a></li>
<li class="chapter" data-level="4.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>4.2.4</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="4.2.5" data-path="biological-analysis.html"><a href="biological-analysis.html#snn-cliq-1"><i class="fa fa-check"></i><b>4.2.5</b> SNN-Cliq</a></li>
<li class="chapter" data-level="4.2.6" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>4.2.6</b> SINCERA</a></li>
<li class="chapter" data-level="4.2.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-5"><i class="fa fa-check"></i><b>4.2.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>4.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="4.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>4.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="4.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>4.3.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="4.3.3" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-methods"><i class="fa fa-check"></i><b>4.3.3</b> Comparing Methods</a></li>
<li class="chapter" data-level="4.3.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-6"><i class="fa fa-check"></i><b>4.3.4</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#pseudotime-analysis"><i class="fa fa-check"></i><b>4.4</b> Pseudotime analysis</a><ul>
<li class="chapter" data-level="4.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>4.4.1</b> TSCAN</a></li>
<li class="chapter" data-level="4.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>4.4.2</b> monocle</a></li>
<li class="chapter" data-level="4.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>4.4.3</b> Diffusion maps</a></li>
<li class="chapter" data-level="4.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>4.4.4</b> SLICER</a></li>
<li class="chapter" data-level="4.4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>4.4.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="4.4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>4.4.6</b> Expression of genes through time</a></li>
<li class="chapter" data-level="4.4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-7"><i class="fa fa-check"></i><b>4.4.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>4.5</b> Imputation</a><ul>
<li class="chapter" data-level="4.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>4.5.1</b> scImpute</a></li>
<li class="chapter" data-level="4.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>4.5.2</b> MAGIC</a></li>
<li class="chapter" data-level="4.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-8"><i class="fa fa-check"></i><b>4.5.3</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#dechapter"><i class="fa fa-check"></i><b>4.6</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="4.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>4.6.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="4.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>4.6.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="4.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>4.6.3</b> Differences in Distribution</a></li>
<li class="chapter" data-level="4.6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rnaseq-data"><i class="fa fa-check"></i><b>4.6.4</b> Models of single-cell RNASeq data</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-a-real-dataset"><i class="fa fa-check"></i><b>4.7</b> DE in a real dataset</a><ul>
<li class="chapter" data-level="4.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-7"><i class="fa fa-check"></i><b>4.7.1</b> Introduction</a></li>
<li class="chapter" data-level="4.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>4.7.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="4.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>4.7.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="4.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>4.7.4</b> edgeR</a></li>
<li class="chapter" data-level="4.7.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>4.7.5</b> Monocle</a></li>
<li class="chapter" data-level="4.7.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>4.7.6</b> MAST</a></li>
<li class="chapter" data-level="4.7.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-methods-1h-to-run"><i class="fa fa-check"></i><b>4.7.7</b> Slow Methods (&gt;1h to run)</a></li>
<li class="chapter" data-level="4.7.8" data-path="biological-analysis.html"><a href="biological-analysis.html#bpsc"><i class="fa fa-check"></i><b>4.7.8</b> BPSC</a></li>
<li class="chapter" data-level="4.7.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>4.7.9</b> SCDE</a></li>
<li class="chapter" data-level="4.7.10" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-9"><i class="fa fa-check"></i><b>4.7.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="biological-analysis.html"><a href="biological-analysis.html#projecting-scrna-seq-data"><i class="fa fa-check"></i><b>4.8</b> Projecting scRNA-seq data</a><ul>
<li class="chapter" data-level="4.8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#datasets"><i class="fa fa-check"></i><b>4.8.1</b> Datasets</a></li>
<li class="chapter" data-level="4.8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#run-scmap"><i class="fa fa-check"></i><b>4.8.2</b> Run <code>scmap</code></a></li>
<li class="chapter" data-level="4.8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#results"><i class="fa fa-check"></i><b>4.8.3</b> Results</a></li>
<li class="chapter" data-level="4.8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#creating-a-precomputed-reference"><i class="fa fa-check"></i><b>4.8.4</b> Creating a precomputed Reference</a></li>
<li class="chapter" data-level="4.8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-10"><i class="fa fa-check"></i><b>4.8.5</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="biological-analysis.html"><a href="biological-analysis.html#search-scrna-seq-data"><i class="fa fa-check"></i><b>4.9</b> Search scRNA-Seq data</a><ul>
<li class="chapter" data-level="4.9.1" data-path="biological-analysis.html"><a href="biological-analysis.html#about"><i class="fa fa-check"></i><b>4.9.1</b> About</a></li>
<li class="chapter" data-level="4.9.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dataset"><i class="fa fa-check"></i><b>4.9.2</b> Dataset</a></li>
<li class="chapter" data-level="4.9.3" data-path="biological-analysis.html"><a href="biological-analysis.html#gene-index"><i class="fa fa-check"></i><b>4.9.3</b> Gene Index</a></li>
<li class="chapter" data-level="4.9.4" data-path="biological-analysis.html"><a href="biological-analysis.html#marker-genes"><i class="fa fa-check"></i><b>4.9.4</b> Marker genes</a></li>
<li class="chapter" data-level="4.9.5" data-path="biological-analysis.html"><a href="biological-analysis.html#search-cells-by-a-gene-list"><i class="fa fa-check"></i><b>4.9.5</b> Search cells by a gene list</a></li>
<li class="chapter" data-level="4.9.6" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-11"><i class="fa fa-check"></i><b>4.9.6</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="seurat-chapter.html"><a href="seurat-chapter.html"><i class="fa fa-check"></i><b>5</b> Seurat</a><ul>
<li class="chapter" data-level="5.1" data-path="seurat-chapter.html"><a href="seurat-chapter.html#seurat-object-class"><i class="fa fa-check"></i><b>5.1</b> <code>Seurat</code> object class</a></li>
<li class="chapter" data-level="5.2" data-path="seurat-chapter.html"><a href="seurat-chapter.html#expression-qc"><i class="fa fa-check"></i><b>5.2</b> Expression QC</a></li>
<li class="chapter" data-level="5.3" data-path="seurat-chapter.html"><a href="seurat-chapter.html#normalization"><i class="fa fa-check"></i><b>5.3</b> Normalization</a></li>
<li class="chapter" data-level="5.4" data-path="seurat-chapter.html"><a href="seurat-chapter.html#highly-variable-genes-1"><i class="fa fa-check"></i><b>5.4</b> Highly variable genes</a></li>
<li class="chapter" data-level="5.5" data-path="seurat-chapter.html"><a href="seurat-chapter.html#dealing-with-confounders-1"><i class="fa fa-check"></i><b>5.5</b> Dealing with confounders</a></li>
<li class="chapter" data-level="5.6" data-path="seurat-chapter.html"><a href="seurat-chapter.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>5.6</b> Linear dimensionality reduction</a></li>
<li class="chapter" data-level="5.7" data-path="seurat-chapter.html"><a href="seurat-chapter.html#significant-pcs"><i class="fa fa-check"></i><b>5.7</b> Significant PCs</a></li>
<li class="chapter" data-level="5.8" data-path="seurat-chapter.html"><a href="seurat-chapter.html#clustering-cells"><i class="fa fa-check"></i><b>5.8</b> Clustering cells</a></li>
<li class="chapter" data-level="5.9" data-path="seurat-chapter.html"><a href="seurat-chapter.html#marker-genes-1"><i class="fa fa-check"></i><b>5.9</b> Marker genes</a></li>
<li class="chapter" data-level="5.10" data-path="seurat-chapter.html"><a href="seurat-chapter.html#sessioninfo-12"><i class="fa fa-check"></i><b>5.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><i class="fa fa-check"></i><b>6</b> “Ideal” scRNAseq pipeline (as of Oct 2017)</a><ul>
<li class="chapter" data-level="6.1" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#experimental-design"><i class="fa fa-check"></i><b>6.1</b> Experimental Design</a></li>
<li class="chapter" data-level="6.2" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#processing-reads"><i class="fa fa-check"></i><b>6.2</b> Processing Reads</a></li>
<li class="chapter" data-level="6.3" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#preparing-expression-matrix"><i class="fa fa-check"></i><b>6.3</b> Preparing Expression Matrix</a></li>
<li class="chapter" data-level="6.4" data-path="ideal-scrnaseq-pipeline-as-of-oct-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-oct-2017.html#biological-interpretation"><i class="fa fa-check"></i><b>6.4</b> Biological Interpretation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-exercises.html"><a href="advanced-exercises.html"><i class="fa fa-check"></i><b>7</b> Advanced exercises</a></li>
<li class="chapter" data-level="8" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>8</b> Resources</a><ul>
<li class="chapter" data-level="8.1" data-path="resources.html"><a href="resources.html#scrna-seq-protocols"><i class="fa fa-check"></i><b>8.1</b> scRNA-seq protocols</a></li>
<li class="chapter" data-level="8.2" data-path="resources.html"><a href="resources.html#external-rna-control-consortium-ercc"><i class="fa fa-check"></i><b>8.2</b> External RNA Control Consortium (ERCC)</a></li>
<li class="chapter" data-level="8.3" data-path="resources.html"><a href="resources.html#scrna-seq-analysis-tools"><i class="fa fa-check"></i><b>8.3</b> scRNA-seq analysis tools</a></li>
<li class="chapter" data-level="8.4" data-path="resources.html"><a href="resources.html#scrna-seq-public-datasets"><i class="fa fa-check"></i><b>8.4</b> scRNA-seq public datasets</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>9</b> References</a></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab, 2017</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="construction-of-expression-matrix" class="section level1">
<h1><span class="header-section-number">3</span> Construction of expression matrix</h1>
<p>Many analyses of scRNA-seq data take as their starting point an <strong>expression matrix</strong>. By convention, the each row of the expression matrix represents a gene and each column represents a cell (although some authors use the transpose). Each entry represents the expression level of a particular gene in a given cell. The units by which the expression is meassured depends on the protocol and the normalization strategy used.</p>
<div id="reads-qc" class="section level2">
<h2><span class="header-section-number">3.1</span> Reads QC</h2>
<p>The output from a scRNA-seq experiment is a large collection of cDNA reads. The first step is to ensure that the reads are of high quality. The quality control can be performed by using standard tools, such as <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> or <a href="http://www.ebi.ac.uk/research/enright/software/kraken">Kraken</a>.</p>
<p>Assuming that our reads are in experiment.bam, we run FastQC as</p>
<pre><code>$&lt;path_to_fastQC&gt;/fastQC experiment.bam</code></pre>
<p>Below is an example of the output from FastQC for a dataset of 125 bp reads. The plot reveals a technical error which resulted in a couple of bases failing to be read correctly in the centre of the read. However, since the rest of the read was of high quality this error will most likely have a negligible effect on mapping efficiency.</p>
<div class="figure" style="text-align: center"><span id="fig:exprs-constr-fastqc"></span>
<img src="figures/per_base_quality.png" alt="Example of FastQC output" width="90%" />
<p class="caption">
Figure 3.1: Example of FastQC output
</p>
</div>
<p>Additionally, it is often helpful to visualize the data using the <a href="https://www.broadinstitute.org/igv/">Integrative Genomics Browser (IGV)</a> or <a href="http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/">SeqMonk</a>.</p>
</div>
<div id="reads-alignment" class="section level2">
<h2><span class="header-section-number">3.2</span> Reads alignment</h2>
<p>After trimming low quality bases from the reads, the remaining sequences can be mapped to a reference genome. Again, there is no need for a special purpose method for this, so we can use the <a href="https://github.com/alexdobin/STAR">STAR</a> or the <a href="https://ccb.jhu.edu/software/tophat/index.shtml">TopHat</a> aligner. For large full-transcript datasets from well annotated organisms (e.g. mouse, human) pseudo-alignment methods (e.g. <a href="https://pachterlab.github.io/kallisto/">Kallisto</a>, <a href="http://salmon.readthedocs.io/en/latest/salmon.html">Salmon</a>) may out-perform conventional alignment. For drop-seq based datasets with tens- or hundreds of thousands of reads pseudoaligners become more appealing since their run-time can be several orders of magnitude less than traditional aligners.</p>
<p>An example of how to map reads.bam to using STAR is</p>
<pre><code>$&lt;path_to_STAR&gt;/STAR --runThreadN 1 --runMode alignReads
--readFilesIn reads1.fq.gz reads2.fq.gz --readFilesCommand zcat --genomeDir &lt;path&gt;
--parametersFiles FileOfMoreParameters.txt --outFileNamePrefix &lt;outpath&gt;/output</code></pre>
<p><strong>Note</strong>, if the <em>spike-ins</em> are used, the reference sequence should be augmented with the DNA sequence of the <em>spike-in</em> molecules prior to mapping.</p>
<p><strong>Note</strong>, when UMIs are used, their barcodes should be removed from the read sequence. A common practice is to add the barcode to the read name.</p>
<p>Once the reads for each cell have been mapped to the reference genome, we need to make sure that a sufficient number of reads from each cell could be mapped to the reference genome. In our experience, the fraction of mappable reads for mouse or human cells is 60-70%. However, this result may vary depending on protocol, read length and settings for the read alignment. As a general rule, we expect all cells to have a similar fraction of mapped reads, so any outliers should be inspected and possibly removed. A low proportion of mappable reads usually indicates contamination.</p>
<p>An example of how to quantify expression using Salmon is</p>
<pre><code>$&lt;path_to_Salmon&gt;/salmon quant -i salmon_transcript_index -1 reads1.fq.gz -2 reads2.fq.gz -p #threads -l A -g genome.gtf --seqBias --gcBias --posBias</code></pre>
<p><strong>Note</strong> Salmon produces estimated read counts and estimated transcripts per million (tpm) in our experience the latter over corrects the expression of long genes for scRNASeq, thus we recommend using read counts.</p>
</div>
<div id="alignment-example" class="section level2">
<h2><span class="header-section-number">3.3</span> Alignment example</h2>
<p>The histogram below shows the total number of reads mapped to each cell for an scRNA-seq experiment. Each bar represents one cell, and they have been sorted in ascending order by the total number of reads per cell. The three red arrows indicate cells that are outliers in terms of their coverage and they should be removed from further analysis. The two yellow arrows point to cells with a surprisingly large number of unmapped reads. In this example we kept the cells during the alignment QC step, but they were later removed during cell QC due to a high proportion of ribosomal RNA reads.</p>
<div class="figure" style="text-align: center"><span id="fig:exprs-constr-total-num-cells"></span>
<img src="figures/Bergiers_exp1_mapping_by_cell.png" alt="Example of the total number of reads mapped to each cell." width="90%" />
<p class="caption">
Figure 3.2: Example of the total number of reads mapped to each cell.
</p>
</div>
</div>
<div id="mapping-qc" class="section level2">
<h2><span class="header-section-number">3.4</span> Mapping QC</h2>
<p>After mapping the raw sequencing to the genome we need to evaluate the quality of the mapping. There are many ways to measure the mapping quality, including: amount of reads mapping to rRNA/tRNAs, proportion of uniquely mapping reads, reads mapping across splice junctions, read depth along the transcripts. Methods developed for bulk RNA-seq, such as <a href="http://rseqc.sourceforge.net/">RSeQC</a>, are applicable to single-cell data:</p>
<pre><code>python &lt;RSeQCpath&gt;/geneBody_coverage.py -i input.bam -r genome.bed -o output.txt
python &lt;RSeQCpath&gt;/bam_stat.py -i input.bam -r genome.bed -o output.txt
python &lt;RSeQCpath&gt;/split_bam.py -i input.bam -r rRNAmask.bed -o output.txt</code></pre>
<p>However the expected results will depend on the experimental protocol, e.g. many scRNA-seq methods use poly-A selection to avoid sequencing rRNAs which results in a 3’ bias in the read coverage across the genes (aka gene body coverage). The figure below shows this 3’ bias as well as three cells which were outliers and removed from the dataset:</p>
<div class="figure" style="text-align: center"><span id="fig:exprs-constr-3-bias"></span>
<img src="figures/Exp1_RSEQC_geneBodyCoverage_plot_Combined.png" alt="Example of the 3' bias in the read coverage." width="90%" />
<p class="caption">
Figure 3.3: Example of the 3’ bias in the read coverage.
</p>
</div>
</div>
<div id="reads-quantification" class="section level2">
<h2><span class="header-section-number">3.5</span> Reads quantification</h2>
<p>The next step is to quantify the expression level of each gene for each cell. For mRNA data, we can use one of the tools which has been developed for bulk RNA-seq data, e.g. <a href="http://www-huber.embl.de/users/anders/HTSeq/">HT-seq</a> or <a href="http://subread.sourceforge.net/">FeatureCounts</a></p>
<pre><code># include multimapping
&lt;featureCounts_path&gt;/featureCounts -O -M -Q 30 -p -a genome.gtf -o outputfile input.bam
# exclude multimapping
&lt;featureCounts_path&gt;/featureCounts -Q 30 -p -a genome.gtf -o outputfile input.bam</code></pre>
<p><a href="http://www.nature.com/nmeth/journal/v9/n1/full/nmeth.1778.html">Unique molecular identifiers (UMIs)</a> make it possible to count the absolute number of molecules and they have proven popular for <a href="http://www.nature.com/nmeth/journal/v11/n2/full/nmeth.2772.html">scRNA-seq</a>. We will discuss how UMIs can be processed in the next chapter.</p>

</div>
<div id="umichapter" class="section level2">
<h2><span class="header-section-number">3.6</span> Unique Molecular Identifiers (UMIs)</h2>
<p>Thanks to Andreas Buness from <a href="https://www.embl.it/services/bioinformatics/">EMBL Monterotondo</a> for collaboration on this section.</p>
<div id="introduction" class="section level3">
<h3><span class="header-section-number">3.6.1</span> Introduction</h3>
<p>Unique Molecular Identifiers are short (4-10bp) random barcodes added to transcripts during reverse-transcription. They enable sequencing reads to be assigned to individual transcript molecules and thus the removal of amplification noise and biases from scRNASeq data.</p>
<div class="figure" style="text-align: center"><span id="fig:intro-umi-protocol"></span>
<img src="figures/UMI-Seq-protocol.png" alt="UMI sequencing protocol" width="90%" />
<p class="caption">
Figure 3.4: UMI sequencing protocol
</p>
</div>
<p>When sequencing UMI containing data, techniques are used to specifically sequence only the end of the transcript containing the UMI (usually the 3’ end).</p>
</div>
<div id="mapping-barcodes" class="section level3">
<h3><span class="header-section-number">3.6.2</span> Mapping Barcodes</h3>
<p>Since the number of unique barcodes (<span class="math inline">\(4^N\)</span>, where <span class="math inline">\(N\)</span> is the length of UMI) is much smaller than the total number of molecules per cell (~<span class="math inline">\(10^6\)</span>), each barcode will typically be assigned to multiple transcripts. Hence, to identify unique molecules both barcode and mapping location (transcript) must be used. The first step is to map UMI reads, for which we recommend using STAR since it is fast and outputs good quality BAM-alignments. Moreover, mapping locations can be useful for eg. identifying poorly-annotated 3’ UTRs of transcripts.</p>
<p>UMI-sequencing typically consists of paired-end reads where one read from each pair captures the cell and UMI barcodes while the other read consists of exonic sequence from the transcript (Figure <a href="construction-of-expression-matrix.html#fig:intro-umi-reads">3.5</a>). Note that trimming and/or filtering to remove reads containing poly-A sequence is recommended to avoid erors due to these read mapping to genes/transcripts with internal poly-A/poly-T sequences.</p>
<p>After processing the reads from a UMI experiment, the following conventions are often used:</p>
<ol style="list-style-type: decimal">
<li><p>The UMI is added to the read name of the other paired read.</p></li>
<li>Reads are sorted into separate files by cell barcode
<ul>
<li>For extremely large, shallow datasets, the cell barcode may be added to the read name as well to reduce the number of files.</li>
</ul></li>
</ol>
<div class="figure" style="text-align: center"><span id="fig:intro-umi-reads"></span>
<img src="figures/UMI-Seq-reads.png" alt="UMI sequencing reads, red lightning bolts represent different fragmentation locations" width="90%" />
<p class="caption">
Figure 3.5: UMI sequencing reads, red lightning bolts represent different fragmentation locations
</p>
</div>
</div>
<div id="counting-barcodes" class="section level3">
<h3><span class="header-section-number">3.6.3</span> Counting Barcodes</h3>
<p>In theory, every unique UMI-transcript pair should represent all reads originating from a single RNA molecule. However, in practice this is frequently not the case and the most common reasons are:</p>
<ol style="list-style-type: decimal">
<li><strong>Different UMI does not necessarily mean different molecule</strong>
<ul>
<li>Due to PCR or sequencing errors, base-pair substitution events can result in new UMI sequences. Longer UMIs give more opportunity for errors to arise and based on estimates from cell barcodes we expect 7-10% of 10bp UMIs to contain at least one error. If not corrected for, this type of error will result in an overestimate of the number of transcripts.</li>
</ul></li>
<li><strong>Different transcript does not necessarily mean different molecule</strong>
<ul>
<li>Mapping errors and/or multimapping reads may result in some UMIs being assigned to the wrong gene/transcript. This type of error will also result in an overestimate of the number of transcripts.</li>
</ul></li>
<li><strong>Same UMI does not necessarily mean same molecule</strong>
<ul>
<li>Biases in UMI frequency and short UMIs can result in the same UMI being attached to different mRNA molecules from the same gene. Thus, the number of transcripts may be underestimated.</li>
</ul></li>
</ol>
<div class="figure" style="text-align: center"><span id="fig:intro-umi-errors"></span>
<img src="figures/UMI-Seq-errors.png" alt="Potential Errors in UMIs" width="90%" />
<p class="caption">
Figure 3.6: Potential Errors in UMIs
</p>
</div>
</div>
<div id="correcting-for-errors" class="section level3">
<h3><span class="header-section-number">3.6.4</span> Correcting for Errors</h3>
<p>How to best account for errors in UMIs remains an active area of research. The best approaches that we are aware of for resolving the issues mentioned above are:</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://github.com/CGATOxford/UMI-tools">UMI-tools’</a> directional-adjacency method implements a procedure which considers both the number of mismatches and the relative frequency of similar UMIs to identify likely PCR/sequencing errors.</p></li>
<li><p>Currently an open question. The problem may be mitigated by removing UMIs with few reads to support their association with a particular transcript, or by removing all multi-mapping reads.</p></li>
<li><p>Simple saturation (aka “collision probability”) correction proposed by <a href="http://www.nature.com/nmeth/journal/v11/n6/full/nmeth.2930.html#methods">Grun, Kester and van Oudenaarden (2014)</a> to estimate the true number of molecules <span class="math inline">\(M\)</span>:</p></li>
</ol>
<p><span class="math display">\[M \approx -N*log(1 - \frac{n}{N})\]</span> where N = total number of unique UMI barcodes and n = number of observed barcodes.</p>
<p>An important caveat of this method is that it assumes that all UMIs are equally frequent. In most cases this is incorrect, since there is often a bias related to the GC content.</p>
<div class="figure" style="text-align: center"><span id="fig:intro-umi-amp"></span>
<img src="figures/UMI-Seq-amp.png" alt="Per gene amplification rate" width="60%" />
<p class="caption">
Figure 3.7: Per gene amplification rate
</p>
</div>
<p>Determining how to best process and use UMIs is currently an active area of research in the bioinformatics community. We are aware of several methods that have recently been developed, including:</p>
<ul>
<li><a href="https://github.com/CGATOxford/UMI-tools">UMI-tools</a></li>
<li><a href="https://github.com/tallulandrews/PoissonUMIs">PoissonUMIs</a></li>
<li><a href="https://github.com/sdparekh/zUMIs">zUMIs</a></li>
<li><a href="https://github.com/hms-dbmi/dropEst">dropEst</a></li>
</ul>
</div>
<div id="downstream-analysis" class="section level3">
<h3><span class="header-section-number">3.6.5</span> Downstream Analysis</h3>
<p>Current UMI platforms (DropSeq, InDrop, ICell8) exhibit low and highly variable capture efficiency as shown in the figure below.</p>
<div class="figure" style="text-align: center"><span id="fig:intro-umi-capture"></span>
<img src="figures/UMI-Seq-capture.png" alt="Variability in Capture Efficiency" width="70%" />
<p class="caption">
Figure 3.8: Variability in Capture Efficiency
</p>
</div>
<p>This variability can introduce strong biases and it needs to be considered in downstream analysis. Recent analyses often pool cells/genes together based on cell-type or biological pathway to increase the power. Robust statistical analyses of this data is still an open research question and it remains to be determined how to best adjust for biases.</p>
<p><strong>Exercise 1</strong> We have provided you with UMI counts and read counts from induced pluripotent stem cells generated from three different individuals <span class="citation">(Tung et al. <a href="#ref-Tung2017-ba">2017</a>)</span> (see: Chapter <a href="construction-of-expression-matrix.html#exprs-qc">3.8</a> for details of this dataset).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi_counts &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/molecules.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
read_counts &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/reads.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>Using this data:</p>
<ol style="list-style-type: decimal">
<li><p>Plot the variability in capture efficiency</p></li>
<li><p>Determine the amplification rate: average number of reads per UMI.</p></li>
</ol>

</div>
</div>
<div id="bioconductor-singlecellexperiment-and-scater" class="section level2">
<h2><span class="header-section-number">3.7</span> Bioconductor, <code>SingleCellExperiment</code> and <code>scater</code></h2>
<div id="bioconductor" class="section level3">
<h3><span class="header-section-number">3.7.1</span> Bioconductor</h3>
<p>From <a href="https://en.wikipedia.org/wiki/Bioconductor">Wikipedia</a>: <a href="https://www.bioconductor.org/">Bioconductor</a> is a free, open source and open development software project for the analysis and comprehension of genomic data generated by wet lab experiments in molecular biology. Bioconductor is based primarily on the statistical R programming language, but does contain contributions in other programming languages. It has two releases each year that follow the semiannual releases of R. At any one time there is a release version, which corresponds to the released version of R, and a development version, which corresponds to the development version of R. Most users will find the release version appropriate for their needs.</p>
<p>We strongly recommend all new comers and even experienced high-throughput data analysts to use well developed and maintained <a href="https://www.bioconductor.org/developers/how-to/commonMethodsAndClasses/">Bioconductor methods and classes</a>.</p>
</div>
<div id="singlecellexperiment-class" class="section level3">
<h3><span class="header-section-number">3.7.2</span> <code>SingleCellExperiment</code> class</h3>
<p><a href="http://bioconductor.org/packages/SingleCellExperiment"><code>SingleCellExperiment</code></a> (SCE) is a S4 class for storing data from single-cell experiments. This includes specialized methods to store and retrieve spike-in information, dimensionality reduction coordinates and size factors for each cell, along with the usual metadata for genes and libraries.</p>
<p>In practice, an object of this class can be created using its constructor:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)
counts &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">100</span>, <span class="dt">lambda =</span> <span class="dv">10</span>), <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">nrow=</span><span class="dv">10</span>)
<span class="kw">rownames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">colnames</span>(counts) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;cell&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(
    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> counts),
    <span class="dt">rowData =</span> <span class="kw">data.frame</span>(<span class="dt">gene_names =</span> <span class="kw">paste</span>(<span class="st">&quot;gene_name&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)),
    <span class="dt">colData =</span> <span class="kw">data.frame</span>(<span class="dt">cell_names =</span> <span class="kw">paste</span>(<span class="st">&quot;cell_name&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>))
)
sce</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 10 10 
## metadata(0):
## assays(1): counts
## rownames(10): gene1 gene2 ... gene9 gene10
## rowData names(1): gene_names
## colnames(10): cell1 cell2 ... cell9 cell10
## colData names(1): cell_names
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>In the <code>SingleCellExperiment</code>, users can assign arbitrary names to entries of assays. To assist interoperability between packages, some suggestions for what the names should be for particular types of data are provided by the authors:</p>
<ul>
<li><strong>counts</strong>: Raw count data, e.g., number of reads or transcripts for a particular gene.</li>
<li><strong>normcounts</strong>: Normalized values on the same scale as the original counts. For example, counts divided by cell-specific size factors that are centred at unity.</li>
<li><strong>logcounts</strong>: Log-transformed counts or count-like values. In most cases, this will be defined as log-transformed normcounts, e.g., using log base 2 and a pseudo-count of 1.</li>
<li><strong>cpm</strong>: Counts-per-million. This is the read count for each gene in each cell, divided by the library size of each cell in millions.</li>
<li><strong>tpm</strong>: Transcripts-per-million. This is the number of transcripts for each gene in each cell, divided by the total number of transcripts in that cell (in millions).</li>
</ul>
<p>Each of these suggested names has an appropriate getter/setter method for convenient manipulation of the <code>SingleCellExperiment</code>. For example, we can take the (very specifically named) <code>counts</code> slot, normalise it and assign it to <code>normcounts</code> instead:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">normcounts</span>(sce) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">counts</span>(sce) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
sce</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 10 10 
## metadata(0):
## assays(2): counts normcounts
## rownames(10): gene1 gene2 ... gene9 gene10
## rowData names(1): gene_names
## colnames(10): cell1 cell2 ... cell9 cell10
## colData names(1): cell_names
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(<span class="kw">normcounts</span>(sce))</code></pre></div>
<pre><code>## [1] 10 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">normcounts</span>(sce))</code></pre></div>
<pre><code>##          cell1    cell2    cell3    cell4    cell5    cell6    cell7
## gene1 3.169925 3.169925 2.000000 2.584963 2.584963 3.321928 3.584963
## gene2 3.459432 1.584963 3.584963 3.807355 3.700440 3.700440 3.000000
## gene3 3.000000 3.169925 3.807355 3.169925 3.321928 3.321928 3.321928
## gene4 3.584963 3.459432 3.000000 3.807355 3.700440 3.700440 3.700440
## gene5 3.906891 3.000000 3.169925 3.321928 3.584963 3.459432 3.807355
## gene6 3.700440 3.700440 3.584963 4.000000 3.169925 3.000000 3.459432
##          cell8    cell9   cell10
## gene1 3.321928 3.807355 2.807355
## gene2 3.807355 3.700440 4.000000
## gene3 2.584963 4.000000 3.700440
## gene4 3.169925 3.584963 3.700440
## gene5 3.807355 2.584963 3.584963
## gene6 3.321928 3.459432 4.000000</code></pre>
</div>
<div id="scater-package" class="section level3">
<h3><span class="header-section-number">3.7.3</span> <code>scater</code> package</h3>
<p><a href="http://bioconductor.org/packages/scater/"><code>scater</code></a> is a R package for single-cell RNA-seq analysis <span class="citation">(McCarthy et al. <a href="#ref-McCarthy2017-kb">2017</a>)</span>. The package contains several useful methods for quality control, visualisation and pre-processing of data prior to further downstream analysis.</p>
<p><code>scater</code> features the following functionality:</p>
<ul>
<li>Automated computation of QC metrics</li>
<li>Transcript quantification from read data with pseudo-alignment</li>
<li>Data format standardisation</li>
<li>Rich visualizations for exploratory analysis</li>
<li>Seamless integration into the Bioconductor universe</li>
<li>Simple normalisation methods</li>
</ul>
<p>We highly recommend to use <code>scater</code> for all single-cell RNA-seq analyses and <code>scater</code> is the basis of the first part of the course.</p>
<p>As illustrated in the figure below, <code>scater</code> will help you with quality control, filtering and normalization of your expression matrix following mapping and alignment. <span style="color:red">Keep in mind that this figure represents the original version of <code>scater</code> where an <code>SCESet</code> class was used. In the newest version this figure is still correct, except that <code>SCESet</code> can be substituted with the <code>SingleCellExperiment</code> class.</span></p>
<div class="figure">
<img src="figures/scater_qc_workflow.png" />

</div>

</div>
</div>
<div id="exprs-qc" class="section level2">
<h2><span class="header-section-number">3.8</span> Expression QC (UMI)</h2>
<div id="introduction-1" class="section level3">
<h3><span class="header-section-number">3.8.1</span> Introduction</h3>
<p>Once gene expression has been quantified it is summarized as an <strong>expression matrix</strong> where each row corresponds to a gene (or transcript) and each column corresponds to a single cell. This matrix should be examined to remove poor quality cells which were not detected in either read QC or mapping QC steps. Failure to remove low quality cells at this stage may add technical noise which has the potential to obscure the biological signals of interest in the downstream analysis.</p>
<p>Since there is currently no standard method for performing scRNASeq the expected values for the various QC measures that will be presented here can vary substantially from experiment to experiment. Thus, to perform QC we will be looking for cells which are outliers with respect to the rest of the dataset rather than comparing to independent quality standards. Consequently, care should be taken when comparing quality metrics across datasets collected using different protocols.</p>
</div>
<div id="tung-dataset" class="section level3">
<h3><span class="header-section-number">3.8.2</span> Tung dataset</h3>
<p>To illustrate cell QC, we consider a <a href="http://jdblischak.github.io/singleCellSeq/analysis/">dataset</a> of induced pluripotent stem cells generated from three different individuals <span class="citation">(Tung et al. <a href="#ref-Tung2017-ba">2017</a>)</span> in <a href="http://giladlab.uchicago.edu/">Yoav Gilad</a>’s lab at the University of Chicago. The experiments were carried out on the Fluidigm C1 platform and to facilitate the quantification both unique molecular identifiers (UMIs) and ERCC <em>spike-ins</em> were used. The data files are located in the <code>tung</code> folder in your working directory. These files are the copies of the original files made on the 15/03/16. We will use these copies for reproducibility purposes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scater)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Load the data and annotations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">molecules &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/molecules.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
anno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/annotation.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Inspect a small portion of the expression matrix</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(molecules[ , <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</code></pre></div>
<pre><code>##                 NA19098.r1.A01 NA19098.r1.A02 NA19098.r1.A03
## ENSG00000237683              0              0              0
## ENSG00000187634              0              0              0
## ENSG00000188976              3              6              1
## ENSG00000187961              0              0              0
## ENSG00000187583              0              0              0
## ENSG00000187642              0              0              0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(anno)</code></pre></div>
<pre><code>##   individual replicate well      batch      sample_id
## 1    NA19098        r1  A01 NA19098.r1 NA19098.r1.A01
## 2    NA19098        r1  A02 NA19098.r1 NA19098.r1.A02
## 3    NA19098        r1  A03 NA19098.r1 NA19098.r1.A03
## 4    NA19098        r1  A04 NA19098.r1 NA19098.r1.A04
## 5    NA19098        r1  A05 NA19098.r1 NA19098.r1.A05
## 6    NA19098        r1  A06 NA19098.r1 NA19098.r1.A06</code></pre>
<p>The data consists of 3 individuals and 3 replicates and therefore has 9 batches in total.</p>
<p>We standardize the analysis by using both <code>SingleCellExperiment</code> (SCE) and <code>scater</code> packages. First, create the SCE object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(
    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">as.matrix</span>(molecules)), 
    <span class="dt">colData =</span> anno
)</code></pre></div>
<p>Remove genes that are not expressed in any cell:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">keep_feature &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">counts</span>(umi) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>
umi &lt;-<span class="st"> </span>umi[keep_feature, ]</code></pre></div>
<p>Define control features (genes) - ERCC spike-ins and mitochondrial genes (<a href="http://jdblischak.github.io/singleCellSeq/analysis/qc-filter-ipsc.html">provided</a> by the authors):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">isSpike</span>(umi, <span class="st">&quot;ERCC&quot;</span>) &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^ERCC-&quot;</span>, <span class="kw">rownames</span>(umi))
<span class="kw">isSpike</span>(umi, <span class="st">&quot;MT&quot;</span>) &lt;-<span class="st"> </span><span class="kw">rownames</span>(umi) <span class="op">%in%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">c</span>(<span class="st">&quot;ENSG00000198899&quot;</span>, <span class="st">&quot;ENSG00000198727&quot;</span>, <span class="st">&quot;ENSG00000198888&quot;</span>,
    <span class="st">&quot;ENSG00000198886&quot;</span>, <span class="st">&quot;ENSG00000212907&quot;</span>, <span class="st">&quot;ENSG00000198786&quot;</span>,
    <span class="st">&quot;ENSG00000198695&quot;</span>, <span class="st">&quot;ENSG00000198712&quot;</span>, <span class="st">&quot;ENSG00000198804&quot;</span>,
    <span class="st">&quot;ENSG00000198763&quot;</span>, <span class="st">&quot;ENSG00000228253&quot;</span>, <span class="st">&quot;ENSG00000198938&quot;</span>,
    <span class="st">&quot;ENSG00000198840&quot;</span>)</code></pre></div>
<p>Calculate the quality metrics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(
    umi,
    <span class="dt">feature_controls =</span> <span class="kw">list</span>(
        <span class="dt">ERCC =</span> <span class="kw">isSpike</span>(umi, <span class="st">&quot;ERCC&quot;</span>), 
        <span class="dt">MT =</span> <span class="kw">isSpike</span>(umi, <span class="st">&quot;MT&quot;</span>)
    )
)</code></pre></div>
</div>
<div id="cell-qc" class="section level3">
<h3><span class="header-section-number">3.8.3</span> Cell QC</h3>
<div id="library-size" class="section level4">
<h4><span class="header-section-number">3.8.3.1</span> Library size</h4>
<p>Next we consider the total number of RNA molecules detected per sample (if we were using read counts rather than UMI counts this would be the total number of reads). Wells with few reads/molecules are likely to have been broken or failed to capture a cell, and should thus be removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    umi<span class="op">$</span>total_counts,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">25000</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:total-counts-hist"></span>
<img src="07-exprs-qc_files/figure-html/total-counts-hist-1.png" alt="Histogram of library sizes for all cells" width="90%" />
<p class="caption">
Figure 3.9: Histogram of library sizes for all cells
</p>
</div>
<p><strong>Exercise 1</strong></p>
<ol style="list-style-type: decimal">
<li><p>How many cells does our filter remove?</p></li>
<li><p>What distribution do you expect that the total number of molecules for each cell should follow?</p></li>
</ol>
<p><strong>Our answer</strong></p>
<pre><code>## filter_by_total_counts
## FALSE  TRUE 
##    46   818</code></pre>
</div>
<div id="detected-genes" class="section level4">
<h4><span class="header-section-number">3.8.3.2</span> Detected genes</h4>
<p>In addition to ensuring sufficient sequencing depth for each sample, we also want to make sure that the reads are distributed across the transcriptome. Thus, we count the total number of unique genes detected in each sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    umi<span class="op">$</span>total_features,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">7000</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:total-features-hist"></span>
<img src="07-exprs-qc_files/figure-html/total-features-hist-1.png" alt="Histogram of the number of detected genes in all cells" width="90%" />
<p class="caption">
Figure 3.10: Histogram of the number of detected genes in all cells
</p>
</div>
<p>From the plot we conclude that most cells have between 7,000-10,000 detected genes, which is normal for high-depth scRNA-seq. However, this varies by experimental protocol and sequencing depth. For example, droplet-based methods or samples with lower sequencing-depth typically detect fewer genes per cell. The most notable feature in the above plot is the <strong>“heavy tail”</strong> on the left hand side of the distribution. If detection rates were equal across the cells then the distribution should be approximately normal. Thus we remove those cells in the tail of the distribution (fewer than 7,000 detected genes).</p>
<p><strong>Exercise 2</strong></p>
<p>How many cells does our filter remove?</p>
<p><strong>Our answer</strong></p>
<pre><code>## filter_by_expr_features
## FALSE  TRUE 
##   116   748</code></pre>
</div>
<div id="erccs-and-mts" class="section level4">
<h4><span class="header-section-number">3.8.3.3</span> ERCCs and MTs</h4>
<p>Another measure of cell quality is the ratio between ERCC <em>spike-in</em> RNAs and endogenous RNAs. This ratio can be used to estimate the total amount of RNA in the captured cells. Cells with a high level of <em>spike-in</em> RNAs had low starting amounts of RNA, likely due to the cell being dead or stressed which may result in the RNA being degraded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPhenoData</span>(
    umi,
    <span class="kw">aes_string</span>(
        <span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;pct_counts_MT&quot;</span>,
        <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>
    )
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:mt-vs-counts"></span>
<img src="07-exprs-qc_files/figure-html/mt-vs-counts-1.png" alt="Percentage of counts in MT genes" width="90%" />
<p class="caption">
Figure 3.11: Percentage of counts in MT genes
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPhenoData</span>(
    umi,
    <span class="kw">aes_string</span>(
        <span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;pct_counts_ERCC&quot;</span>,
        <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>
    )
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ercc-vs-counts"></span>
<img src="07-exprs-qc_files/figure-html/ercc-vs-counts-1.png" alt="Percentage of counts in ERCCs" width="90%" />
<p class="caption">
Figure 3.12: Percentage of counts in ERCCs
</p>
</div>
<p>The above analysis shows that majority of the cells from NA19098.r2 batch have a very high ERCC/Endo ratio. Indeed, it has been shown by the authors that this batch contains cells of smaller size.</p>
<p><strong>Exercise 3</strong></p>
<p>Create filters for removing batch NA19098.r2 and cells with high expression of mitochondrial genes (&gt;10% of total counts in a cell).</p>
<p><strong>Our answer</strong></p>
<pre><code>## filter_by_ERCC
## FALSE  TRUE 
##    96   768</code></pre>
<pre><code>## filter_by_MT
## FALSE  TRUE 
##    31   833</code></pre>
<p><strong>Exercise 4</strong></p>
<p>What would you expect to see in the ERCC vs counts plot if you were examining a dataset containing cells of different sizes (eg. normal &amp; senescent cells)?</p>
<p><strong>Answer</strong></p>
<p>You would expect to see a group corresponding to the smaller cells (normal) with a higher fraction of ERCC reads than a separate group corresponding to the larger cells (senescent).</p>
</div>
</div>
<div id="cell-filtering" class="section level3">
<h3><span class="header-section-number">3.8.4</span> Cell filtering</h3>
<div id="manual" class="section level4">
<h4><span class="header-section-number">3.8.4.1</span> Manual</h4>
<p>Now we can define a cell filter based on our previous analysis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi<span class="op">$</span>use &lt;-<span class="st"> </span>(
    <span class="co"># sufficient features (genes)</span>
    filter_by_expr_features <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># sufficient molecules counted</span>
<span class="st">    </span>filter_by_total_counts <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># sufficient endogenous RNA</span>
<span class="st">    </span>filter_by_ERCC <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># remove cells with unusual number of reads in MT genes</span>
<span class="st">    </span>filter_by_MT
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(umi<span class="op">$</span>use)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   207   657</code></pre>
</div>
<div id="automatic" class="section level4">
<h4><span class="header-section-number">3.8.4.2</span> Automatic</h4>
<p>Another option available in <code>scater</code> is to conduct PCA on a set of QC metrics and then use automatic outlier detection to identify potentially problematic cells.</p>
<p>By default, the following metrics are used for PCA-based outlier detection:</p>
<ul>
<li><strong>pct_counts_top_100_features</strong></li>
<li><strong>total_features</strong></li>
<li><strong>pct_counts_feature_controls</strong></li>
<li><strong>n_detected_feature_controls</strong></li>
<li><strong>log10_counts_endogenous_features</strong></li>
<li><strong>log10_counts_feature_controls</strong></li>
</ul>
<p><code>scater</code> first creates a matrix where the rows represent cells and the columns represent the different QC metrics. Here, the PCA plot provides a 2D representation of cells ordered by their quality metrics. The outliers are then detected using methods from the mvoutlier package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(
    umi,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>, 
    <span class="dt">shape_by =</span> <span class="st">&quot;use&quot;</span>,
    <span class="dt">pca_data_input =</span> <span class="st">&quot;pdata&quot;</span>,
    <span class="dt">detect_outliers =</span> <span class="ot">TRUE</span>,
    <span class="dt">return_SCE =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:auto-cell-filt"></span>
<img src="07-exprs-qc_files/figure-html/auto-cell-filt-1.png" alt="PCA plot used for automatic detection of cell outliers" width="90%" />
<p class="caption">
Figure 3.13: PCA plot used for automatic detection of cell outliers
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(umi<span class="op">$</span>outlier)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   819    45</code></pre>
</div>
</div>
<div id="compare-filterings" class="section level3">
<h3><span class="header-section-number">3.8.5</span> Compare filterings</h3>
<p><strong>Exercise 5</strong></p>
<p>Compare the default, automatic and manual cell filters. Plot a Venn diagram of the outlier cells from these filterings.</p>
<p><strong>Hint</strong>: Use <code>vennCounts</code> and <code>vennDiagram</code> functions from the <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> package to make a Venn diagram.</p>
<p><strong>Answer</strong></p>
<pre><code>## 
## Attaching package: &#39;limma&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:scater&#39;:
## 
##     plotMDS</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     plotMA</code></pre>
<div class="figure" style="text-align: center"><span id="fig:cell-filt-comp"></span>
<img src="07-exprs-qc_files/figure-html/cell-filt-comp-1.png" alt="Comparison of the default, automatic and manual cell filters" width="90%" />
<p class="caption">
Figure 3.14: Comparison of the default, automatic and manual cell filters
</p>
</div>
</div>
<div id="gene-analysis" class="section level3">
<h3><span class="header-section-number">3.8.6</span> Gene analysis</h3>
<div id="gene-expression" class="section level4">
<h4><span class="header-section-number">3.8.6.1</span> Gene expression</h4>
<p>In addition to removing cells with poor quality, it is usually a good idea to exclude genes where we suspect that technical artefacts may have skewed the results. Moreover, inspection of the gene expression profiles may provide insights about how the experimental procedures could be improved.</p>
<p>It is often instructive to consider the number of reads consumed by the top 50 expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotQC</span>(umi, <span class="dt">type =</span> <span class="st">&quot;highest-expression&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:top50-gene-expr"></span>
<img src="07-exprs-qc_files/figure-html/top50-gene-expr-1.png" alt="Number of total counts consumed by the top 50 expressed genes" width="90%" />
<p class="caption">
Figure 3.15: Number of total counts consumed by the top 50 expressed genes
</p>
</div>
<p>The distributions are relatively flat indicating (but not guaranteeing!) good coverage of the full transcriptome of these cells. However, there are several spike-ins in the top 15 genes which suggests a greater dilution of the spike-ins may be preferrable if the experiment is to be repeated.</p>
</div>
<div id="gene-filtering" class="section level4">
<h4><span class="header-section-number">3.8.6.2</span> Gene filtering</h4>
<p>It is typically a good idea to remove genes whose expression level is considered <strong>“undetectable”</strong>. We define a gene as detectable if at least two cells contain more than 1 transcript from the gene. If we were considering read counts rather than UMI counts a reasonable threshold is to require at least five reads in at least two cells. However, in both cases the threshold strongly depends on the sequencing depth. It is important to keep in mind that genes must be filtered after cell filtering since some genes may only be detected in poor quality cells (<strong>note</strong> <code>colData(umi)$use</code> filter applied to the <code>umi</code> dataset).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_genes &lt;-<span class="st"> </span><span class="kw">apply</span>(
    <span class="kw">counts</span>(umi[ , <span class="kw">colData</span>(umi)<span class="op">$</span>use]), 
    <span class="dv">1</span>, 
    <span class="cf">function</span>(x) <span class="kw">length</span>(x[x <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>]) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span>
)
<span class="kw">rowData</span>(umi)<span class="op">$</span>use &lt;-<span class="st"> </span>filter_genes</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(filter_genes)</code></pre></div>
<pre><code>## filter_genes
## FALSE  TRUE 
##  4660 14066</code></pre>
<p>Depending on the cell-type, protocol and sequencing depth, other cut-offs may be appropriate.</p>
</div>
</div>
<div id="save-the-data" class="section level3">
<h3><span class="header-section-number">3.8.7</span> Save the data</h3>
<p>Dimensions of the QCed dataset (do not forget about the gene filter we defined above):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(umi[<span class="kw">rowData</span>(umi)<span class="op">$</span>use, <span class="kw">colData</span>(umi)<span class="op">$</span>use])</code></pre></div>
<pre><code>## [1] 14066   657</code></pre>
<p>Let’s create an additional slot with log-transformed counts (we will need it in the next chapters) and remove saved PCA results from the <code>reducedDim</code> slot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assay</span>(umi, <span class="st">&quot;logcounts_raw&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">counts</span>(umi) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">reducedDim</span>(umi) &lt;-<span class="st"> </span><span class="ot">NULL</span></code></pre></div>
<p>Save the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(umi, <span class="dt">file =</span> <span class="st">&quot;tung/umi.rds&quot;</span>)</code></pre></div>
</div>
<div id="big-exercise" class="section level3">
<h3><span class="header-section-number">3.8.8</span> Big Exercise</h3>
<p>Perform exactly the same QC analysis with read counts of the same Blischak data. Use <code>tung/reads.txt</code> file to load the reads. Once you have finished please compare your results to ours (next chapter).</p>
</div>
<div id="sessioninfo" class="section level3">
<h3><span class="header-section-number">3.8.9</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] limma_3.34.1               scater_1.6.0              
##  [3] ggplot2_2.2.1              SingleCellExperiment_1.0.0
##  [5] SummarizedExperiment_1.8.0 DelayedArray_0.4.1        
##  [7] matrixStats_0.52.2         Biobase_2.38.0            
##  [9] GenomicRanges_1.30.0       GenomeInfoDb_1.14.0       
## [11] IRanges_2.12.0             S4Vectors_0.16.0          
## [13] BiocGenerics_0.24.0        knitr_1.17                
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0        minqa_1.2.4            
##   [3] colorspace_1.3-2        mvoutlier_2.0.8        
##   [5] rjson_0.2.15            modeltools_0.2-21      
##   [7] class_7.3-14            mclust_5.3             
##   [9] rprojroot_1.2           XVector_0.18.0         
##  [11] pls_2.6-0               cvTools_0.3.2          
##  [13] MatrixModels_0.4-1      flexmix_2.3-14         
##  [15] bit64_0.9-7             AnnotationDbi_1.40.0   
##  [17] mvtnorm_1.0-6           sROC_0.1-2             
##  [19] splines_3.4.2           tximport_1.6.0         
##  [21] robustbase_0.92-8       nloptr_1.0.4           
##  [23] robCompositions_2.0.6   pbkrtest_0.4-7         
##  [25] kernlab_0.9-25          cluster_2.0.6          
##  [27] shinydashboard_0.6.1    shiny_1.0.5            
##  [29] rrcov_1.4-3             compiler_3.4.2         
##  [31] backports_1.1.1         assertthat_0.2.0       
##  [33] Matrix_1.2-7.1          lazyeval_0.2.1         
##  [35] htmltools_0.3.6         quantreg_5.34          
##  [37] prettyunits_1.0.2       tools_3.4.2            
##  [39] bindrcpp_0.2            gtable_0.2.0           
##  [41] glue_1.2.0              GenomeInfoDbData_0.99.1
##  [43] reshape2_1.4.2          dplyr_0.7.4            
##  [45] Rcpp_0.12.13            trimcluster_0.1-2      
##  [47] sgeostat_1.0-27         nlme_3.1-129           
##  [49] fpc_2.1-10              lmtest_0.9-35          
##  [51] laeken_0.4.6            stringr_1.2.0          
##  [53] lme4_1.1-14             mime_0.5               
##  [55] XML_3.98-1.9            edgeR_3.20.1           
##  [57] DEoptimR_1.0-8          zoo_1.8-0              
##  [59] zlibbioc_1.24.0         MASS_7.3-45            
##  [61] scales_0.5.0            VIM_4.7.0              
##  [63] rhdf5_2.22.0            SparseM_1.77           
##  [65] RColorBrewer_1.1-2      yaml_2.1.14            
##  [67] memoise_1.1.0           gridExtra_2.3          
##  [69] biomaRt_2.34.0          reshape_0.8.7          
##  [71] stringi_1.1.5           RSQLite_2.0            
##  [73] highr_0.6               pcaPP_1.9-72           
##  [75] e1071_1.6-8             boot_1.3-18            
##  [77] prabclus_2.2-6          rlang_0.1.4            
##  [79] pkgconfig_2.0.1         bitops_1.0-6           
##  [81] evaluate_0.10.1         lattice_0.20-34        
##  [83] bindr_0.1               labeling_0.3           
##  [85] cowplot_0.9.0           bit_1.1-12             
##  [87] GGally_1.3.2            plyr_1.8.4             
##  [89] magrittr_1.5            bookdown_0.5           
##  [91] R6_2.2.2                DBI_0.7                
##  [93] mgcv_1.8-22             RCurl_1.95-4.8         
##  [95] sp_1.2-5                nnet_7.3-12            
##  [97] tibble_1.3.4            car_2.1-5              
##  [99] rmarkdown_1.7           viridis_0.4.0          
## [101] progress_1.1.2          locfit_1.5-9.1         
## [103] grid_3.4.2              data.table_1.10.4-3    
## [105] blob_1.1.0              diptest_0.75-7         
## [107] vcd_1.4-3               digest_0.6.12          
## [109] xtable_1.8-2            httpuv_1.3.5           
## [111] munsell_0.4.3           beeswarm_0.2.3         
## [113] viridisLite_0.2.0       vipor_0.4.5</code></pre>

</div>
</div>
<div id="expression-qc-reads" class="section level2">
<h2><span class="header-section-number">3.9</span> Expression QC (Reads)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scater)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reads &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/reads.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
anno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;tung/annotation.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(reads[ , <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</code></pre></div>
<pre><code>##                 NA19098.r1.A01 NA19098.r1.A02 NA19098.r1.A03
## ENSG00000237683              0              0              0
## ENSG00000187634              0              0              0
## ENSG00000188976             57            140              1
## ENSG00000187961              0              0              0
## ENSG00000187583              0              0              0
## ENSG00000187642              0              0              0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(anno)</code></pre></div>
<pre><code>##   individual replicate well      batch      sample_id
## 1    NA19098        r1  A01 NA19098.r1 NA19098.r1.A01
## 2    NA19098        r1  A02 NA19098.r1 NA19098.r1.A02
## 3    NA19098        r1  A03 NA19098.r1 NA19098.r1.A03
## 4    NA19098        r1  A04 NA19098.r1 NA19098.r1.A04
## 5    NA19098        r1  A05 NA19098.r1 NA19098.r1.A05
## 6    NA19098        r1  A06 NA19098.r1 NA19098.r1.A06</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reads &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(
    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">as.matrix</span>(reads)), 
    <span class="dt">colData =</span> anno
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">keep_feature &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">counts</span>(reads) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>
reads &lt;-<span class="st"> </span>reads[keep_feature, ]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">isSpike</span>(reads, <span class="st">&quot;ERCC&quot;</span>) &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^ERCC-&quot;</span>, <span class="kw">rownames</span>(reads))
<span class="kw">isSpike</span>(reads, <span class="st">&quot;MT&quot;</span>) &lt;-<span class="st"> </span><span class="kw">rownames</span>(reads) <span class="op">%in%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">c</span>(<span class="st">&quot;ENSG00000198899&quot;</span>, <span class="st">&quot;ENSG00000198727&quot;</span>, <span class="st">&quot;ENSG00000198888&quot;</span>,
    <span class="st">&quot;ENSG00000198886&quot;</span>, <span class="st">&quot;ENSG00000212907&quot;</span>, <span class="st">&quot;ENSG00000198786&quot;</span>,
    <span class="st">&quot;ENSG00000198695&quot;</span>, <span class="st">&quot;ENSG00000198712&quot;</span>, <span class="st">&quot;ENSG00000198804&quot;</span>,
    <span class="st">&quot;ENSG00000198763&quot;</span>, <span class="st">&quot;ENSG00000228253&quot;</span>, <span class="st">&quot;ENSG00000198938&quot;</span>,
    <span class="st">&quot;ENSG00000198840&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reads &lt;-<span class="st"> </span><span class="kw">calculateQCMetrics</span>(
    reads,
    <span class="dt">feature_controls =</span> <span class="kw">list</span>(
        <span class="dt">ERCC =</span> <span class="kw">isSpike</span>(reads, <span class="st">&quot;ERCC&quot;</span>), 
        <span class="dt">MT =</span> <span class="kw">isSpike</span>(reads, <span class="st">&quot;MT&quot;</span>)
    )
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    reads<span class="op">$</span>total_counts,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="fl">1.3e6</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:total-counts-hist-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/total-counts-hist-reads-1.png" alt="Histogram of library sizes for all cells" width="90%" />
<p class="caption">
Figure 3.16: Histogram of library sizes for all cells
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_total_counts &lt;-<span class="st"> </span>(reads<span class="op">$</span>total_counts <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.3e6</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(filter_by_total_counts)</code></pre></div>
<pre><code>## filter_by_total_counts
## FALSE  TRUE 
##   180   684</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    reads<span class="op">$</span>total_features,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">7000</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:total-features-hist-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/total-features-hist-reads-1.png" alt="Histogram of the number of detected genes in all cells" width="90%" />
<p class="caption">
Figure 3.17: Histogram of the number of detected genes in all cells
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_expr_features &lt;-<span class="st"> </span>(reads<span class="op">$</span>total_features <span class="op">&gt;</span><span class="st"> </span><span class="dv">7000</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(filter_by_expr_features)</code></pre></div>
<pre><code>## filter_by_expr_features
## FALSE  TRUE 
##   116   748</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPhenoData</span>(
    reads,
    <span class="kw">aes_string</span>(
        <span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;pct_counts_MT&quot;</span>,
        <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>
    )
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:mt-vs-counts-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/mt-vs-counts-reads-1.png" alt="Percentage of counts in MT genes" width="90%" />
<p class="caption">
Figure 3.18: Percentage of counts in MT genes
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPhenoData</span>(
    reads,
    <span class="kw">aes_string</span>(
        <span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;pct_counts_ERCC&quot;</span>,
        <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>
    )
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ercc-vs-counts-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/ercc-vs-counts-reads-1.png" alt="Percentage of counts in ERCCs" width="90%" />
<p class="caption">
Figure 3.19: Percentage of counts in ERCCs
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_ERCC &lt;-<span class="st"> </span>
<span class="st">    </span>reads<span class="op">$</span>batch <span class="op">!=</span><span class="st"> &quot;NA19098.r2&quot;</span> <span class="op">&amp;</span><span class="st"> </span>reads<span class="op">$</span>pct_counts_ERCC <span class="op">&lt;</span><span class="st"> </span><span class="dv">25</span>
<span class="kw">table</span>(filter_by_ERCC)</code></pre></div>
<pre><code>## filter_by_ERCC
## FALSE  TRUE 
##   103   761</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_MT &lt;-<span class="st"> </span>reads<span class="op">$</span>pct_counts_MT <span class="op">&lt;</span><span class="st"> </span><span class="dv">30</span>
<span class="kw">table</span>(filter_by_MT)</code></pre></div>
<pre><code>## filter_by_MT
## FALSE  TRUE 
##    18   846</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reads<span class="op">$</span>use &lt;-<span class="st"> </span>(
    <span class="co"># sufficient features (genes)</span>
    filter_by_expr_features <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># sufficient molecules counted</span>
<span class="st">    </span>filter_by_total_counts <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># sufficient endogenous RNA</span>
<span class="st">    </span>filter_by_ERCC <span class="op">&amp;</span>
<span class="st">    </span><span class="co"># remove cells with unusual number of reads in MT genes</span>
<span class="st">    </span>filter_by_MT
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(reads<span class="op">$</span>use)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   258   606</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">reads &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(
    reads,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>, 
    <span class="dt">shape_by =</span> <span class="st">&quot;use&quot;</span>,
    <span class="dt">pca_data_input =</span> <span class="st">&quot;pdata&quot;</span>,
    <span class="dt">detect_outliers =</span> <span class="ot">TRUE</span>,
    <span class="dt">return_SCE =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:auto-cell-filt-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/auto-cell-filt-reads-1.png" alt="PCA plot used for automatic detection of cell outliers" width="90%" />
<p class="caption">
Figure 3.20: PCA plot used for automatic detection of cell outliers
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(reads<span class="op">$</span>outlier)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   756   108</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)</code></pre></div>
<pre><code>## 
## Attaching package: &#39;limma&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:scater&#39;:
## 
##     plotMDS</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     plotMA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">auto &lt;-<span class="st"> </span><span class="kw">colnames</span>(reads)[reads<span class="op">$</span>outlier]
man &lt;-<span class="st"> </span><span class="kw">colnames</span>(reads)[<span class="op">!</span>reads<span class="op">$</span>use]
venn.diag &lt;-<span class="st"> </span><span class="kw">vennCounts</span>(
    <span class="kw">cbind</span>(<span class="kw">colnames</span>(reads) <span class="op">%in%</span><span class="st"> </span>auto,
    <span class="kw">colnames</span>(reads) <span class="op">%in%</span><span class="st"> </span>man)
)
<span class="kw">vennDiagram</span>(
    venn.diag,
    <span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;Automatic&quot;</span>, <span class="st">&quot;Manual&quot;</span>),
    <span class="dt">circle.col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>)
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:cell-filt-comp-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/cell-filt-comp-reads-1.png" alt="Comparison of the default, automatic and manual cell filters" width="90%" />
<p class="caption">
Figure 3.21: Comparison of the default, automatic and manual cell filters
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotQC</span>(reads, <span class="dt">type =</span> <span class="st">&quot;highest-expression&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:top50-gene-expr-reads"></span>
<img src="08-exprs-qc-reads_files/figure-html/top50-gene-expr-reads-1.png" alt="Number of total counts consumed by the top 50 expressed genes" width="90%" />
<p class="caption">
Figure 3.22: Number of total counts consumed by the top 50 expressed genes
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_genes &lt;-<span class="st"> </span><span class="kw">apply</span>(
    <span class="kw">counts</span>(reads[, <span class="kw">colData</span>(reads)<span class="op">$</span>use]), 
    <span class="dv">1</span>, 
    <span class="cf">function</span>(x) <span class="kw">length</span>(x[x <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>]) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span>
)
<span class="kw">rowData</span>(reads)<span class="op">$</span>use &lt;-<span class="st"> </span>filter_genes</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(filter_genes)</code></pre></div>
<pre><code>## filter_genes
## FALSE  TRUE 
##  2664 16062</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(reads[<span class="kw">rowData</span>(reads)<span class="op">$</span>use, <span class="kw">colData</span>(reads)<span class="op">$</span>use])</code></pre></div>
<pre><code>## [1] 16062   606</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assay</span>(reads, <span class="st">&quot;logcounts_raw&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">counts</span>(reads) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">reducedDim</span>(reads) &lt;-<span class="st"> </span><span class="ot">NULL</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(reads, <span class="dt">file =</span> <span class="st">&quot;tung/reads.rds&quot;</span>)</code></pre></div>
<p>By comparing Figure <a href="construction-of-expression-matrix.html#fig:cell-filt-comp">3.14</a> and Figure <a href="construction-of-expression-matrix.html#fig:cell-filt-comp-reads">3.21</a>, it is clear that the reads based filtering removed more cells than the UMI based analysis. If you go back and compare the results you should be able to conclude that the ERCC and MT filters are more strict for the reads-based analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] limma_3.34.1               scater_1.6.0              
##  [3] ggplot2_2.2.1              SingleCellExperiment_1.0.0
##  [5] SummarizedExperiment_1.8.0 DelayedArray_0.4.1        
##  [7] matrixStats_0.52.2         Biobase_2.38.0            
##  [9] GenomicRanges_1.30.0       GenomeInfoDb_1.14.0       
## [11] IRanges_2.12.0             S4Vectors_0.16.0          
## [13] BiocGenerics_0.24.0        knitr_1.17                
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0        minqa_1.2.4            
##   [3] colorspace_1.3-2        mvoutlier_2.0.8        
##   [5] rjson_0.2.15            modeltools_0.2-21      
##   [7] class_7.3-14            mclust_5.3             
##   [9] rprojroot_1.2           XVector_0.18.0         
##  [11] pls_2.6-0               cvTools_0.3.2          
##  [13] MatrixModels_0.4-1      flexmix_2.3-14         
##  [15] bit64_0.9-7             AnnotationDbi_1.40.0   
##  [17] mvtnorm_1.0-6           sROC_0.1-2             
##  [19] splines_3.4.2           tximport_1.6.0         
##  [21] robustbase_0.92-8       nloptr_1.0.4           
##  [23] robCompositions_2.0.6   pbkrtest_0.4-7         
##  [25] kernlab_0.9-25          cluster_2.0.6          
##  [27] shinydashboard_0.6.1    shiny_1.0.5            
##  [29] rrcov_1.4-3             compiler_3.4.2         
##  [31] backports_1.1.1         assertthat_0.2.0       
##  [33] Matrix_1.2-7.1          lazyeval_0.2.1         
##  [35] htmltools_0.3.6         quantreg_5.34          
##  [37] prettyunits_1.0.2       tools_3.4.2            
##  [39] bindrcpp_0.2            gtable_0.2.0           
##  [41] glue_1.2.0              GenomeInfoDbData_0.99.1
##  [43] reshape2_1.4.2          dplyr_0.7.4            
##  [45] Rcpp_0.12.13            trimcluster_0.1-2      
##  [47] sgeostat_1.0-27         nlme_3.1-129           
##  [49] fpc_2.1-10              lmtest_0.9-35          
##  [51] laeken_0.4.6            stringr_1.2.0          
##  [53] lme4_1.1-14             mime_0.5               
##  [55] XML_3.98-1.9            edgeR_3.20.1           
##  [57] DEoptimR_1.0-8          zoo_1.8-0              
##  [59] zlibbioc_1.24.0         MASS_7.3-45            
##  [61] scales_0.5.0            VIM_4.7.0              
##  [63] rhdf5_2.22.0            SparseM_1.77           
##  [65] RColorBrewer_1.1-2      yaml_2.1.14            
##  [67] memoise_1.1.0           gridExtra_2.3          
##  [69] biomaRt_2.34.0          reshape_0.8.7          
##  [71] stringi_1.1.5           RSQLite_2.0            
##  [73] highr_0.6               pcaPP_1.9-72           
##  [75] e1071_1.6-8             boot_1.3-18            
##  [77] prabclus_2.2-6          rlang_0.1.4            
##  [79] pkgconfig_2.0.1         bitops_1.0-6           
##  [81] evaluate_0.10.1         lattice_0.20-34        
##  [83] bindr_0.1               labeling_0.3           
##  [85] cowplot_0.9.0           bit_1.1-12             
##  [87] GGally_1.3.2            plyr_1.8.4             
##  [89] magrittr_1.5            bookdown_0.5           
##  [91] R6_2.2.2                DBI_0.7                
##  [93] mgcv_1.8-22             RCurl_1.95-4.8         
##  [95] sp_1.2-5                nnet_7.3-12            
##  [97] tibble_1.3.4            car_2.1-5              
##  [99] rmarkdown_1.7           viridis_0.4.0          
## [101] progress_1.1.2          locfit_1.5-9.1         
## [103] grid_3.4.2              data.table_1.10.4-3    
## [105] blob_1.1.0              diptest_0.75-7         
## [107] vcd_1.4-3               digest_0.6.12          
## [109] xtable_1.8-2            httpuv_1.3.5           
## [111] munsell_0.4.3           beeswarm_0.2.3         
## [113] viridisLite_0.2.0       vipor_0.4.5</code></pre>

</div>
<div id="data-visualization" class="section level2">
<h2><span class="header-section-number">3.10</span> Data visualization</h2>
<div id="introduction-2" class="section level3">
<h3><span class="header-section-number">3.10.1</span> Introduction</h3>
<p>In this chapter we will continue to work with the filtered <code>Tung</code> dataset produced in the previous chapter. We will explore different ways of visualizing the data to allow you to asses what happened to the expression matrix after the quality control step. <code>scater</code> package provides several very useful functions to simplify visualisation.</p>
<p>One important aspect of single-cell RNA-seq is to control for batch effects. Batch effects are technical artefacts that are added to the samples during handling. For example, if two sets of samples were prepared in different labs or even on different days in the same lab, then we may observe greater similarities between the samples that were handled together. In the worst case scenario, batch effects may be <a href="http://f1000research.com/articles/4-121/v1">mistaken</a> for true biological variation. The <code>Tung</code> data allows us to explore these issues in a controlled manner since some of the salient aspects of how the samples were handled have been recorded. Ideally, we expect to see batches from the same individual grouping together and distinct groups corresponding to each individual.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scater)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
umi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/umi.rds&quot;</span>)
umi.qc &lt;-<span class="st"> </span>umi[<span class="kw">rowData</span>(umi)<span class="op">$</span>use, <span class="kw">colData</span>(umi)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(umi.qc)<span class="op">$</span>is_feature_control</code></pre></div>
</div>
<div id="visual-pca" class="section level3">
<h3><span class="header-section-number">3.10.2</span> PCA plot</h3>
<p>The easiest way to overview the data is by transforming it using the principal component analysis and then visualize the first two principal components.</p>
<p><a href="https://en.wikipedia.org/wiki/Principal_component_analysis">Principal component analysis (PCA)</a> is a statistical procedure that uses a transformation to convert a set of observations into a set of values of linearly uncorrelated variables called principal components (PCs). The number of principal components is less than or equal to the number of original variables.</p>
<p>Mathematically, the PCs correspond to the eigenvectors of the covariance matrix. The eigenvectors are sorted by eigenvalue so that the first principal component accounts for as much of the variability in the data as possible, and each succeeding component in turn has the highest variance possible under the constraint that it is orthogonal to the preceding components (the figure below is taken from <a href="http://www.nlpca.org/pca_principal_component_analysis.html">here</a>).</p>
<div class="figure" style="text-align: center"><span id="fig:clust-pca"></span>
<img src="figures/pca.png" alt="Schematic representation of PCA dimensionality reduction" width="100%" />
<p class="caption">
Figure 3.23: Schematic representation of PCA dimensionality reduction
</p>
</div>
<div id="before-qc" class="section level4">
<h4><span class="header-section-number">3.10.2.1</span> Before QC</h4>
<p>Without log-transformation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-before-qc1"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-pca-before-qc1-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.24: PCA plot of the tung data
</p>
</div>
<p>With log-transformation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-before-qc2"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-pca-before-qc2-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.25: PCA plot of the tung data
</p>
</div>
<p>Clearly log-transformation is benefitial for our data - it reduces the variance on the first principal component and already separates some biological effects. Moreover, it makes the distribution of the expression values more normal. In the following analysis and chapters we will be using log-transformed raw counts by default.</p>
<p><strong>However, note that just a log-transformation is not enough to account for different technical factors between the cells (e.g. sequencing depth). Therefore, please do not use <code>logcounts_raw</code> for your downstream analysis, instead as a minimum suitable data use the <code>logcounts</code> slot of the <code>SingleCellExperiment</code> object, which not just log-transformed, but also normalised by library size (e.g. CPM normalisation). In the course we use <code>logcounts_raw</code> only for demonstration purposes!</strong></p>
</div>
<div id="after-qc" class="section level4">
<h4><span class="header-section-number">3.10.2.2</span> After QC</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-after-qc"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-pca-after-qc-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.26: PCA plot of the tung data
</p>
</div>
<p>Comparing Figure <a href="construction-of-expression-matrix.html#fig:expr-overview-pca-before-qc2">3.25</a> and Figure <a href="construction-of-expression-matrix.html#fig:expr-overview-pca-after-qc">3.26</a>, it is clear that after quality control the NA19098.r2 cells no longer form a group of outliers.</p>
<p>By default only the top 500 most variable genes are used by scater to calculate the PCA. This can be adjusted by changing the <code>ntop</code> argument.</p>
<p><strong>Exercise 1</strong> How do the PCA plots change if when all 14,214 genes are used? Or when only top 50 genes are used? Why does the fraction of variance accounted for by the first PC change so dramatically?</p>
<p><strong>Hint</strong> Use <code>ntop</code> argument of the <code>plotPCA</code> function.</p>
<p><strong>Our answer</strong></p>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-after-qc-exercise1-1"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-pca-after-qc-exercise1-1-1.png" alt="PCA plot of the tung data (14214 genes)" width="90%" />
<p class="caption">
Figure 3.27: PCA plot of the tung data (14214 genes)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-after-qc-exercise1-2"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-pca-after-qc-exercise1-2-1.png" alt="PCA plot of the tung data (50 genes)" width="90%" />
<p class="caption">
Figure 3.28: PCA plot of the tung data (50 genes)
</p>
</div>
<p>If your answers are different please compare your code with <a href="https://github.com/hemberg-lab/scRNA.seq.course/blob/master/07-exprs-overview.Rmd">ours</a> (you need to search for this exercise in the opened file).</p>
</div>
</div>
<div id="visual-tsne" class="section level3">
<h3><span class="header-section-number">3.10.3</span> tSNE map</h3>
<p>An alternative to PCA for visualizing scRNASeq data is a tSNE plot. <a href="https://lvdmaaten.github.io/tsne/">tSNE</a> (t-Distributed Stochastic Neighbor Embedding) combines dimensionality reduction (e.g. PCA) with random walks on the nearest-neighbour network to map high dimensional data (i.e. our 14,214 dimensional expression matrix) to a 2-dimensional space while preserving local distances between cells. In contrast with PCA, tSNE is a stochastic algorithm which means running the method multiple times on the same dataset will result in different plots. Due to the non-linear and stochastic nature of the algorithm, tSNE is more difficult to intuitively interpret tSNE. To ensure reproducibility, we fix the “seed” of the random-number generator in the code below so that we always get the same plot.</p>
<div id="before-qc-1" class="section level4">
<h4><span class="header-section-number">3.10.3.1</span> Before QC</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotTSNE</span>(
    umi[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">perplexity =</span> <span class="dv">130</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
    <span class="dt">rand_seed =</span> <span class="dv">123456</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-before-qc"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-tsne-before-qc-1.png" alt="tSNE map of the tung data" width="90%" />
<p class="caption">
Figure 3.29: tSNE map of the tung data
</p>
</div>
</div>
<div id="after-qc-1" class="section level4">
<h4><span class="header-section-number">3.10.3.2</span> After QC</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotTSNE</span>(
    umi.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">perplexity =</span> <span class="dv">130</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
    <span class="dt">rand_seed =</span> <span class="dv">123456</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-tsne-after-qc-1.png" alt="tSNE map of the tung data" width="90%" />
<p class="caption">
Figure 3.30: tSNE map of the tung data
</p>
</div>
<p>Interpreting PCA and tSNE plots is often challenging and due to their stochastic and non-linear nature, they are less intuitive. However, in this case it is clear that they provide a similar picture of the data. Comparing Figure <a href="construction-of-expression-matrix.html#fig:expr-overview-tsne-before-qc">3.29</a> and <a href="construction-of-expression-matrix.html#fig:expr-overview-tsne-after-qc">3.30</a>, it is again clear that the samples from NA19098.r2 are no longer outliers after the QC filtering.</p>
<p>Furthermore tSNE requires you to provide a value of <code>perplexity</code> which reflects the number of neighbours used to build the nearest-neighbour network; a high value creates a dense network which clumps cells together while a low value makes the network more sparse allowing groups of cells to separate from each other. <code>scater</code> uses a default perplexity of the total number of cells divided by five (rounded down).</p>
<p>You can read more about the pitfalls of using tSNE <a href="http://distill.pub/2016/misread-tsne/">here</a>.</p>
<p><strong>Exercise 2</strong> How do the tSNE plots change when a perplexity of 10 or 200 is used? How does the choice of perplexity affect the interpretation of the results?</p>
<p><strong>Our answer</strong></p>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc-exercise2-1"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-tsne-after-qc-exercise2-1-1.png" alt="tSNE map of the tung data (perplexity = 10)" width="90%" />
<p class="caption">
Figure 3.31: tSNE map of the tung data (perplexity = 10)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc-exercise2-2"></span>
<img src="09-exprs-overview_files/figure-html/expr-overview-tsne-after-qc-exercise2-2-1.png" alt="tSNE map of the tung data (perplexity = 200)" width="90%" />
<p class="caption">
Figure 3.32: tSNE map of the tung data (perplexity = 200)
</p>
</div>
</div>
</div>
<div id="big-exercise-1" class="section level3">
<h3><span class="header-section-number">3.10.4</span> Big Exercise</h3>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>tung/reads.rds</code> file to load the reads SCE object. Once you have finished please compare your results to ours (next chapter).</p>
</div>
<div id="sessioninfo-1" class="section level3">
<h3><span class="header-section-number">3.10.5</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] scater_1.6.0               ggplot2_2.2.1             
##  [3] SingleCellExperiment_1.0.0 SummarizedExperiment_1.8.0
##  [5] DelayedArray_0.4.1         matrixStats_0.52.2        
##  [7] Biobase_2.38.0             GenomicRanges_1.30.0      
##  [9] GenomeInfoDb_1.14.0        IRanges_2.12.0            
## [11] S4Vectors_0.16.0           BiocGenerics_0.24.0       
## [13] knitr_1.17                
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.4.0           edgeR_3.20.1           
##  [3] bit64_0.9-7             viridisLite_0.2.0      
##  [5] shiny_1.0.5             assertthat_0.2.0       
##  [7] highr_0.6               blob_1.1.0             
##  [9] GenomeInfoDbData_0.99.1 vipor_0.4.5            
## [11] yaml_2.1.14             progress_1.1.2         
## [13] RSQLite_2.0             backports_1.1.1        
## [15] lattice_0.20-34         glue_1.2.0             
## [17] limma_3.34.1            digest_0.6.12          
## [19] XVector_0.18.0          colorspace_1.3-2       
## [21] cowplot_0.9.0           htmltools_0.3.6        
## [23] httpuv_1.3.5            Matrix_1.2-7.1         
## [25] plyr_1.8.4              XML_3.98-1.9           
## [27] pkgconfig_2.0.1         biomaRt_2.34.0         
## [29] bookdown_0.5            zlibbioc_1.24.0        
## [31] xtable_1.8-2            scales_0.5.0           
## [33] Rtsne_0.13              tibble_1.3.4           
## [35] lazyeval_0.2.1          magrittr_1.5           
## [37] mime_0.5                memoise_1.1.0          
## [39] evaluate_0.10.1         beeswarm_0.2.3         
## [41] shinydashboard_0.6.1    tools_3.4.2            
## [43] data.table_1.10.4-3     prettyunits_1.0.2      
## [45] stringr_1.2.0           munsell_0.4.3          
## [47] locfit_1.5-9.1          AnnotationDbi_1.40.0   
## [49] bindrcpp_0.2            compiler_3.4.2         
## [51] rlang_0.1.4             rhdf5_2.22.0           
## [53] grid_3.4.2              RCurl_1.95-4.8         
## [55] tximport_1.6.0          rjson_0.2.15           
## [57] labeling_0.3            bitops_1.0-6           
## [59] rmarkdown_1.7           gtable_0.2.0           
## [61] DBI_0.7                 reshape2_1.4.2         
## [63] R6_2.2.2                gridExtra_2.3          
## [65] dplyr_0.7.4             bit_1.1-12             
## [67] bindr_0.1               rprojroot_1.2          
## [69] stringi_1.1.5           ggbeeswarm_0.6.0       
## [71] Rcpp_0.12.13</code></pre>

</div>
</div>
<div id="data-visualization-reads" class="section level2">
<h2><span class="header-section-number">3.11</span> Data visualization (Reads)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
reads &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/reads.rds&quot;</span>)
reads.qc &lt;-<span class="st"> </span>reads[<span class="kw">rowData</span>(reads)<span class="op">$</span>use, <span class="kw">colData</span>(reads)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(reads.qc)<span class="op">$</span>is_feature_control</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    reads[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-before-qc-reads1"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads1-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.33: PCA plot of the tung data
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    reads[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-before-qc-reads2"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads2-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.34: PCA plot of the tung data
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    reads.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-pca-after-qc-reads"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-pca-after-qc-reads-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.35: PCA plot of the tung data
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotTSNE</span>(
    reads[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">perplexity =</span> <span class="dv">130</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
    <span class="dt">rand_seed =</span> <span class="dv">123456</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-before-qc-reads"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-tsne-before-qc-reads-1.png" alt="tSNE map of the tung data" width="90%" />
<p class="caption">
Figure 3.36: tSNE map of the tung data
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotTSNE</span>(
    reads.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">perplexity =</span> <span class="dv">130</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
    <span class="dt">rand_seed =</span> <span class="dv">123456</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc-reads"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-reads-1.png" alt="tSNE map of the tung data" width="90%" />
<p class="caption">
Figure 3.37: tSNE map of the tung data
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc-exercise2-1"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-1-1.png" alt="tSNE map of the tung data (perplexity = 10)" width="90%" />
<p class="caption">
Figure 3.31: tSNE map of the tung data (perplexity = 10)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:expr-overview-tsne-after-qc-exercise2-2"></span>
<img src="10-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-2-1.png" alt="tSNE map of the tung data (perplexity = 200)" width="90%" />
<p class="caption">
Figure 3.32: tSNE map of the tung data (perplexity = 200)
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] knitr_1.17                 scater_1.6.0              
##  [3] SingleCellExperiment_1.0.0 SummarizedExperiment_1.8.0
##  [5] DelayedArray_0.4.1         matrixStats_0.52.2        
##  [7] GenomicRanges_1.30.0       GenomeInfoDb_1.14.0       
##  [9] IRanges_2.12.0             S4Vectors_0.16.0          
## [11] ggplot2_2.2.1              Biobase_2.38.0            
## [13] BiocGenerics_0.24.0       
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.4.0           edgeR_3.20.1           
##  [3] bit64_0.9-7             viridisLite_0.2.0      
##  [5] shiny_1.0.5             assertthat_0.2.0       
##  [7] highr_0.6               blob_1.1.0             
##  [9] GenomeInfoDbData_0.99.1 vipor_0.4.5            
## [11] yaml_2.1.14             progress_1.1.2         
## [13] RSQLite_2.0             backports_1.1.1        
## [15] lattice_0.20-34         glue_1.2.0             
## [17] limma_3.34.1            digest_0.6.12          
## [19] XVector_0.18.0          colorspace_1.3-2       
## [21] cowplot_0.9.0           htmltools_0.3.6        
## [23] httpuv_1.3.5            Matrix_1.2-7.1         
## [25] plyr_1.8.4              XML_3.98-1.9           
## [27] pkgconfig_2.0.1         biomaRt_2.34.0         
## [29] bookdown_0.5            zlibbioc_1.24.0        
## [31] xtable_1.8-2            scales_0.5.0           
## [33] Rtsne_0.13              tibble_1.3.4           
## [35] lazyeval_0.2.1          magrittr_1.5           
## [37] mime_0.5                memoise_1.1.0          
## [39] evaluate_0.10.1         beeswarm_0.2.3         
## [41] shinydashboard_0.6.1    tools_3.4.2            
## [43] data.table_1.10.4-3     prettyunits_1.0.2      
## [45] stringr_1.2.0           munsell_0.4.3          
## [47] locfit_1.5-9.1          AnnotationDbi_1.40.0   
## [49] bindrcpp_0.2            compiler_3.4.2         
## [51] rlang_0.1.4             rhdf5_2.22.0           
## [53] grid_3.4.2              RCurl_1.95-4.8         
## [55] tximport_1.6.0          rjson_0.2.15           
## [57] labeling_0.3            bitops_1.0-6           
## [59] rmarkdown_1.7           gtable_0.2.0           
## [61] DBI_0.7                 reshape2_1.4.2         
## [63] R6_2.2.2                gridExtra_2.3          
## [65] dplyr_0.7.4             bit_1.1-12             
## [67] bindr_0.1               rprojroot_1.2          
## [69] stringi_1.1.5           ggbeeswarm_0.6.0       
## [71] Rcpp_0.12.13</code></pre>

</div>
<div id="identifying-confounding-factors" class="section level2">
<h2><span class="header-section-number">3.12</span> Identifying confounding factors</h2>
<div id="introduction-3" class="section level3">
<h3><span class="header-section-number">3.12.1</span> Introduction</h3>
<p>There is a large number of potential confounders, artifacts and biases in sc-RNA-seq data. One of the main challenges in analyzing scRNA-seq data stems from the fact that it is difficult to carry out a true technical replicate (why?) to distinguish biological and technical variability. In the previous chapters we considered batch effects and in this chapter we will continue to explore how experimental artifacts can be identified and removed. We will continue using the <code>scater</code> package since it provides a set of methods specifically for quality control of experimental and explanatory variables. Moreover, we will continue to work with the Blischak data that was used in the previous chapter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
umi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/umi.rds&quot;</span>)
umi.qc &lt;-<span class="st"> </span>umi[<span class="kw">rowData</span>(umi)<span class="op">$</span>use, <span class="kw">colData</span>(umi)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(umi.qc)<span class="op">$</span>is_feature_control</code></pre></div>
<p>The <code>umi.qc</code> dataset contains filtered cells and genes. Our next step is to explore technical drivers of variability in the data to inform data normalisation before downstream analysis.</p>
</div>
<div id="correlations-with-pcs" class="section level3">
<h3><span class="header-section-number">3.12.2</span> Correlations with PCs</h3>
<p>Let’s first look again at the PCA plot of the QCed dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confound-pca"></span>
<img src="11-confounders_files/figure-html/confound-pca-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.38: PCA plot of the tung data
</p>
</div>
<p><code>scater</code> allows one to identify principal components that correlate with experimental and QC variables of interest (it ranks principle components by <span class="math inline">\(R^2\)</span> from a linear model regressing PC value against the variable of interest).</p>
<p>Let’s test whether some of the variables correlate with any of the PCs.</p>
<div id="detected-genes-1" class="section level4">
<h4><span class="header-section-number">3.12.2.1</span> Detected genes</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotQC</span>(
    umi.qc[endog_genes, ],
    <span class="dt">type =</span> <span class="st">&quot;find-pcs&quot;</span>,
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">variable =</span> <span class="st">&quot;total_features&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confound-find-pcs-total-features"></span>
<img src="11-confounders_files/figure-html/confound-find-pcs-total-features-1.png" alt="PC correlation with the number of detected genes" width="90%" />
<p class="caption">
Figure 3.39: PC correlation with the number of detected genes
</p>
</div>
<p>Indeed, we can see that <code>PC1</code> can be almost completely explained by the number of detected genes. In fact, it was also visible on the PCA plot above. This is a well-known issue in scRNA-seq and was described <a href="http://biorxiv.org/content/early/2015/12/27/025528">here</a>.</p>
</div>
</div>
<div id="explanatory-variables" class="section level3">
<h3><span class="header-section-number">3.12.3</span> Explanatory variables</h3>
<p><code>scater</code> can also compute the marginal <span class="math inline">\(R^2\)</span> for each variable when fitting a linear model regressing expression values for each gene against just that variable, and display a density plot of the gene-wise marginal <span class="math inline">\(R^2\)</span> values for the variables.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotQC</span>(
    umi.qc[endog_genes, ],
    <span class="dt">type =</span> <span class="st">&quot;expl&quot;</span>,
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">variables =</span> <span class="kw">c</span>(
        <span class="st">&quot;total_features&quot;</span>,
        <span class="st">&quot;total_counts&quot;</span>,
        <span class="st">&quot;batch&quot;</span>,
        <span class="st">&quot;individual&quot;</span>,
        <span class="st">&quot;pct_counts_ERCC&quot;</span>,
        <span class="st">&quot;pct_counts_MT&quot;</span>
    )
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confound-find-expl-vars"></span>
<img src="11-confounders_files/figure-html/confound-find-expl-vars-1.png" alt="Explanatory variables" width="90%" />
<p class="caption">
Figure 3.40: Explanatory variables
</p>
</div>
<p>This analysis indicates that the number of detected genes (again) and also the sequencing depth (number of counts) have substantial explanatory power for many genes, so these variables are good candidates for conditioning out in a normalisation step, or including in downstream statistical models. Expression of ERCCs also appears to be an important explanatory variable and one notable feature of the above plot is that batch explains more than individual. What does that tell us about the technical and biological variability of the data?</p>
</div>
<div id="other-confounders" class="section level3">
<h3><span class="header-section-number">3.12.4</span> Other confounders</h3>
<p>In addition to correcting for batch, there are other factors that one may want to compensate for. As with batch correction, these adjustments require extrinsic information. One popular method is <a href="https://github.com/PMBio/scLVM">scLVM</a> which allows you to identify and subtract the effect from processes such as cell-cycle or apoptosis.</p>
<p>In addition, protocols may differ in terms of their coverage of each transcript, their bias based on the average content of <strong>A/T</strong> nucleotides, or their ability to capture short transcripts. Ideally, we would like to compensate for all of these differences and biases.</p>
</div>
<div id="exercise" class="section level3">
<h3><span class="header-section-number">3.12.5</span> Exercise</h3>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>tung/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter).</p>
</div>
<div id="sessioninfo-2" class="section level3">
<h3><span class="header-section-number">3.12.6</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] scater_1.6.0               SingleCellExperiment_1.0.0
##  [3] SummarizedExperiment_1.8.0 DelayedArray_0.4.1        
##  [5] matrixStats_0.52.2         GenomicRanges_1.30.0      
##  [7] GenomeInfoDb_1.14.0        IRanges_2.12.0            
##  [9] S4Vectors_0.16.0           ggplot2_2.2.1             
## [11] Biobase_2.38.0             BiocGenerics_0.24.0       
## [13] knitr_1.17                
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.4.0           edgeR_3.20.1           
##  [3] bit64_0.9-7             viridisLite_0.2.0      
##  [5] shiny_1.0.5             assertthat_0.2.0       
##  [7] highr_0.6               blob_1.1.0             
##  [9] GenomeInfoDbData_0.99.1 vipor_0.4.5            
## [11] yaml_2.1.14             progress_1.1.2         
## [13] RSQLite_2.0             backports_1.1.1        
## [15] lattice_0.20-34         glue_1.2.0             
## [17] limma_3.34.1            digest_0.6.12          
## [19] XVector_0.18.0          colorspace_1.3-2       
## [21] cowplot_0.9.0           htmltools_0.3.6        
## [23] httpuv_1.3.5            Matrix_1.2-7.1         
## [25] plyr_1.8.4              XML_3.98-1.9           
## [27] pkgconfig_2.0.1         biomaRt_2.34.0         
## [29] bookdown_0.5            zlibbioc_1.24.0        
## [31] xtable_1.8-2            scales_0.5.0           
## [33] tibble_1.3.4            lazyeval_0.2.1         
## [35] magrittr_1.5            mime_0.5               
## [37] memoise_1.1.0           evaluate_0.10.1        
## [39] beeswarm_0.2.3          shinydashboard_0.6.1   
## [41] tools_3.4.2             data.table_1.10.4-3    
## [43] prettyunits_1.0.2       stringr_1.2.0          
## [45] munsell_0.4.3           locfit_1.5-9.1         
## [47] AnnotationDbi_1.40.0    bindrcpp_0.2           
## [49] compiler_3.4.2          rlang_0.1.4            
## [51] rhdf5_2.22.0            grid_3.4.2             
## [53] RCurl_1.95-4.8          tximport_1.6.0         
## [55] rjson_0.2.15            labeling_0.3           
## [57] bitops_1.0-6            rmarkdown_1.7          
## [59] gtable_0.2.0            DBI_0.7                
## [61] reshape2_1.4.2          R6_2.2.2               
## [63] gridExtra_2.3           dplyr_0.7.4            
## [65] bit_1.1-12              bindr_0.1              
## [67] rprojroot_1.2           stringi_1.1.5          
## [69] ggbeeswarm_0.6.0        Rcpp_0.12.13</code></pre>

</div>
</div>
<div id="identifying-confounding-factors-reads" class="section level2">
<h2><span class="header-section-number">3.13</span> Identifying confounding factors (Reads)</h2>
<div class="figure" style="text-align: center"><span id="fig:confound-pca-reads"></span>
<img src="12-confounders-reads_files/figure-html/confound-pca-reads-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.41: PCA plot of the tung data
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-find-pcs-total-features-reads"></span>
<img src="12-confounders-reads_files/figure-html/confound-find-pcs-total-features-reads-1.png" alt="PC correlation with the number of detected genes" width="90%" />
<p class="caption">
Figure 3.42: PC correlation with the number of detected genes
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-find-expl-vars-reads"></span>
<img src="12-confounders-reads_files/figure-html/confound-find-expl-vars-reads-1.png" alt="Explanatory variables" width="90%" />
<p class="caption">
Figure 3.43: Explanatory variables
</p>
</div>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] knitr_1.17                 scater_1.6.0              
##  [3] SingleCellExperiment_1.0.0 SummarizedExperiment_1.8.0
##  [5] DelayedArray_0.4.1         matrixStats_0.52.2        
##  [7] GenomicRanges_1.30.0       GenomeInfoDb_1.14.0       
##  [9] IRanges_2.12.0             S4Vectors_0.16.0          
## [11] ggplot2_2.2.1              Biobase_2.38.0            
## [13] BiocGenerics_0.24.0       
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.4.0           edgeR_3.20.1           
##  [3] bit64_0.9-7             viridisLite_0.2.0      
##  [5] shiny_1.0.5             assertthat_0.2.0       
##  [7] highr_0.6               blob_1.1.0             
##  [9] GenomeInfoDbData_0.99.1 vipor_0.4.5            
## [11] yaml_2.1.14             progress_1.1.2         
## [13] RSQLite_2.0             backports_1.1.1        
## [15] lattice_0.20-34         glue_1.2.0             
## [17] limma_3.34.1            digest_0.6.12          
## [19] XVector_0.18.0          colorspace_1.3-2       
## [21] cowplot_0.9.0           htmltools_0.3.6        
## [23] httpuv_1.3.5            Matrix_1.2-7.1         
## [25] plyr_1.8.4              XML_3.98-1.9           
## [27] pkgconfig_2.0.1         biomaRt_2.34.0         
## [29] bookdown_0.5            zlibbioc_1.24.0        
## [31] xtable_1.8-2            scales_0.5.0           
## [33] tibble_1.3.4            lazyeval_0.2.1         
## [35] magrittr_1.5            mime_0.5               
## [37] memoise_1.1.0           evaluate_0.10.1        
## [39] beeswarm_0.2.3          shinydashboard_0.6.1   
## [41] tools_3.4.2             data.table_1.10.4-3    
## [43] prettyunits_1.0.2       stringr_1.2.0          
## [45] munsell_0.4.3           locfit_1.5-9.1         
## [47] AnnotationDbi_1.40.0    bindrcpp_0.2           
## [49] compiler_3.4.2          rlang_0.1.4            
## [51] rhdf5_2.22.0            grid_3.4.2             
## [53] RCurl_1.95-4.8          tximport_1.6.0         
## [55] rjson_0.2.15            labeling_0.3           
## [57] bitops_1.0-6            rmarkdown_1.7          
## [59] gtable_0.2.0            DBI_0.7                
## [61] reshape2_1.4.2          R6_2.2.2               
## [63] gridExtra_2.3           dplyr_0.7.4            
## [65] bit_1.1-12              bindr_0.1              
## [67] rprojroot_1.2           stringi_1.1.5          
## [69] ggbeeswarm_0.6.0        Rcpp_0.12.13</code></pre>

</div>
<div id="normalization-theory" class="section level2">
<h2><span class="header-section-number">3.14</span> Normalization theory</h2>
<div id="introduction-4" class="section level3">
<h3><span class="header-section-number">3.14.1</span> Introduction</h3>
<p>In the previous chapter we identified important confounding factors and explanatory variables. <code>scater</code> allows one to account for these variables in subsequent statistical models or to condition them out using <code>normaliseExprs()</code>, if so desired. This can be done by providing a design matrix to <code>normaliseExprs()</code>. We are not covering this topic here, but you can try to do it yourself as an exercise.</p>
<p>Instead we will explore how simple size-factor normalisations correcting for library size can remove the effects of some of the confounders and explanatory variables.</p>
</div>
<div id="library-size-1" class="section level3">
<h3><span class="header-section-number">3.14.2</span> Library size</h3>
<p>Library sizes vary because scRNA-seq data is often sequenced on highly multiplexed platforms the total reads which are derived from each cell may differ substantially. Some quantification methods (eg. <a href="http://cole-trapnell-lab.github.io/cufflinks/"><code>Cufflinks</code></a>, <a href="http://deweylab.github.io/RSEM/"><code>RSEM</code></a>) incorporated library size when determining gene expression estimates thus do not require this normalization.</p>
<p>However, if another quantification method was used then library size must be corrected for by multiplying or dividing each column of the expression matrix by a “normalization factor” which is an estimate of the library size relative to the other cells. Many methods to correct for library size have been developped for bulk RNA-seq and can be equally applied to scRNA-seq (eg. <strong>UQ</strong>, <strong>SF</strong>, <strong>CPM</strong>, <strong>RPKM</strong>, <strong>FPKM</strong>, <strong>TPM</strong>).</p>
</div>
<div id="normalisations" class="section level3">
<h3><span class="header-section-number">3.14.3</span> Normalisations</h3>
<div id="cpm" class="section level4">
<h4><span class="header-section-number">3.14.3.1</span> CPM</h4>
<p>The simplest way to normalize this data is to convert it to counts per million (<strong>CPM</strong>) by dividing each column by its total then multiplying by 1,000,000. Note that spike-ins should be excluded from the calculation of total expression in order to correct for total cell RNA content, therefore we will only use endogenous genes. Example of a <strong>CPM</strong> function in <code>R</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">calc_cpm &lt;-
<span class="cf">function</span> (expr_mat, <span class="dt">spikes =</span> <span class="ot">NULL</span>) 
{
    norm_factor &lt;-<span class="st"> </span><span class="kw">colSums</span>(expr_mat[<span class="op">-</span>spikes, ])
    <span class="kw">return</span>(<span class="kw">t</span>(<span class="kw">t</span>(expr_mat)<span class="op">/</span>norm_factor)) <span class="op">*</span><span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>
}</code></pre></div>
<p>One potential drawback of <strong>CPM</strong> is if your sample contains genes that are both very highly expressed and differentially expressed across the cells. In this case, the total molecules in the cell may depend of whether such genes are on/off in the cell and normalizing by total molecules may hide the differential expression of those genes and/or falsely create differential expression for the remaining genes.</p>
<p><strong>Note</strong> <strong>RPKM</strong>, <strong>FPKM</strong> and <strong>TPM</strong> are variants on <strong>CPM</strong> which further adjust counts by the length of the respective gene/transcript.</p>
<p>To deal with this potentiality several other measures were devised.</p>
</div>
<div id="rle-sf" class="section level4">
<h4><span class="header-section-number">3.14.3.2</span> RLE (SF)</h4>
<p>The <strong>size factor (SF)</strong> was proposed and popularized by DESeq <span class="citation">(Anders and Huber <a href="#ref-Anders2010-jr">2010</a>)</span>. First the geometric mean of each gene across all cells is calculated. The size factor for each cell is the median across genes of the ratio of the expression to the gene’s geometric mean. A drawback to this method is that since it uses the geometric mean only genes with non-zero expression values across all cells can be used in its calculation, making it unadvisable for large low-depth scRNASeq experiments. <code>edgeR</code> &amp; <code>scater</code> call this method <strong>RLE</strong> for “relative log expression”. Example of a <strong>SF</strong> function in <code>R</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">calc_sf &lt;-
<span class="cf">function</span> (expr_mat, <span class="dt">spikes =</span> <span class="ot">NULL</span>) 
{
    geomeans &lt;-<span class="st"> </span><span class="kw">exp</span>(<span class="kw">rowMeans</span>(<span class="kw">log</span>(expr_mat[<span class="op">-</span>spikes, ])))
    SF &lt;-<span class="st"> </span><span class="cf">function</span>(cnts) {
        <span class="kw">median</span>((cnts<span class="op">/</span>geomeans)[(<span class="kw">is.finite</span>(geomeans) <span class="op">&amp;</span><span class="st"> </span>geomeans <span class="op">&gt;</span><span class="st"> </span>
<span class="st">            </span><span class="dv">0</span>)])
    }
    norm_factor &lt;-<span class="st"> </span><span class="kw">apply</span>(expr_mat[<span class="op">-</span>spikes, ], <span class="dv">2</span>, SF)
    <span class="kw">return</span>(<span class="kw">t</span>(<span class="kw">t</span>(expr_mat)<span class="op">/</span>norm_factor))
}</code></pre></div>
</div>
<div id="uq" class="section level4">
<h4><span class="header-section-number">3.14.3.3</span> UQ</h4>
<p>The <strong>upperquartile (UQ)</strong> was proposed by <span class="citation">(Bullard et al. <a href="#ref-Bullard2010-eb">2010</a>)</span>. Here each column is divided by the 75% quantile of the counts for each library. Often the calculated quantile is scaled by the median across cells to keep the absolute level of expression relatively consistent. A drawback to this method is that for low-depth scRNASeq experiments the large number of undetected genes may result in the 75% quantile being zero (or close to it). This limitation can be overcome by generalizing the idea and using a higher quantile (eg. the 99% quantile is the default in scater) or by excluding zeros prior to calculating the 75% quantile. Example of a <strong>UQ</strong> function in <code>R</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">calc_uq &lt;-
<span class="cf">function</span> (expr_mat, <span class="dt">spikes =</span> <span class="ot">NULL</span>) 
{
    UQ &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
        <span class="kw">quantile</span>(x[x <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>], <span class="fl">0.75</span>)
    }
    uq &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">apply</span>(expr_mat[<span class="op">-</span>spikes, ], <span class="dv">2</span>, UQ))
    norm_factor &lt;-<span class="st"> </span>uq<span class="op">/</span><span class="kw">median</span>(uq)
    <span class="kw">return</span>(<span class="kw">t</span>(<span class="kw">t</span>(expr_mat)<span class="op">/</span>norm_factor))
}</code></pre></div>
</div>
<div id="tmm" class="section level4">
<h4><span class="header-section-number">3.14.3.4</span> TMM</h4>
<p>Another method is called <strong>TMM</strong> is the weighted trimmed mean of M-values (to the reference) proposed by <span class="citation">(Robinson and Oshlack <a href="#ref-Robinson2010-hz">2010</a>)</span>. The M-values in question are the gene-wise log2 fold changes between individual cells. One cell is used as the reference then the M-values for each other cell is calculated compared to this reference. These values are then trimmed by removing the top and bottom ~30%, and the average of the remaining values is calculated by weighting them to account for the effect of the log scale on variance. Each non-reference cell is multiplied by the calculated factor. Two potential issues with this method are insufficient non-zero genes left after trimming, and the assumption that most genes are not differentially expressed.</p>
</div>
<div id="scran" class="section level4">
<h4><span class="header-section-number">3.14.3.5</span> scran</h4>
<p><code>scran</code> package implements a variant on <strong>CPM</strong> specialized for single-cell data <span class="citation">(L. Lun, Bach, and Marioni <a href="#ref-L_Lun2016-pq">2016</a>)</span>. Briefly this method deals with the problem of vary large numbers of zero values per cell by pooling cells together calculating a normalization factor (similar to <strong>CPM</strong>) for the sum of each pool. Since each cell is found in many different pools, cell-specific factors can be deconvoluted from the collection of pool-specific factors using linear algebra.</p>
</div>
<div id="downsampling" class="section level4">
<h4><span class="header-section-number">3.14.3.6</span> Downsampling</h4>
<p>A final way to correct for library size is to downsample the expression matrix so that each cell has approximately the same total number of molecules. The benefit of this method is that zero values will be introduced by the down sampling thus eliminating any biases due to differing numbers of detected genes. However, the major drawback is that the process is not deterministic so each time the downsampling is run the resulting expression matrix is slightly different. Thus, often analyses must be run on multiple downsamplings to ensure results are robust. Example of a <strong>downsampling</strong> function in <code>R</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Down_Sample_Matrix &lt;-
<span class="cf">function</span> (expr_mat) 
{
    min_lib_size &lt;-<span class="st"> </span><span class="kw">min</span>(<span class="kw">colSums</span>(expr_mat))
    down_sample &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
        prob &lt;-<span class="st"> </span>min_lib_size<span class="op">/</span><span class="kw">sum</span>(x)
        <span class="kw">return</span>(<span class="kw">unlist</span>(<span class="kw">lapply</span>(x, <span class="cf">function</span>(y) {
            <span class="kw">rbinom</span>(<span class="dv">1</span>, y, prob)
        })))
    }
    down_sampled_mat &lt;-<span class="st"> </span><span class="kw">apply</span>(expr_mat, <span class="dv">2</span>, down_sample)
    <span class="kw">return</span>(down_sampled_mat)
}</code></pre></div>
</div>
</div>
<div id="effectiveness" class="section level3">
<h3><span class="header-section-number">3.14.4</span> Effectiveness</h3>
<p>to compare the efficiency of different normalization methods we will use visual inspection of <code>PCA</code> plots and calculation of cell-wise <em>relative log expression</em> via <code>scater</code>’s <code>plotRLE()</code> function. Namely, cells with many (few) reads have higher (lower) than median expression for most genes resulting in a positive (negative) <em>RLE</em> across the cell, whereas normalized cells have an <em>RLE</em> close to zero. Example of a <em>RLE</em> function in <code>R</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">calc_cell_RLE &lt;-
<span class="cf">function</span> (expr_mat, <span class="dt">spikes =</span> <span class="ot">NULL</span>) 
{
    RLE_gene &lt;-<span class="st"> </span><span class="cf">function</span>(x) {
        <span class="cf">if</span> (<span class="kw">median</span>(<span class="kw">unlist</span>(x)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
            <span class="kw">log</span>((x <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)<span class="op">/</span>(<span class="kw">median</span>(<span class="kw">unlist</span>(x)) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>))<span class="op">/</span><span class="kw">log</span>(<span class="dv">2</span>)
        }
        <span class="cf">else</span> {
            <span class="kw">rep</span>(<span class="ot">NA</span>, <span class="dt">times =</span> <span class="kw">length</span>(x))
        }
    }
    <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(spikes)) {
        RLE_matrix &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(expr_mat[<span class="op">-</span>spikes, ], <span class="dv">1</span>, RLE_gene))
    }
    <span class="cf">else</span> {
        RLE_matrix &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(expr_mat, <span class="dv">1</span>, RLE_gene))
    }
    cell_RLE &lt;-<span class="st"> </span><span class="kw">apply</span>(RLE_matrix, <span class="dv">2</span>, median, <span class="dt">na.rm =</span> T)
    <span class="kw">return</span>(cell_RLE)
}</code></pre></div>
<p><strong>Note</strong> The <strong>RLE</strong>, <strong>TMM</strong>, and <strong>UQ</strong> size-factor methods were developed for bulk RNA-seq data and, depending on the experimental context, may not be appropriate for single-cell RNA-seq data, as their underlying assumptions may be problematically violated.</p>
<p><strong>Note</strong> <code>scater</code> acts as a wrapper for the <code>calcNormFactors</code> function from <code>edgeR</code> which implements several library size normalization methods making it easy to apply any of these methods to our data.</p>
<p><strong>Note</strong> <code>edgeR</code> makes extra adjustments to some of the normalization methods which may result in somewhat different results than if the original methods are followed exactly, e.g. edgeR’s and scater’s “RLE” method which is based on the “size factor” used by <a href="http://bioconductor.org/packages/DESeq">DESeq</a> may give different results to the <code>estimateSizeFactorsForMatrix</code> method in the <code>DESeq</code>/<code>DESeq2</code> packages. In addition, some versions of <code>edgeR</code> will not calculate the normalization factors correctly unless <code>lib.size</code> is set at 1 for all cells.</p>
<p><strong>Note</strong> For <strong>CPM</strong> normalisation we use <code>scater</code>’s <code>calculateCPM()</code> function. For <strong>RLE</strong>, <strong>UQ</strong> and <strong>TMM</strong> we use <code>scater</code>’s <code>normaliseExprs()</code> function. For <strong>scran</strong> we use <code>scran</code> package to calculate size factors (it also operates on <code>SingleCellExperiment</code> class) and <code>scater</code>’s <code>normalize()</code> to normalise the data. All these normalization functions save the results to the <code>logcounts</code> slot of the <code>SCE</code> object. For <strong>downsampling</strong> we use our own functions shown above.</p>
</div>
</div>
<div id="normalization-practice-umi" class="section level2">
<h2><span class="header-section-number">3.15</span> Normalization practice (UMI)</h2>
<p>We will continue to work with the <code>tung</code> data that was used in the previous chapter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNA.seq.funcs)
<span class="kw">library</span>(scater)
<span class="kw">library</span>(scran)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
<span class="kw">set.seed</span>(<span class="dv">1234567</span>)
umi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/umi.rds&quot;</span>)
umi.qc &lt;-<span class="st"> </span>umi[<span class="kw">rowData</span>(umi)<span class="op">$</span>use, <span class="kw">colData</span>(umi)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(umi.qc)<span class="op">$</span>is_feature_control</code></pre></div>
<div id="raw" class="section level3">
<h3><span class="header-section-number">3.15.1</span> Raw</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-raw"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-raw-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.44: PCA plot of the tung data
</p>
</div>
</div>
<div id="cpm-1" class="section level3">
<h3><span class="header-section-number">3.15.2</span> CPM</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">logcounts</span>(umi.qc) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">calculateCPM</span>(umi.qc, <span class="dt">use.size.factors =</span> <span class="ot">FALSE</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-cpm"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-cpm-1.png" alt="PCA plot of the tung data after CPM normalisation" width="90%" />
<p class="caption">
Figure 3.45: PCA plot of the tung data after CPM normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">CPM =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-cpm"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-cpm-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.46: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="size-factor-rle" class="section level3">
<h3><span class="header-section-number">3.15.3</span> Size-factor (RLE)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi.qc &lt;-<span class="st"> </span><span class="kw">normaliseExprs</span>(
    umi.qc,
    <span class="dt">method =</span> <span class="st">&quot;RLE&quot;</span>, 
    <span class="dt">feature_set =</span> endog_genes,
    <span class="dt">return_log =</span> <span class="ot">TRUE</span>,
    <span class="dt">return_norm_as_exprs =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-rle"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-rle-1.png" alt="PCA plot of the tung data after RLE normalisation" width="90%" />
<p class="caption">
Figure 3.47: PCA plot of the tung data after RLE normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">RLE =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-rle"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-rle-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.48: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="upperquantile" class="section level3">
<h3><span class="header-section-number">3.15.4</span> Upperquantile</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi.qc &lt;-<span class="st"> </span><span class="kw">normaliseExprs</span>(
    umi.qc,
    <span class="dt">method =</span> <span class="st">&quot;upperquartile&quot;</span>, 
    <span class="dt">feature_set =</span> endog_genes,
    <span class="dt">p =</span> <span class="fl">0.99</span>,
    <span class="dt">return_log =</span> <span class="ot">TRUE</span>,
    <span class="dt">return_norm_as_exprs =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-uq"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-uq-1.png" alt="PCA plot of the tung data after UQ normalisation" width="90%" />
<p class="caption">
Figure 3.49: PCA plot of the tung data after UQ normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">UQ =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-uq"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-uq-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.50: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="tmm-1" class="section level3">
<h3><span class="header-section-number">3.15.5</span> TMM</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi.qc &lt;-<span class="st"> </span><span class="kw">normaliseExprs</span>(
    umi.qc,
    <span class="dt">method =</span> <span class="st">&quot;TMM&quot;</span>,
    <span class="dt">feature_set =</span> endog_genes,
    <span class="dt">return_log =</span> <span class="ot">TRUE</span>,
    <span class="dt">return_norm_as_exprs =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-tmm"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-tmm-1.png" alt="PCA plot of the tung data after TMM normalisation" width="90%" />
<p class="caption">
Figure 3.51: PCA plot of the tung data after TMM normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">TMM =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-tmm"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-tmm-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.52: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="scran-1" class="section level3">
<h3><span class="header-section-number">3.15.6</span> scran</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">qclust &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(umi.qc, <span class="dt">min.size =</span> <span class="dv">30</span>)
umi.qc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(umi.qc, <span class="dt">sizes =</span> <span class="dv">15</span>, <span class="dt">clusters =</span> qclust)
umi.qc &lt;-<span class="st"> </span><span class="kw">normalize</span>(umi.qc)</code></pre></div>
<pre><code>## Warning in .local(object, ...): spike-in transcripts in &#39;ERCC&#39; should have
## their own size factors</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-lsf"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-lsf-1.png" alt="PCA plot of the tung data after LSF normalisation" width="90%" />
<p class="caption">
Figure 3.53: PCA plot of the tung data after LSF normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">scran =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-scran"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-scran-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.54: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="downsampling-1" class="section level3">
<h3><span class="header-section-number">3.15.7</span> Downsampling</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">logcounts</span>(umi.qc) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">Down_Sample_Matrix</span>(<span class="kw">counts</span>(umi.qc)) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">plotPCA</span>(
    umi.qc[endog_genes, ],
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-downsample"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-downsample-1.png" alt="PCA plot of the tung data after downsampling" width="90%" />
<p class="caption">
Figure 3.55: PCA plot of the tung data after downsampling
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotRLE</span>(
    umi.qc[endog_genes, ], 
    <span class="dt">exprs_mats =</span> <span class="kw">list</span>(<span class="dt">Raw =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="dt">DownSample =</span> <span class="st">&quot;logcounts&quot;</span>),
    <span class="dt">exprs_logged =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>),
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-downsample"></span>
<img src="13-exprs-norm_files/figure-html/norm-ours-rle-downsample-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.56: Cell-wise RLE of the tung data
</p>
</div>
</div>
<div id="normalisation-for-genetranscript-length" class="section level3">
<h3><span class="header-section-number">3.15.8</span> Normalisation for gene/transcript length</h3>
<p>Some methods combine library size and fragment/gene length normalization such as:</p>
<ul>
<li><strong>RPKM</strong> - Reads Per Kilobase Million (for single-end sequencing)</li>
<li><strong>FPKM</strong> - Fragments Per Kilobase Million (same as <strong>RPKM</strong> but for paired-end sequencing, makes sure that paired ends mapped to the same fragment are not counted twice)</li>
<li><strong>TPM</strong> - Transcripts Per Kilobase Million (same as <strong>RPKM</strong>, but the order of normalizations is reversed - length first and sequencing depth second)</li>
</ul>
<p>These methods are not applicable to our dataset since the end of the transcript which contains the UMI was preferentially sequenced. Furthermore in general these should only be calculated using appropriate quantification software from aligned BAM files not from read counts since often only a portion of the entire gene/transcript is sequenced, not the entire length. If in doubt check for a relationship between gene/transcript length and expression level.</p>
<p>However, here we show how these normalisations can be calculated using <code>scater</code>. First, we need to find the effective transcript length in Kilobases. However, our dataset containes only gene IDs, therefore we will be using the gene lengths instead of transcripts. <code>scater</code> uses the <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package, which allows one to annotate genes by other attributes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi.qc &lt;-<span class="st"> </span><span class="kw">getBMFeatureAnnos</span>(
    umi.qc,
    <span class="dt">filters =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>, 
    <span class="dt">attributes =</span> <span class="kw">c</span>(
        <span class="st">&quot;ensembl_gene_id&quot;</span>,
        <span class="st">&quot;hgnc_symbol&quot;</span>,
        <span class="st">&quot;chromosome_name&quot;</span>,
        <span class="st">&quot;start_position&quot;</span>,
        <span class="st">&quot;end_position&quot;</span>
    ), 
    <span class="dt">feature_symbol =</span> <span class="st">&quot;hgnc_symbol&quot;</span>,
    <span class="dt">feature_id =</span> <span class="st">&quot;ensembl_gene_id&quot;</span>,
    <span class="dt">biomart =</span> <span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, 
    <span class="dt">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>,
    <span class="dt">host =</span> <span class="st">&quot;www.ensembl.org&quot;</span>
)

<span class="co"># If you have mouse data, change the arguments based on this example:</span>
<span class="co"># getBMFeatureAnnos(</span>
<span class="co">#     object,</span>
<span class="co">#     filters = &quot;ensembl_transcript_id&quot;,</span>
<span class="co">#     attributes = c(</span>
<span class="co">#         &quot;ensembl_transcript_id&quot;,</span>
<span class="co">#         &quot;ensembl_gene_id&quot;, </span>
<span class="co">#         &quot;mgi_symbol&quot;,</span>
<span class="co">#         &quot;chromosome_name&quot;,</span>
<span class="co">#         &quot;transcript_biotype&quot;,</span>
<span class="co">#         &quot;transcript_start&quot;,</span>
<span class="co">#         &quot;transcript_end&quot;,</span>
<span class="co">#         &quot;transcript_count&quot;</span>
<span class="co">#     ),</span>
<span class="co">#     feature_symbol = &quot;mgi_symbol&quot;,</span>
<span class="co">#     feature_id = &quot;ensembl_gene_id&quot;,</span>
<span class="co">#     biomart = &quot;ENSEMBL_MART_ENSEMBL&quot;,</span>
<span class="co">#     dataset = &quot;mmusculus_gene_ensembl&quot;,</span>
<span class="co">#     host = &quot;www.ensembl.org&quot;</span>
<span class="co"># )</span></code></pre></div>
<p>Some of the genes were not annotated, therefore we filter them out:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi.qc.ann &lt;-<span class="st"> </span>umi.qc[<span class="op">!</span><span class="kw">is.na</span>(<span class="kw">rowData</span>(umi.qc)<span class="op">$</span>ensembl_gene_id), ]</code></pre></div>
<p>Now we compute the total gene length in Kilobases by using the <code>end_position</code> and <code>start_position</code> fields:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">eff_length &lt;-<span class="st"> </span>
<span class="st">    </span><span class="kw">abs</span>(<span class="kw">rowData</span>(umi.qc.ann)<span class="op">$</span>end_position <span class="op">-</span><span class="st"> </span><span class="kw">rowData</span>(umi.qc.ann)<span class="op">$</span>start_position) <span class="op">/</span><span class="st"> </span><span class="dv">1000</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(eff_length, <span class="kw">rowMeans</span>(<span class="kw">counts</span>(umi.qc.ann)))</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:length-vs-mean"></span>
<img src="13-exprs-norm_files/figure-html/length-vs-mean-1.png" alt="Gene length vs Mean Expression for the raw data" width="90%" />
<p class="caption">
Figure 3.57: Gene length vs Mean Expression for the raw data
</p>
</div>
<p>There is no relationship between gene length and mean expression so __FPKM__s &amp; __TPM__s are inappropriate for this dataset. But we will demonstrate them anyway.</p>
<p><strong>Note</strong> Here calculate the total gene length instead of the total exon length. Many genes will contain lots of introns so their <code>eff_length</code> will be very different from what we have calculated. Please consider our calculation as approximation. If you want to use the total exon lengths, please refer to <a href="https://www.biostars.org/p/83901/">this page</a>.</p>
<p>Now we are ready to perform the normalisations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tpm</span>(umi.qc.ann) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">calculateTPM</span>(umi.qc.ann, eff_length) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<p>Plot the results as a PCA plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc.ann,
    <span class="dt">exprs_values =</span> <span class="st">&quot;tpm&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-fpkm"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-fpkm-1.png" alt="PCA plot of the tung data after TPM normalisation" width="90%" />
<p class="caption">
Figure 3.58: PCA plot of the tung data after TPM normalisation
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tpm</span>(umi.qc.ann) &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">calculateFPKM</span>(umi.qc.ann, eff_length) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotPCA</span>(
    umi.qc.ann,
    <span class="dt">exprs_values =</span> <span class="st">&quot;tpm&quot;</span>,
    <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
    <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
    <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>
)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-tpm"></span>
<img src="13-exprs-norm_files/figure-html/norm-pca-tpm-1.png" alt="PCA plot of the tung data after FPKM normalisation" width="90%" />
<p class="caption">
Figure 3.59: PCA plot of the tung data after FPKM normalisation
</p>
</div>
<p><strong>Note</strong> The <code>PCA</code> looks for differences between cells. Gene length is the same across cells for each gene thus <strong>FPKM</strong> is almost identical to the <strong>CPM</strong> plot (it is just rotated) since it performs <strong>CPM</strong> first then normalizes gene length. Whereas, <strong>TPM</strong> is different because it weights genes by their length before performing <strong>CPM</strong>.</p>
</div>
<div id="exercise-1" class="section level3">
<h3><span class="header-section-number">3.15.9</span> Exercise</h3>
<p>Perform the same analysis with read counts of the <code>tung</code> data. Use <code>tung/reads.rds</code> file to load the reads <code>SCE</code> object. Once you have finished please compare your results to ours (next chapter).</p>
</div>
<div id="sessioninfo-3" class="section level3">
<h3><span class="header-section-number">3.15.10</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] scran_1.6.2                BiocParallel_1.12.0       
##  [3] scater_1.6.0               SingleCellExperiment_1.0.0
##  [5] SummarizedExperiment_1.8.0 DelayedArray_0.4.1        
##  [7] matrixStats_0.52.2         GenomicRanges_1.30.0      
##  [9] GenomeInfoDb_1.14.0        IRanges_2.12.0            
## [11] S4Vectors_0.16.0           ggplot2_2.2.1             
## [13] Biobase_2.38.0             BiocGenerics_0.24.0       
## [15] knitr_1.17                 scRNA.seq.funcs_0.1.0     
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6            bit64_0.9-7            
##  [3] progress_1.1.2          rprojroot_1.2          
##  [5] dynamicTreeCut_1.63-1   tools_3.4.2            
##  [7] backports_1.1.1         DT_0.2                 
##  [9] R6_2.2.2                hypergeo_1.2-13        
## [11] vipor_0.4.5             DBI_0.7                
## [13] lazyeval_0.2.1          colorspace_1.3-2       
## [15] gridExtra_2.3           prettyunits_1.0.2      
## [17] moments_0.14            bit_1.1-12             
## [19] compiler_3.4.2          orthopolynom_1.0-5     
## [21] labeling_0.3            bookdown_0.5           
## [23] scales_0.5.0            stringr_1.2.0          
## [25] digest_0.6.12           rmarkdown_1.7          
## [27] XVector_0.18.0          pkgconfig_2.0.1        
## [29] htmltools_0.3.6         highr_0.6              
## [31] limma_3.34.1            htmlwidgets_0.9        
## [33] rlang_0.1.4             RSQLite_2.0            
## [35] FNN_1.1                 shiny_1.0.5            
## [37] bindr_0.1               zoo_1.8-0              
## [39] dplyr_0.7.4             RCurl_1.95-4.8         
## [41] magrittr_1.5            GenomeInfoDbData_0.99.1
## [43] Matrix_1.2-7.1          Rcpp_0.12.13           
## [45] ggbeeswarm_0.6.0        munsell_0.4.3          
## [47] viridis_0.4.0           stringi_1.1.5          
## [49] yaml_2.1.14             edgeR_3.20.1           
## [51] MASS_7.3-45             zlibbioc_1.24.0        
## [53] rhdf5_2.22.0            Rtsne_0.13             
## [55] plyr_1.8.4              grid_3.4.2             
## [57] blob_1.1.0              shinydashboard_0.6.1   
## [59] contfrac_1.1-11         lattice_0.20-34        
## [61] cowplot_0.9.0           splines_3.4.2          
## [63] locfit_1.5-9.1          igraph_1.1.2           
## [65] rjson_0.2.15            reshape2_1.4.2         
## [67] biomaRt_2.34.0          XML_3.98-1.9           
## [69] glue_1.2.0              evaluate_0.10.1        
## [71] data.table_1.10.4-3     deSolve_1.20           
## [73] httpuv_1.3.5            gtable_0.2.0           
## [75] assertthat_0.2.0        mime_0.5               
## [77] xtable_1.8-2            viridisLite_0.2.0      
## [79] tibble_1.3.4            elliptic_1.3-7         
## [81] AnnotationDbi_1.40.0    beeswarm_0.2.3         
## [83] memoise_1.1.0           tximport_1.6.0         
## [85] bindrcpp_0.2            statmod_1.4.30</code></pre>

</div>
</div>
<div id="normalization-practice-reads" class="section level2">
<h2><span class="header-section-number">3.16</span> Normalization practice (Reads)</h2>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-raw-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-raw-reads-1.png" alt="PCA plot of the tung data" width="90%" />
<p class="caption">
Figure 3.60: PCA plot of the tung data
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-cpm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-cpm-reads-1.png" alt="PCA plot of the tung data after CPM normalisation" width="90%" />
<p class="caption">
Figure 3.61: PCA plot of the tung data after CPM normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-cpm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-cpm-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.62: Cell-wise RLE of the tung data
</p>
</div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-rle-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-rle-reads-1.png" alt="PCA plot of the tung data after RLE normalisation" width="90%" />
<p class="caption">
Figure 3.63: PCA plot of the tung data after RLE normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-rle-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-rle-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.64: Cell-wise RLE of the tung data
</p>
</div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-uq-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-uq-reads-1.png" alt="PCA plot of the tung data after UQ normalisation" width="90%" />
<p class="caption">
Figure 3.65: PCA plot of the tung data after UQ normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-uq-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-uq-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.66: Cell-wise RLE of the tung data
</p>
</div>
<pre><code>## Warning in normalizeSCE(object, exprs_values = exprs_values, return_log
## = return_log, : spike-in transcripts in &#39;ERCC&#39; should have their own size
## factors</code></pre>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-tmm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-tmm-reads-1.png" alt="PCA plot of the tung data after TMM normalisation" width="90%" />
<p class="caption">
Figure 3.67: PCA plot of the tung data after TMM normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-tmm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-tmm-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.68: Cell-wise RLE of the tung data
</p>
</div>
<pre><code>## Warning in .local(object, ...): spike-in transcripts in &#39;ERCC&#39; should have
## their own size factors</code></pre>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-lsf-umi"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-lsf-umi-1.png" alt="PCA plot of the tung data after LSF normalisation" width="90%" />
<p class="caption">
Figure 3.69: PCA plot of the tung data after LSF normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-scran-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-scran-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.70: Cell-wise RLE of the tung data
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-downsample-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-downsample-reads-1.png" alt="PCA plot of the tung data after downsampling" width="90%" />
<p class="caption">
Figure 3.71: PCA plot of the tung data after downsampling
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-ours-rle-downsample-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-ours-rle-downsample-reads-1.png" alt="Cell-wise RLE of the tung data" width="90%" />
<p class="caption">
Figure 3.72: Cell-wise RLE of the tung data
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-tpm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-tpm-reads-1.png" alt="PCA plot of the tung data after TPM normalisation" width="90%" />
<p class="caption">
Figure 3.73: PCA plot of the tung data after TPM normalisation
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:norm-pca-fpkm-reads"></span>
<img src="14-exprs-norm-reads_files/figure-html/norm-pca-fpkm-reads-1.png" alt="PCA plot of the tung data after FPKM normalisation" width="90%" />
<p class="caption">
Figure 3.74: PCA plot of the tung data after FPKM normalisation
</p>
</div>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] knitr_1.17                 scran_1.6.2               
##  [3] BiocParallel_1.12.0        scater_1.6.0              
##  [5] SingleCellExperiment_1.0.0 SummarizedExperiment_1.8.0
##  [7] DelayedArray_0.4.1         matrixStats_0.52.2        
##  [9] GenomicRanges_1.30.0       GenomeInfoDb_1.14.0       
## [11] IRanges_2.12.0             S4Vectors_0.16.0          
## [13] ggplot2_2.2.1              Biobase_2.38.0            
## [15] BiocGenerics_0.24.0        scRNA.seq.funcs_0.1.0     
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6            bit64_0.9-7            
##  [3] progress_1.1.2          rprojroot_1.2          
##  [5] dynamicTreeCut_1.63-1   tools_3.4.2            
##  [7] backports_1.1.1         DT_0.2                 
##  [9] R6_2.2.2                hypergeo_1.2-13        
## [11] vipor_0.4.5             DBI_0.7                
## [13] lazyeval_0.2.1          colorspace_1.3-2       
## [15] gridExtra_2.3           prettyunits_1.0.2      
## [17] moments_0.14            bit_1.1-12             
## [19] compiler_3.4.2          orthopolynom_1.0-5     
## [21] labeling_0.3            bookdown_0.5           
## [23] scales_0.5.0            stringr_1.2.0          
## [25] digest_0.6.12           rmarkdown_1.7          
## [27] XVector_0.18.0          pkgconfig_2.0.1        
## [29] htmltools_0.3.6         highr_0.6              
## [31] limma_3.34.1            htmlwidgets_0.9        
## [33] rlang_0.1.4             RSQLite_2.0            
## [35] FNN_1.1                 shiny_1.0.5            
## [37] bindr_0.1               zoo_1.8-0              
## [39] dplyr_0.7.4             RCurl_1.95-4.8         
## [41] magrittr_1.5            GenomeInfoDbData_0.99.1
## [43] Matrix_1.2-7.1          Rcpp_0.12.13           
## [45] ggbeeswarm_0.6.0        munsell_0.4.3          
## [47] viridis_0.4.0           stringi_1.1.5          
## [49] yaml_2.1.14             edgeR_3.20.1           
## [51] MASS_7.3-45             zlibbioc_1.24.0        
## [53] rhdf5_2.22.0            Rtsne_0.13             
## [55] plyr_1.8.4              grid_3.4.2             
## [57] blob_1.1.0              shinydashboard_0.6.1   
## [59] contfrac_1.1-11         lattice_0.20-34        
## [61] cowplot_0.9.0           splines_3.4.2          
## [63] locfit_1.5-9.1          igraph_1.1.2           
## [65] rjson_0.2.15            reshape2_1.4.2         
## [67] biomaRt_2.34.0          XML_3.98-1.9           
## [69] glue_1.2.0              evaluate_0.10.1        
## [71] data.table_1.10.4-3     deSolve_1.20           
## [73] httpuv_1.3.5            gtable_0.2.0           
## [75] assertthat_0.2.0        mime_0.5               
## [77] xtable_1.8-2            viridisLite_0.2.0      
## [79] tibble_1.3.4            elliptic_1.3-7         
## [81] AnnotationDbi_1.40.0    beeswarm_0.2.3         
## [83] memoise_1.1.0           tximport_1.6.0         
## [85] bindrcpp_0.2            statmod_1.4.30</code></pre>

</div>
<div id="dealing-with-confounders" class="section level2">
<h2><span class="header-section-number">3.17</span> Dealing with confounders</h2>
<div id="introduction-5" class="section level3">
<h3><span class="header-section-number">3.17.1</span> Introduction</h3>
<p>In the previous chapter we normalized for library size, effectively removing it as a confounder. Now we will consider removing other less well defined confounders from our data. Technical confounders (aka batch effects) can arise from difference in reagents, isolation methods, the lab/experimenter who performed the experiment, even which day/time the experiment was performed. Accounting for technical confounders, and batch effects particularly, is a large topic that also involves principles of experimental design. Here we address approaches that can be taken to account for confounders when the experimental design is appropriate.</p>
<p>Fundamentally, accounting for technical confounders involves identifying and, ideally, removing sources of variation in the expression data that are not related to (i.e. are confounding) the biological signal of interest. Various approaches exist, some of which use spike-in or housekeeping genes, and some of which use endogenous genes.</p>
<div id="advantages-and-disadvantages-of-using-spike-ins-to-remove-confounders" class="section level4">
<h4><span class="header-section-number">3.17.1.1</span> Advantages and disadvantages of using spike-ins to remove confounders</h4>
<p>The use of spike-ins as control genes is appealing, since the same amount of ERCC (or other) spike-in was added to each cell in our experiment. In principle, all the variablity we observe for these genes is due to technical noise; whereas endogenous genes are affected by both technical noise and biological variability. Technical noise can be removed by fitting a model to the spike-ins and “substracting” this from the endogenous genes. There are several methods available based on this premise (eg. <a href="https://github.com/catavallejos/BASiCS">BASiCS</a>, <a href="https://github.com/PMBio/scLVM">scLVM</a>, <a href="http://bioconductor.org/packages/release/bioc/html/RUVSeq.html">RUVg</a>); each using different noise models and different fitting procedures. Alternatively, one can identify genes which exhibit significant variation beyond technical noise (eg. Distance to median, <a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html">Highly variable genes</a>). However, there are issues with the use of spike-ins for normalisation (particularly ERCCs, derived from bacterial sequences), including that their variability can, for various reasons, actually be <em>higher</em> than that of endogenous genes.</p>
<p>Given the issues with using spike-ins, better results can often be obtained by using endogenous genes instead. Where we have a large number of endogenous genes that, on average, do not vary systematically between cells and where we expect technical effects to affect a large number of genes (a very common and reasonable assumption), then such methods (for example, the RUVs method) can perform well.</p>
<p>We explore both general approaches below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNA.seq.funcs)
<span class="kw">library</span>(RUVSeq)
<span class="kw">library</span>(scater)
<span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scran)
<span class="kw">library</span>(kBET)
<span class="kw">library</span>(sva) <span class="co"># Combat</span>
<span class="kw">library</span>(edgeR)
<span class="kw">set.seed</span>(<span class="dv">1234567</span>)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
umi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/umi.rds&quot;</span>)
umi.qc &lt;-<span class="st"> </span>umi[<span class="kw">rowData</span>(umi)<span class="op">$</span>use, <span class="kw">colData</span>(umi)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(umi.qc)<span class="op">$</span>is_feature_control
erccs &lt;-<span class="st"> </span><span class="kw">rowData</span>(umi.qc)<span class="op">$</span>is_feature_control

qclust &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(umi.qc, <span class="dt">min.size =</span> <span class="dv">30</span>)
umi.qc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(umi.qc, <span class="dt">sizes =</span> <span class="dv">15</span>, <span class="dt">clusters =</span> qclust)
umi.qc &lt;-<span class="st"> </span><span class="kw">normalize</span>(umi.qc)</code></pre></div>
</div>
</div>
<div id="remove-unwanted-variation" class="section level3">
<h3><span class="header-section-number">3.17.2</span> Remove Unwanted Variation</h3>
<p>Factors contributing to technical noise frequently appear as “batch effects” where cells processed on different days or by different technicians systematically vary from one another. Removing technical noise and correcting for batch effects can frequently be performed using the same tool or slight variants on it. We will be considering the <a href="http://bioconductor.org/packages/RUVSeq">Remove Unwanted Variation (RUVSeq)</a>. Briefly, RUVSeq works as follows. For <span class="math inline">\(n\)</span> samples and <span class="math inline">\(J\)</span> genes, consider the following generalized linear model (GLM), where the RNA-Seq read counts are regressed on both the known covariates of interest and unknown factors of unwanted variation: <span class="math display">\[\log E[Y|W,X,O] = W\alpha + X\beta + O\]</span> Here, <span class="math inline">\(Y\)</span> is the <span class="math inline">\(n \times J\)</span> matrix of observed gene-level read counts, <span class="math inline">\(W\)</span> is an <span class="math inline">\(n \times k\)</span> matrix corresponding to the factors of “unwanted variation” and <span class="math inline">\(O\)</span> is an <span class="math inline">\(n \times J\)</span> matrix of offsets that can either be set to zero or estimated with some other normalization procedure (such as upper-quartile normalization). The simultaneous estimation of <span class="math inline">\(W\)</span>, <span class="math inline">\(\alpha\)</span>, <span class="math inline">\(\beta\)</span>, and <span class="math inline">\(k\)</span> is infeasible. For a given <span class="math inline">\(k\)</span>, instead the following three approaches to estimate the factors of unwanted variation <span class="math inline">\(W\)</span> are used:</p>
<ul>
<li><em>RUVg</em> uses negative control genes (e.g. ERCCs), assumed to have constant expression across samples;</li>
<li><em>RUVs</em> uses centered (technical) replicate/negative control samples for which the covariates of interest are constant;</li>
<li><em>RUVr</em> uses residuals, e.g., from a first-pass GLM regression of the counts on the covariates of interest.</li>
</ul>
<p>We will concentrate on the first two approaches.</p>
<div id="ruvg" class="section level4">
<h4><span class="header-section-number">3.17.2.1</span> RUVg</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(umi.qc), erccs, <span class="dt">k =</span> <span class="dv">1</span>)
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;ruvg1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(umi.qc), erccs, <span class="dt">k =</span> <span class="dv">10</span>)
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;ruvg10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
</div>
<div id="ruvs" class="section level4">
<h4><span class="header-section-number">3.17.2.2</span> RUVs</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scIdx &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw">max</span>(<span class="kw">table</span>(umi.qc<span class="op">$</span>individual)), <span class="dt">nrow =</span> <span class="dv">3</span>)
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>)
scIdx[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>)
scIdx[<span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>)
scIdx[<span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
cIdx &lt;-<span class="st"> </span><span class="kw">rownames</span>(umi.qc)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(umi.qc), cIdx, <span class="dt">k =</span> <span class="dv">1</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;ruvs1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(umi.qc), cIdx, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;ruvs10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
</div>
</div>
<div id="combat" class="section level3">
<h3><span class="header-section-number">3.17.3</span> Combat</h3>
<p>If you have an experiment with a balanced design, <code>Combat</code> can be used to eliminate batch effects while preserving biological effects by specifying the biological effects using the <code>mod</code> parameter. However the <code>Tung</code> data contains multiple experimental replicates rather than a balanced design so using <code>mod1</code> to preserve biological variability will result in an error.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">combat_data &lt;-<span class="st"> </span><span class="kw">logcounts</span>(umi.qc)
mod_data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(combat_data))
<span class="co"># Basic batch removal</span>
mod0 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">data =</span> mod_data) 
<span class="co"># Preserve biological variability</span>
mod1 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>umi.qc<span class="op">$</span>individual, <span class="dt">data =</span> mod_data) 
<span class="co"># adjust for total genes detected</span>
mod2 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>umi.qc<span class="op">$</span>total_features, <span class="dt">data =</span> mod_data)
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;combat&quot;</span>) &lt;-<span class="st"> </span><span class="kw">ComBat</span>(
    <span class="dt">dat =</span> <span class="kw">t</span>(mod_data), 
    <span class="dt">batch =</span> <span class="kw">factor</span>(umi.qc<span class="op">$</span>batch), 
    <span class="dt">mod =</span> mod0,
    <span class="dt">par.prior =</span> <span class="ot">TRUE</span>,
    <span class="dt">prior.plots =</span> <span class="ot">FALSE</span>
)</code></pre></div>
<pre><code>## Standardizing Data across genes</code></pre>
<p><strong>Exercise 1</strong></p>
<p>Perform <code>ComBat</code> correction accounting for total features as a co-variate. Store the corrected matrix in the <code>combat_tf</code> slot.</p>
<pre><code>## Standardizing Data across genes</code></pre>
</div>
<div id="mnncorrect" class="section level3">
<h3><span class="header-section-number">3.17.4</span> mnnCorrect</h3>
<p><code>mnnCorrect</code> <span class="citation">(Haghverdi et al. <a href="#ref-Haghverdi2017-vh">2017</a>)</span> assumes that each batch shares at least one biological condition with each other batch. Thus it works well for a variety of balanced experimental designs. However, the <code>Tung</code> data contains multiple replicates for each invidividual rather than balanced batches, thus we will normalized each individual separately. Note that this will remove batch effects between batches within the same individual but not the batch effects between batches in different individuals, due to the confounded experimental design.</p>
<p>Thus we will merge a replicate from each individual to form three batches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">do_mnn &lt;-<span class="st"> </span><span class="cf">function</span>(data.qc) {
    batch1 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r1&quot;</span>]
    batch2 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r2&quot;</span>]
    batch3 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r3&quot;</span>]
    
    <span class="cf">if</span> (<span class="kw">ncol</span>(batch2) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch2), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]], x<span class="op">$</span>corrected[[<span class="dv">3</span>]]))
    } <span class="cf">else</span> {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]]))
    }
}

indi1 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(umi.qc[, umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>])
indi2 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(umi.qc[, umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>])
indi3 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(umi.qc[, umi.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>])

<span class="kw">assay</span>(umi.qc, <span class="st">&quot;mnn&quot;</span>) &lt;-<span class="st"> </span><span class="kw">cbind</span>(indi1, indi2, indi3);

<span class="co"># For a balanced design: </span>
<span class="co">#assay(umi.qc, &quot;mnn&quot;) &lt;- mnnCorrect(</span>
<span class="co">#    list(B1 = logcounts(batch1), B2 = logcounts(batch2), B3 = logcounts(batch3)),  </span>
<span class="co">#    k = 20,</span>
<span class="co">#    sigma = 0.1,</span>
<span class="co">#    cos.norm = TRUE,</span>
<span class="co">#    svd.dim = 2</span>
<span class="co">#)</span></code></pre></div>
</div>
<div id="glm" class="section level3">
<h3><span class="header-section-number">3.17.5</span> GLM</h3>
<p>A general linear model is a simpler version of <code>Combat</code>. It can correct for batches while preserving biological effects if you have a balanced design. In a confounded/replicate design biological effects will not be fit/preserved. Similar to <code>mnnCorrect</code> we could remove batch effects from each individual separately in order to preserve biological (and technical) variance between individuals. For demonstation purposes we will naively correct all cofounded batch effects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">glm_fun &lt;-<span class="st"> </span><span class="cf">function</span>(g, batch, indi) {
  model &lt;-<span class="st"> </span><span class="kw">glm</span>(g <span class="op">~</span><span class="st"> </span>batch <span class="op">+</span><span class="st"> </span>indi)
  model<span class="op">$</span>coef[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="dv">0</span> <span class="co"># replace intercept with 0 to preserve reference batch.</span>
  <span class="kw">return</span>(model<span class="op">$</span>coef)
}
effects &lt;-<span class="st"> </span><span class="kw">apply</span>(
    <span class="kw">logcounts</span>(umi.qc), 
    <span class="dv">1</span>, 
    glm_fun, 
    <span class="dt">batch =</span> umi.qc<span class="op">$</span>batch, 
    <span class="dt">indi =</span> umi.qc<span class="op">$</span>individual
)
corrected &lt;-<span class="st"> </span><span class="kw">logcounts</span>(umi.qc) <span class="op">-</span><span class="st"> </span><span class="kw">t</span>(effects[<span class="kw">as.numeric</span>(<span class="kw">factor</span>(umi.qc<span class="op">$</span>batch)), ])
<span class="kw">assay</span>(umi.qc, <span class="st">&quot;glm&quot;</span>) &lt;-<span class="st"> </span>corrected</code></pre></div>
<p><strong>Exercise 2</strong></p>
<p>Perform GLM correction for each individual separately. Store the final corrected matrix in the <code>glm_indi</code> slot.</p>
</div>
<div id="how-to-evaluate-and-compare-confounder-removal-strategies" class="section level3">
<h3><span class="header-section-number">3.17.6</span> How to evaluate and compare confounder removal strategies</h3>
<p>A key question when considering the different methods for removing confounders is how to quantitatively determine which one is the most effective. The main reason why comparisons are challenging is because it is often difficult to know what corresponds to technical counfounders and what is interesting biological variability. Here, we consider three different metrics which are all reasonable based on our knowledge of the experimental design. Depending on the biological question that you wish to address, it is important to choose a metric that allows you to evaluate the confounders that are likely to be the biggest concern for the given situation.</p>
<div id="effectiveness-1" class="section level4">
<h4><span class="header-section-number">3.17.6.1</span> Effectiveness 1</h4>
<p>We evaluate the effectiveness of the normalization by inspecting the PCA plot where colour corresponds the technical replicates and shape corresponds to different biological samples (individuals). Separation of biological samples and interspersed batches indicates that technical variation has been removed. We always use log2-cpm normalized data to match the assumptions of PCA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(umi.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotPCA</span>(
            umi.qc[endog_genes, ],
            <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
            <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
            <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
            <span class="dt">exprs_values =</span> n
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<p><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-3.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-4.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-5.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-6.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-7.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-8.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-9.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-10.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-11.png" width="672" style="display: block; margin: auto;" /><img src="15-remove-conf_files/figure-html/unnamed-chunk-10-12.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Exercise 3</strong></p>
<p>Consider different <code>ks</code> for RUV normalizations. Which gives the best results?</p>
</div>
<div id="effectiveness-2" class="section level4">
<h4><span class="header-section-number">3.17.6.2</span> Effectiveness 2</h4>
<p>We can also examine the effectiveness of correction using the relative log expression (RLE) across cells to confirm technical noise has been removed from the dataset. Note RLE only evaluates whether the number of genes higher and lower than average are equal for each cell - i.e. systemic technical effects. Random technical noise between batches may not be detected by RLE.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(umi.qc)) {
    res[[n]] &lt;-<span class="st"> </span><span class="kw">suppressWarnings</span>(<span class="kw">calc_cell_RLE</span>(<span class="kw">assay</span>(umi.qc, n), erccs))
}
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">boxplot</span>(res, <span class="dt">las=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="15-remove-conf_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effectiveness-3" class="section level4">
<h4><span class="header-section-number">3.17.6.3</span> Effectiveness 3</h4>
<p>We can repeat the analysis from Chapter 12 to check whether batch effects have been removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(umi.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotQC</span>(
            umi.qc[endog_genes, ],
            <span class="dt">type =</span> <span class="st">&quot;expl&quot;</span>,
            <span class="dt">exprs_values =</span> n,
            <span class="dt">variables =</span> <span class="kw">c</span>(
                <span class="st">&quot;total_features&quot;</span>,
                <span class="st">&quot;total_counts&quot;</span>,
                <span class="st">&quot;batch&quot;</span>,
                <span class="st">&quot;individual&quot;</span>,
                <span class="st">&quot;pct_counts_ERCC&quot;</span>,
                <span class="st">&quot;pct_counts_MT&quot;</span>
            )
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-1.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-2.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-3.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-4.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-5.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-6.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-7.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-8.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-9.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-10.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-11.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:confound-cpm"></span>
<img src="15-remove-conf_files/figure-html/confound-cpm-12.png" alt="Explanatory variables (mnn)" width="672" />
<p class="caption">
Figure 3.75: Explanatory variables (mnn)
</p>
</div>
<p><strong>Exercise 4</strong></p>
<p>Perform the above analysis for each normalization/batch correction method. Which method(s) are most/least effective? Why is the variance accounted for by batch never lower than the variance accounted for by individual?</p>
</div>
<div id="effectiveness-4" class="section level4">
<h4><span class="header-section-number">3.17.6.4</span> Effectiveness 4</h4>
<p>Another method to check the efficacy of batch-effect correction is to consider the intermingling of points from different batches in local subsamples of the data. If there are no batch-effects then proportion of cells from each batch in any local region should be equal to the global proportion of cells in each batch.</p>
<p><code>kBET</code> <span class="citation">(Buttner et al. <a href="#ref-Buttner2017-ds">2017</a>)</span> takes <code>kNN</code> networks around random cells and tests the number of cells from each batch against a binomial distribution. The rejection rate of these tests indicates the severity of batch-effects still present in the data (high rejection rate = strong batch effects). <code>kBET</code> assumes each batch contains the same complement of biological groups, thus it can only be applied to the entire dataset if a perfectly balanced design has been used. However, <code>kBET</code> can also be applied to replicate-data if it is applied to each biological group separately. In the case of the Tung data, we will apply <code>kBET</code> to each individual independently to check for residual batch effects. However, this method will not identify residual batch-effects which are confounded with biological conditions. In addition, <code>kBET</code> does not determine if biological signal has been preserved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">compare_kBET_results &lt;-<span class="st"> </span><span class="cf">function</span>(sce){
    indiv &lt;-<span class="st"> </span><span class="kw">unique</span>(sce<span class="op">$</span>individual)
    norms &lt;-<span class="st"> </span><span class="kw">assayNames</span>(sce) <span class="co"># Get all normalizations</span>
    results &lt;-<span class="st"> </span><span class="kw">list</span>()
    <span class="cf">for</span> (i <span class="cf">in</span> indiv){ 
        <span class="cf">for</span> (j <span class="cf">in</span> norms){
            tmp &lt;-<span class="st"> </span><span class="kw">kBET</span>(
                <span class="dt">df =</span> <span class="kw">t</span>(<span class="kw">assay</span>(sce[,sce<span class="op">$</span>individual<span class="op">==</span><span class="st"> </span>i], j)), 
                <span class="dt">batch =</span> sce<span class="op">$</span>batch[sce<span class="op">$</span>individual<span class="op">==</span>i], 
                <span class="dt">heuristic =</span> <span class="ot">TRUE</span>, 
                <span class="dt">verbose =</span> <span class="ot">FALSE</span>, 
                <span class="dt">addTest =</span> <span class="ot">FALSE</span>, 
                <span class="dt">plot =</span> <span class="ot">FALSE</span>)
            results[[i]][[j]] &lt;-<span class="st"> </span>tmp<span class="op">$</span>summary<span class="op">$</span>kBET.observed[<span class="dv">1</span>]
        }
    }
    <span class="kw">return</span>(<span class="kw">as.data.frame</span>(results))
}

eff_debatching &lt;-<span class="st"> </span><span class="kw">compare_kBET_results</span>(umi.qc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(<span class="st">&quot;reshape2&quot;</span>)
<span class="kw">require</span>(<span class="st">&quot;RColorBrewer&quot;</span>)
<span class="co"># Plot results</span>
dod &lt;-<span class="st"> </span><span class="kw">melt</span>(<span class="kw">as.matrix</span>(eff_debatching),  <span class="dt">value.name =</span> <span class="st">&quot;kBET&quot;</span>)
<span class="kw">colnames</span>(dod)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Normalisation&quot;</span>, <span class="st">&quot;Individual&quot;</span>)

colorset &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;gray&#39;</span>, <span class="kw">brewer.pal</span>(<span class="dt">n =</span> <span class="dv">9</span>, <span class="st">&quot;RdYlBu&quot;</span>))

<span class="kw">ggplot</span>(dod, <span class="kw">aes</span>(Normalisation, Individual, <span class="dt">fill=</span>kBET)) <span class="op">+</span><span class="st">  </span>
<span class="st">    </span><span class="kw">geom_tile</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradient2</span>(
        <span class="dt">na.value =</span> <span class="st">&quot;gray&quot;</span>,
        <span class="dt">low =</span> colorset[<span class="dv">2</span>],
        <span class="dt">mid=</span>colorset[<span class="dv">6</span>],
        <span class="dt">high =</span> colorset[<span class="dv">10</span>],
        <span class="dt">midpoint =</span> <span class="fl">0.5</span>, <span class="dt">limit =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_x_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_y_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(
            <span class="dt">angle =</span> <span class="dv">45</span>, 
            <span class="dt">vjust =</span> <span class="dv">1</span>, 
            <span class="dt">size =</span> <span class="dv">12</span>, 
            <span class="dt">hjust =</span> <span class="dv">1</span>
        )
    ) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Effect of batch regression methods per individual&quot;</span>)</code></pre></div>
<p><img src="15-remove-conf_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Exercise 5</strong></p>
<p>Why do the raw counts appear to have little batch effects?</p>
</div>
</div>
<div id="big-exercise-2" class="section level3">
<h3><span class="header-section-number">3.17.7</span> Big Exercise</h3>
<p>Perform the same analysis with read counts of the <code>tung</code> data. Use <code>tung/reads.rds</code> file to load the reads <code>SCE</code> object. Once you have finished please compare your results to ours (next chapter). Additionally, experiment with other combinations of normalizations and compare the results.</p>
</div>
<div id="sessioninfo-4" class="section level3">
<h3><span class="header-section-number">3.17.8</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] RColorBrewer_1.1-2         reshape2_1.4.2            
##  [3] sva_3.26.0                 genefilter_1.60.0         
##  [5] mgcv_1.8-22                nlme_3.1-129              
##  [7] kBET_0.99.0                scran_1.6.2               
##  [9] scater_1.6.0               SingleCellExperiment_1.0.0
## [11] ggplot2_2.2.1              RUVSeq_1.12.0             
## [13] edgeR_3.20.1               limma_3.34.1              
## [15] EDASeq_2.12.0              ShortRead_1.36.0          
## [17] GenomicAlignments_1.14.0   SummarizedExperiment_1.8.0
## [19] DelayedArray_0.4.1         matrixStats_0.52.2        
## [21] Rsamtools_1.30.0           GenomicRanges_1.30.0      
## [23] GenomeInfoDb_1.14.0        Biostrings_2.46.0         
## [25] XVector_0.18.0             IRanges_2.12.0            
## [27] S4Vectors_0.16.0           BiocParallel_1.12.0       
## [29] Biobase_2.38.0             BiocGenerics_0.24.0       
## [31] scRNA.seq.funcs_0.1.0      knitr_1.17                
## 
## loaded via a namespace (and not attached):
##  [1] Rtsne_0.13              ggbeeswarm_0.6.0       
##  [3] colorspace_1.3-2        rjson_0.2.15           
##  [5] hwriter_1.3.2           dynamicTreeCut_1.63-1  
##  [7] rprojroot_1.2           DT_0.2                 
##  [9] bit64_0.9-7             AnnotationDbi_1.40.0   
## [11] splines_3.4.2           R.methodsS3_1.7.1      
## [13] tximport_1.6.0          DESeq_1.30.0           
## [15] geneplotter_1.56.0      annotate_1.56.1        
## [17] cluster_2.0.6           R.oo_1.21.0            
## [19] shinydashboard_0.6.1    shiny_1.0.5            
## [21] compiler_3.4.2          backports_1.1.1        
## [23] assertthat_0.2.0        Matrix_1.2-7.1         
## [25] lazyeval_0.2.1          htmltools_0.3.6        
## [27] prettyunits_1.0.2       tools_3.4.2            
## [29] igraph_1.1.2            bindrcpp_0.2           
## [31] gtable_0.2.0            glue_1.2.0             
## [33] GenomeInfoDbData_0.99.1 dplyr_0.7.4            
## [35] Rcpp_0.12.13            rtracklayer_1.38.0     
## [37] stringr_1.2.0           mime_0.5               
## [39] hypergeo_1.2-13         statmod_1.4.30         
## [41] XML_3.98-1.9            zoo_1.8-0              
## [43] zlibbioc_1.24.0         MASS_7.3-45            
## [45] scales_0.5.0            aroma.light_3.8.0      
## [47] rhdf5_2.22.0            yaml_2.1.14            
## [49] memoise_1.1.0           gridExtra_2.3          
## [51] biomaRt_2.34.0          latticeExtra_0.6-28    
## [53] stringi_1.1.5           RSQLite_2.0            
## [55] highr_0.6               RMySQL_0.10.13         
## [57] orthopolynom_1.0-5      GenomicFeatures_1.30.0 
## [59] contfrac_1.1-11         rlang_0.1.4            
## [61] pkgconfig_2.0.1         moments_0.14           
## [63] bitops_1.0-6            evaluate_0.10.1        
## [65] lattice_0.20-34         bindr_0.1              
## [67] labeling_0.3            htmlwidgets_0.9        
## [69] cowplot_0.9.0           bit_1.1-12             
## [71] deSolve_1.20            plyr_1.8.4             
## [73] magrittr_1.5            bookdown_0.5           
## [75] R6_2.2.2                DBI_0.7                
## [77] survival_2.40-1         RCurl_1.95-4.8         
## [79] tibble_1.3.4            rmarkdown_1.7          
## [81] viridis_0.4.0           progress_1.1.2         
## [83] locfit_1.5-9.1          grid_3.4.2             
## [85] data.table_1.10.4-3     FNN_1.1                
## [87] blob_1.1.0              digest_0.6.12          
## [89] xtable_1.8-2            httpuv_1.3.5           
## [91] elliptic_1.3-7          R.utils_2.6.0          
## [93] munsell_0.4.3           beeswarm_0.2.3         
## [95] viridisLite_0.2.0       vipor_0.4.5</code></pre>

</div>
</div>
<div id="dealing-with-confounders-reads" class="section level2">
<h2><span class="header-section-number">3.18</span> Dealing with confounders (Reads)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNA.seq.funcs)
<span class="kw">library</span>(RUVSeq)
<span class="kw">library</span>(scater)
<span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scran)
<span class="kw">library</span>(kBET)
<span class="kw">library</span>(sva) <span class="co"># Combat</span>
<span class="kw">library</span>(edgeR)
<span class="kw">set.seed</span>(<span class="dv">1234567</span>)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
reads &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/reads.rds&quot;</span>)
reads.qc &lt;-<span class="st"> </span>reads[<span class="kw">rowData</span>(reads)<span class="op">$</span>use, <span class="kw">colData</span>(reads)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(reads.qc)<span class="op">$</span>is_feature_control
erccs &lt;-<span class="st"> </span><span class="kw">rowData</span>(reads.qc)<span class="op">$</span>is_feature_control

qclust &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(reads.qc, <span class="dt">min.size =</span> <span class="dv">30</span>)
reads.qc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(reads.qc, <span class="dt">sizes =</span> <span class="dv">15</span>, <span class="dt">clusters =</span> qclust)
reads.qc &lt;-<span class="st"> </span><span class="kw">normalize</span>(reads.qc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(reads.qc), erccs, <span class="dt">k =</span> <span class="dv">1</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvg1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(reads.qc), erccs, <span class="dt">k =</span> <span class="dv">10</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvg10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scIdx &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw">max</span>(<span class="kw">table</span>(reads.qc<span class="op">$</span>individual)), <span class="dt">nrow =</span> <span class="dv">3</span>)
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>)
scIdx[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>)
scIdx[<span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>)
scIdx[<span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
cIdx &lt;-<span class="st"> </span><span class="kw">rownames</span>(reads.qc)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(reads.qc), cIdx, <span class="dt">k =</span> <span class="dv">1</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvs1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(reads.qc), cIdx, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvs10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">combat_data &lt;-<span class="st"> </span><span class="kw">logcounts</span>(reads.qc)
mod_data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(combat_data))
<span class="co"># Basic batch removal</span>
mod0 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">data =</span> mod_data) 
<span class="co"># Preserve biological variability</span>
mod1 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>reads.qc<span class="op">$</span>individual, <span class="dt">data =</span> mod_data) 
<span class="co"># adjust for total genes detected</span>
mod2 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>reads.qc<span class="op">$</span>total_features, <span class="dt">data =</span> mod_data)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;combat&quot;</span>) &lt;-<span class="st"> </span><span class="kw">ComBat</span>(
    <span class="dt">dat =</span> <span class="kw">t</span>(mod_data), 
    <span class="dt">batch =</span> <span class="kw">factor</span>(reads.qc<span class="op">$</span>batch), 
    <span class="dt">mod =</span> mod0,
    <span class="dt">par.prior =</span> <span class="ot">TRUE</span>,
    <span class="dt">prior.plots =</span> <span class="ot">FALSE</span>
)</code></pre></div>
<pre><code>## Standardizing Data across genes</code></pre>
<p><strong>Exercise 1</strong></p>
<pre><code>## Standardizing Data across genes</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">do_mnn &lt;-<span class="st"> </span><span class="cf">function</span>(data.qc) {
    batch1 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r1&quot;</span>]
    batch2 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r2&quot;</span>]
    batch3 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r3&quot;</span>]
    
    <span class="cf">if</span> (<span class="kw">ncol</span>(batch2) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch2), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]], x<span class="op">$</span>corrected[[<span class="dv">3</span>]]))
    } <span class="cf">else</span> {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]]))
    }
}

indi1 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>])
indi2 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>])
indi3 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>])

<span class="kw">assay</span>(reads.qc, <span class="st">&quot;mnn&quot;</span>) &lt;-<span class="st"> </span><span class="kw">cbind</span>(indi1, indi2, indi3);

<span class="co"># For a balanced design: </span>
<span class="co">#assay(reads.qc, &quot;mnn&quot;) &lt;- mnnCorrect(</span>
<span class="co">#    list(B1 = logcounts(batch1), B2 = logcounts(batch2), B3 = logcounts(batch3)),  </span>
<span class="co">#    k = 20,</span>
<span class="co">#    sigma = 0.1,</span>
<span class="co">#    cos.norm = TRUE,</span>
<span class="co">#    svd.dim = 2</span>
<span class="co">#)</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">glm_fun &lt;-<span class="st"> </span><span class="cf">function</span>(g, batch, indi) {
  model &lt;-<span class="st"> </span><span class="kw">glm</span>(g <span class="op">~</span><span class="st"> </span>batch <span class="op">+</span><span class="st"> </span>indi)
  model<span class="op">$</span>coef[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="dv">0</span> <span class="co"># replace intercept with 0 to preserve reference batch.</span>
  <span class="kw">return</span>(model<span class="op">$</span>coef)
}
effects &lt;-<span class="st"> </span><span class="kw">apply</span>(
    <span class="kw">logcounts</span>(reads.qc), 
    <span class="dv">1</span>, 
    glm_fun, 
    <span class="dt">batch =</span> reads.qc<span class="op">$</span>batch, 
    <span class="dt">indi =</span> reads.qc<span class="op">$</span>individual
)
corrected &lt;-<span class="st"> </span><span class="kw">logcounts</span>(reads.qc) <span class="op">-</span><span class="st"> </span><span class="kw">t</span>(effects[<span class="kw">as.numeric</span>(<span class="kw">factor</span>(reads.qc<span class="op">$</span>batch)), ])
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;glm&quot;</span>) &lt;-<span class="st"> </span>corrected</code></pre></div>
<p><strong>Exercise 2</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotPCA</span>(
            reads.qc[endog_genes, ],
            <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
            <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
            <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
            <span class="dt">exprs_values =</span> n
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-3.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-4.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-5.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-6.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-7.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-8.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-9.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-10.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-11.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-12.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    res[[n]] &lt;-<span class="st"> </span><span class="kw">suppressWarnings</span>(<span class="kw">calc_cell_RLE</span>(<span class="kw">assay</span>(reads.qc, n), erccs))
}
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">boxplot</span>(res, <span class="dt">las=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotQC</span>(
            reads.qc[endog_genes, ],
            <span class="dt">type =</span> <span class="st">&quot;expl&quot;</span>,
            <span class="dt">exprs_values =</span> n,
            <span class="dt">variables =</span> <span class="kw">c</span>(
                <span class="st">&quot;total_features&quot;</span>,
                <span class="st">&quot;total_counts&quot;</span>,
                <span class="st">&quot;batch&quot;</span>,
                <span class="st">&quot;individual&quot;</span>,
                <span class="st">&quot;pct_counts_ERCC&quot;</span>,
                <span class="st">&quot;pct_counts_MT&quot;</span>
            )
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-2.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-3.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-4.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-5.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-6.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-7.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-8.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-9.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-10.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-11.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-12.png" width="672" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">compare_kBET_results &lt;-<span class="st"> </span><span class="cf">function</span>(sce){
    indiv &lt;-<span class="st"> </span><span class="kw">unique</span>(sce<span class="op">$</span>individual)
    norms &lt;-<span class="st"> </span><span class="kw">assayNames</span>(sce) <span class="co"># Get all normalizations</span>
    results &lt;-<span class="st"> </span><span class="kw">list</span>()
    <span class="cf">for</span> (i <span class="cf">in</span> indiv){ 
        <span class="cf">for</span> (j <span class="cf">in</span> norms){
            tmp &lt;-<span class="st"> </span><span class="kw">kBET</span>(
                <span class="dt">df =</span> <span class="kw">t</span>(<span class="kw">assay</span>(sce[,sce<span class="op">$</span>individual<span class="op">==</span><span class="st"> </span>i], j)), 
                <span class="dt">batch =</span> sce<span class="op">$</span>batch[sce<span class="op">$</span>individual<span class="op">==</span>i], 
                <span class="dt">heuristic =</span> <span class="ot">TRUE</span>, 
                <span class="dt">verbose =</span> <span class="ot">FALSE</span>, 
                <span class="dt">addTest =</span> <span class="ot">FALSE</span>, 
                <span class="dt">plot =</span> <span class="ot">FALSE</span>)
            results[[i]][[j]] &lt;-<span class="st"> </span>tmp<span class="op">$</span>summary<span class="op">$</span>kBET.observed[<span class="dv">1</span>]
        }
    }
    <span class="kw">return</span>(<span class="kw">as.data.frame</span>(results))
}

eff_debatching &lt;-<span class="st"> </span><span class="kw">compare_kBET_results</span>(reads.qc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(<span class="st">&quot;reshape2&quot;</span>)
<span class="kw">require</span>(<span class="st">&quot;RColorBrewer&quot;</span>)
<span class="co"># Plot results</span>
dod &lt;-<span class="st"> </span><span class="kw">melt</span>(<span class="kw">as.matrix</span>(eff_debatching),  <span class="dt">value.name =</span> <span class="st">&quot;kBET&quot;</span>)
<span class="kw">colnames</span>(dod)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Normalisation&quot;</span>, <span class="st">&quot;Individual&quot;</span>)

colorset &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;gray&#39;</span>, <span class="kw">brewer.pal</span>(<span class="dt">n =</span> <span class="dv">9</span>, <span class="st">&quot;RdYlBu&quot;</span>))

<span class="kw">ggplot</span>(dod, <span class="kw">aes</span>(Normalisation, Individual, <span class="dt">fill=</span>kBET)) <span class="op">+</span><span class="st">  </span>
<span class="st">    </span><span class="kw">geom_tile</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradient2</span>(
        <span class="dt">na.value =</span> <span class="st">&quot;gray&quot;</span>,
        <span class="dt">low =</span> colorset[<span class="dv">2</span>],
        <span class="dt">mid=</span>colorset[<span class="dv">6</span>],
        <span class="dt">high =</span> colorset[<span class="dv">10</span>],
        <span class="dt">midpoint =</span> <span class="fl">0.5</span>, <span class="dt">limit =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_x_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_y_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(
            <span class="dt">angle =</span> <span class="dv">45</span>, 
            <span class="dt">vjust =</span> <span class="dv">1</span>, 
            <span class="dt">size =</span> <span class="dv">12</span>, 
            <span class="dt">hjust =</span> <span class="dv">1</span>
        )
    ) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Effect of batch regression methods per individual&quot;</span>)</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 9 (stretch)
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libblas.so.3
## LAPACK: /usr/lib/libopenblasp-r0.2.19.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] RColorBrewer_1.1-2         reshape2_1.4.2            
##  [3] sva_3.26.0                 genefilter_1.60.0         
##  [5] mgcv_1.8-22                nlme_3.1-129              
##  [7] kBET_0.99.0                scran_1.6.2               
##  [9] scater_1.6.0               SingleCellExperiment_1.0.0
## [11] ggplot2_2.2.1              RUVSeq_1.12.0             
## [13] edgeR_3.20.1               limma_3.34.1              
## [15] EDASeq_2.12.0              ShortRead_1.36.0          
## [17] GenomicAlignments_1.14.0   SummarizedExperiment_1.8.0
## [19] DelayedArray_0.4.1         matrixStats_0.52.2        
## [21] Rsamtools_1.30.0           GenomicRanges_1.30.0      
## [23] GenomeInfoDb_1.14.0        Biostrings_2.46.0         
## [25] XVector_0.18.0             IRanges_2.12.0            
## [27] S4Vectors_0.16.0           BiocParallel_1.12.0       
## [29] Biobase_2.38.0             BiocGenerics_0.24.0       
## [31] scRNA.seq.funcs_0.1.0      knitr_1.17                
## 
## loaded via a namespace (and not attached):
##  [1] Rtsne_0.13              ggbeeswarm_0.6.0       
##  [3] colorspace_1.3-2        rjson_0.2.15           
##  [5] hwriter_1.3.2           dynamicTreeCut_1.63-1  
##  [7] rprojroot_1.2           DT_0.2                 
##  [9] bit64_0.9-7             AnnotationDbi_1.40.0   
## [11] splines_3.4.2           R.methodsS3_1.7.1      
## [13] tximport_1.6.0          DESeq_1.30.0           
## [15] geneplotter_1.56.0      annotate_1.56.1        
## [17] cluster_2.0.6           R.oo_1.21.0            
## [19] shinydashboard_0.6.1    shiny_1.0.5            
## [21] compiler_3.4.2          backports_1.1.1        
## [23] assertthat_0.2.0        Matrix_1.2-7.1         
## [25] lazyeval_0.2.1          htmltools_0.3.6        
## [27] prettyunits_1.0.2       tools_3.4.2            
## [29] igraph_1.1.2            bindrcpp_0.2           
## [31] gtable_0.2.0            glue_1.2.0             
## [33] GenomeInfoDbData_0.99.1 dplyr_0.7.4            
## [35] Rcpp_0.12.13            rtracklayer_1.38.0     
## [37] stringr_1.2.0           mime_0.5               
## [39] hypergeo_1.2-13         statmod_1.4.30         
## [41] XML_3.98-1.9            zoo_1.8-0              
## [43] zlibbioc_1.24.0         MASS_7.3-45            
## [45] scales_0.5.0            aroma.light_3.8.0      
## [47] rhdf5_2.22.0            yaml_2.1.14            
## [49] memoise_1.1.0           gridExtra_2.3          
## [51] biomaRt_2.34.0          latticeExtra_0.6-28    
## [53] stringi_1.1.5           RSQLite_2.0            
## [55] RMySQL_0.10.13          orthopolynom_1.0-5     
## [57] GenomicFeatures_1.30.0  contfrac_1.1-11        
## [59] rlang_0.1.4             pkgconfig_2.0.1        
## [61] moments_0.14            bitops_1.0-6           
## [63] evaluate_0.10.1         lattice_0.20-34        
## [65] bindr_0.1               labeling_0.3           
## [67] htmlwidgets_0.9         cowplot_0.9.0          
## [69] bit_1.1-12              deSolve_1.20           
## [71] plyr_1.8.4              magrittr_1.5           
## [73] bookdown_0.5            R6_2.2.2               
## [75] DBI_0.7                 survival_2.40-1        
## [77] RCurl_1.95-4.8          tibble_1.3.4           
## [79] rmarkdown_1.7           viridis_0.4.0          
## [81] progress_1.1.2          locfit_1.5-9.1         
## [83] grid_3.4.2              data.table_1.10.4-3    
## [85] FNN_1.1                 blob_1.1.0             
## [87] digest_0.6.12           xtable_1.8-2           
## [89] httpuv_1.3.5            elliptic_1.3-7         
## [91] R.utils_2.6.0           munsell_0.4.3          
## [93] beeswarm_0.2.3          viridisLite_0.2.0      
## [95] vipor_0.4.5</code></pre>

</div>
</div>
<h3> References</h3>
<div id="refs" class="references">
<div id="ref-Tung2017-ba">
<p>Tung, Po-Yuan, John D. Blischak, Chiaowen Joyce Hsiao, David A. Knowles, Jonathan E. Burnett, Jonathan K. Pritchard, and Yoav Gilad. 2017. “Batch Effects and the Effective Design of Single-Cell Gene Expression Studies.” <em>Sci. Rep.</em> 7 (January). Springer Nature: 39921. doi:<a href="https://doi.org/10.1038/srep39921">10.1038/srep39921</a>.</p>
</div>
<div id="ref-McCarthy2017-kb">
<p>McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Wills. 2017. “Scater: Pre-processing, Quality Control, Normalization and Visualization of Single-Cell RNA-Seq Data in R.” <em>Bioinformatics</em>, January. Oxford University Press (OUP), btw777. doi:<a href="https://doi.org/10.1093/bioinformatics/btw777">10.1093/bioinformatics/btw777</a>.</p>
</div>
<div id="ref-Anders2010-jr">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biol</em> 11 (10). Springer Nature: R106. doi:<a href="https://doi.org/10.1186/gb-2010-11-10-r106">10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-Bullard2010-eb">
<p>Bullard, James H, Elizabeth Purdom, Kasper D Hansen, and Sandrine Dudoit. 2010. “Evaluation of Statistical Methods for Normalization and Differential Expression in mRNA-Seq Experiments.” <em>BMC Bioinformatics</em> 11 (1). Springer Nature: 94. doi:<a href="https://doi.org/10.1186/1471-2105-11-94">10.1186/1471-2105-11-94</a>.</p>
</div>
<div id="ref-Robinson2010-hz">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.” <em>Genome Biol</em> 11 (3). Springer Nature: R25. doi:<a href="https://doi.org/10.1186/gb-2010-11-3-r25">10.1186/gb-2010-11-3-r25</a>.</p>
</div>
<div id="ref-L_Lun2016-pq">
<p>L. Lun, Aaron T., Karsten Bach, and John C. Marioni. 2016. “Pooling Across Cells to Normalize Single-Cell RNA Sequencing Data with Many Zero Counts.” <em>Genome Biol</em> 17 (1). Springer Nature. doi:<a href="https://doi.org/10.1186/s13059-016-0947-7">10.1186/s13059-016-0947-7</a>.</p>
</div>
<div id="ref-Haghverdi2017-vh">
<p>Haghverdi, Laleh, Aaron T L Lun, Michael D Morgan, and John C Marioni. 2017. “Correcting Batch Effects in Single-Cell RNA Sequencing Data by Matching Mutual Nearest Neighbours.” <em>bioRxiv</em>, July, 165118.</p>
</div>
<div id="ref-Buttner2017-ds">
<p>Buttner, Maren, Zhichao Miao, Alexander Wolf, Sarah A Teichmann, and Fabian J Theis. 2017. “Assessment of Batch-Correction Methods for scRNA-seq Data with a New Test Metric.” <em>bioRxiv</em>, October, 200345.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-single-cell-rna-seq.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="biological-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
