<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data</title>
  <meta name="description" content="Analysis of single cell RNA-seq data">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data" />
  
  
  

<meta name="author" content="Vladimir Kiselev (wikiselev), Tallulah Andrews, Davis McCarthy (davisjmcc), Maren Büttner (marenbuettner) and Martin Hemberg (m_hemberg)">


<meta name="date" content="2017-10-26">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="construction-of-expression-matrix.html">
<link rel="next" href="biological-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#video"><i class="fa fa-check"></i><b>1.1</b> Video</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#docker-image"><i class="fa fa-check"></i><b>1.4</b> Docker image</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.6</b> Prerequisites</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.7</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="technical-requirements.html"><a href="technical-requirements.html"><i class="fa fa-check"></i><b>2</b> Technical requirements</a><ul>
<li class="chapter" data-level="2.1" data-path="technical-requirements.html"><a href="technical-requirements.html#r-based"><i class="fa fa-check"></i><b>2.1</b> R-based</a></li>
<li class="chapter" data-level="2.2" data-path="technical-requirements.html"><a href="technical-requirements.html#docker-image-1"><i class="fa fa-check"></i><b>2.2</b> Docker image</a></li>
<li class="chapter" data-level="2.3" data-path="technical-requirements.html"><a href="technical-requirements.html#manual-installation"><i class="fa fa-check"></i><b>2.3</b> Manual installation</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>3</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>3.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>3.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>3.3</b> Workflow</a></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>3.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="3.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>3.5</b> Challenges</a></li>
<li class="chapter" data-level="3.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>3.6</b> Experimental methods</a><ul>
<li class="chapter" data-level="3.6.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#overview-of-experimental-methods-for-generating-scrna-seq-data"><i class="fa fa-check"></i><b>3.6.1</b> Overview of experimental methods for generating scRNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>3.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>4</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="4.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-qc"><i class="fa fa-check"></i><b>4.1</b> Reads QC</a></li>
<li class="chapter" data-level="4.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>4.2</b> Reads alignment</a></li>
<li class="chapter" data-level="4.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-example"><i class="fa fa-check"></i><b>4.3</b> Alignment example</a></li>
<li class="chapter" data-level="4.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>4.4</b> Mapping QC</a></li>
<li class="chapter" data-level="4.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>4.5</b> Reads quantification</a></li>
<li class="chapter" data-level="4.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>4.6</b> Unique Molecular Identifiers (UMIs)</a><ul>
<li class="chapter" data-level="4.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>4.6.1</b> Introduction</a></li>
<li class="chapter" data-level="4.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-barcodes"><i class="fa fa-check"></i><b>4.6.2</b> Mapping Barcodes</a></li>
<li class="chapter" data-level="4.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#counting-barcodes"><i class="fa fa-check"></i><b>4.6.3</b> Counting Barcodes</a></li>
<li class="chapter" data-level="4.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correcting-for-errors"><i class="fa fa-check"></i><b>4.6.4</b> Correcting for Errors</a></li>
<li class="chapter" data-level="4.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downstream-analysis"><i class="fa fa-check"></i><b>4.6.5</b> Downstream Analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor-singlecellexperiment-and-scater"><i class="fa fa-check"></i><b>4.7</b> Bioconductor, <code>SingleCellExperiment</code> and <code>scater</code></a><ul>
<li class="chapter" data-level="4.7.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor"><i class="fa fa-check"></i><b>4.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="4.7.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>4.7.2</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="4.7.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scater-package"><i class="fa fa-check"></i><b>4.7.3</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>4.8</b> Expression QC (UMI)</a><ul>
<li class="chapter" data-level="4.8.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>4.8.1</b> Introduction</a></li>
<li class="chapter" data-level="4.8.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tung-dataset"><i class="fa fa-check"></i><b>4.8.2</b> Tung dataset</a></li>
<li class="chapter" data-level="4.8.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>4.8.3</b> Cell QC</a></li>
<li class="chapter" data-level="4.8.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>4.8.4</b> Cell filtering</a></li>
<li class="chapter" data-level="4.8.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>4.8.5</b> Compare filterings</a></li>
<li class="chapter" data-level="4.8.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>4.8.6</b> Gene analysis</a></li>
<li class="chapter" data-level="4.8.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#save-the-data"><i class="fa fa-check"></i><b>4.8.7</b> Save the data</a></li>
<li class="chapter" data-level="4.8.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise"><i class="fa fa-check"></i><b>4.8.8</b> Big Exercise</a></li>
<li class="chapter" data-level="4.8.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo"><i class="fa fa-check"></i><b>4.8.9</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#expression-qc-reads"><i class="fa fa-check"></i><b>4.9</b> Expression QC (Reads)</a><ul>
<li class="chapter" data-level="4.9.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-1"><i class="fa fa-check"></i><b>4.9.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>4.10</b> Data visualization</a><ul>
<li class="chapter" data-level="4.10.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>4.10.1</b> Introduction</a></li>
<li class="chapter" data-level="4.10.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>4.10.2</b> PCA plot</a></li>
<li class="chapter" data-level="4.10.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>4.10.3</b> tSNE map</a></li>
<li class="chapter" data-level="4.10.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise-1"><i class="fa fa-check"></i><b>4.10.4</b> Big Exercise</a></li>
<li class="chapter" data-level="4.10.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-2"><i class="fa fa-check"></i><b>4.10.5</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.11" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization-reads"><i class="fa fa-check"></i><b>4.11</b> Data visualization (Reads)</a><ul>
<li class="chapter" data-level="4.11.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-3"><i class="fa fa-check"></i><b>4.11.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.12" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>4.12</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="4.12.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>4.12.1</b> Introduction</a></li>
<li class="chapter" data-level="4.12.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>4.12.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="4.12.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>4.12.3</b> Explanatory variables</a></li>
<li class="chapter" data-level="4.12.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#other-confounders"><i class="fa fa-check"></i><b>4.12.4</b> Other confounders</a></li>
<li class="chapter" data-level="4.12.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise"><i class="fa fa-check"></i><b>4.12.5</b> Exercise</a></li>
<li class="chapter" data-level="4.12.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-4"><i class="fa fa-check"></i><b>4.12.6</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.13" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors-reads"><i class="fa fa-check"></i><b>4.13</b> Identifying confounding factors (Reads)</a><ul>
<li class="chapter" data-level="4.13.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-5"><i class="fa fa-check"></i><b>4.13.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.14" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>4.14</b> Normalization theory</a><ul>
<li class="chapter" data-level="4.14.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>4.14.1</b> Introduction</a></li>
<li class="chapter" data-level="4.14.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>4.14.2</b> Library size</a></li>
<li class="chapter" data-level="4.14.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>4.14.3</b> Normalisations</a></li>
</ul></li>
<li class="chapter" data-level="4.15" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-umi"><i class="fa fa-check"></i><b>4.15</b> Normalization practice (UMI)</a><ul>
<li class="chapter" data-level="4.15.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw"><i class="fa fa-check"></i><b>4.15.1</b> Raw</a></li>
<li class="chapter" data-level="4.15.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cpm"><i class="fa fa-check"></i><b>4.15.2</b> CPM</a></li>
<li class="chapter" data-level="4.15.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tmm"><i class="fa fa-check"></i><b>4.15.3</b> TMM</a></li>
<li class="chapter" data-level="4.15.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scran"><i class="fa fa-check"></i><b>4.15.4</b> scran</a></li>
<li class="chapter" data-level="4.15.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#size-factor-rle"><i class="fa fa-check"></i><b>4.15.5</b> Size-factor (RLE)</a></li>
<li class="chapter" data-level="4.15.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#upperquantile"><i class="fa fa-check"></i><b>4.15.6</b> Upperquantile</a></li>
<li class="chapter" data-level="4.15.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downsampling"><i class="fa fa-check"></i><b>4.15.7</b> Downsampling</a></li>
<li class="chapter" data-level="4.15.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalizing-for-genetranscript-length"><i class="fa fa-check"></i><b>4.15.8</b> Normalizing for gene/transcript length</a></li>
<li class="chapter" data-level="4.15.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise-1"><i class="fa fa-check"></i><b>4.15.9</b> Exercise</a></li>
<li class="chapter" data-level="4.15.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-6"><i class="fa fa-check"></i><b>4.15.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.16" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-reads"><i class="fa fa-check"></i><b>4.16</b> Normalization practice (Reads)</a><ul>
<li class="chapter" data-level="4.16.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-7"><i class="fa fa-check"></i><b>4.16.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.17" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>4.17</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="4.17.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-5"><i class="fa fa-check"></i><b>4.17.1</b> Introduction</a></li>
<li class="chapter" data-level="4.17.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#remove-unwanted-variation"><i class="fa fa-check"></i><b>4.17.2</b> Remove Unwanted Variation</a></li>
<li class="chapter" data-level="4.17.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#combat"><i class="fa fa-check"></i><b>4.17.3</b> Combat</a></li>
<li class="chapter" data-level="4.17.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>4.17.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="4.17.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#glm"><i class="fa fa-check"></i><b>4.17.5</b> GLM</a></li>
<li class="chapter" data-level="4.17.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-1"><i class="fa fa-check"></i><b>4.17.6</b> Effectiveness 1</a></li>
<li class="chapter" data-level="4.17.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-2"><i class="fa fa-check"></i><b>4.17.7</b> Effectiveness 2</a></li>
<li class="chapter" data-level="4.17.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-3"><i class="fa fa-check"></i><b>4.17.8</b> Effectiveness 3</a></li>
<li class="chapter" data-level="4.17.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-4"><i class="fa fa-check"></i><b>4.17.9</b> Effectiveness 4</a></li>
<li class="chapter" data-level="4.17.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise-2"><i class="fa fa-check"></i><b>4.17.10</b> Exercise</a></li>
<li class="chapter" data-level="4.17.11" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-8"><i class="fa fa-check"></i><b>4.17.11</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html"><i class="fa fa-check"></i><b>5</b> Dealing with confounders (Reads)</a><ul>
<li class="chapter" data-level="5.1" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#remove-unwanted-variation-1"><i class="fa fa-check"></i><b>5.1</b> Remove Unwanted Variation</a><ul>
<li class="chapter" data-level="5.1.1" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#ruvg-1"><i class="fa fa-check"></i><b>5.1.1</b> RUVg</a></li>
<li class="chapter" data-level="5.1.2" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#ruvs-1"><i class="fa fa-check"></i><b>5.1.2</b> RUVs</a></li>
<li class="chapter" data-level="5.1.3" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#combat-1"><i class="fa fa-check"></i><b>5.1.3</b> Combat</a></li>
<li class="chapter" data-level="5.1.4" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#mnncorrect-1"><i class="fa fa-check"></i><b>5.1.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="5.1.5" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#glm-1"><i class="fa fa-check"></i><b>5.1.5</b> GLM</a></li>
<li class="chapter" data-level="5.1.6" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#effectiveness-1-1"><i class="fa fa-check"></i><b>5.1.6</b> Effectiveness 1</a></li>
<li class="chapter" data-level="5.1.7" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#effectiveness-2-1"><i class="fa fa-check"></i><b>5.1.7</b> Effectiveness 2</a></li>
<li class="chapter" data-level="5.1.8" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#effectiveness-3-1"><i class="fa fa-check"></i><b>5.1.8</b> Effectiveness 3</a></li>
<li class="chapter" data-level="5.1.9" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#effectiveness-4-1"><i class="fa fa-check"></i><b>5.1.9</b> Effectiveness 4</a></li>
<li class="chapter" data-level="5.1.10" data-path="dealing-with-confounders-reads.html"><a href="dealing-with-confounders-reads.html#sessioninfo-9"><i class="fa fa-check"></i><b>5.1.10</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>6</b> Biological Analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-introduction"><i class="fa fa-check"></i><b>6.1</b> Clustering Introduction</a><ul>
<li class="chapter" data-level="6.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>6.1.1</b> Introduction</a></li>
<li class="chapter" data-level="6.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reductions"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reductions</a></li>
<li class="chapter" data-level="6.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>6.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="6.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>6.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="6.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#tools-for-scrna-seq-data"><i class="fa fa-check"></i><b>6.1.5</b> Tools for scRNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>6.2</b> Clustering example</a><ul>
<li class="chapter" data-level="6.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>6.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="6.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>6.2.2</b> SC3</a></li>
<li class="chapter" data-level="6.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#pcareduce-1"><i class="fa fa-check"></i><b>6.2.3</b> pcaReduce</a></li>
<li class="chapter" data-level="6.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>6.2.4</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="6.2.5" data-path="biological-analysis.html"><a href="biological-analysis.html#snn-cliq-1"><i class="fa fa-check"></i><b>6.2.5</b> SNN-Cliq</a></li>
<li class="chapter" data-level="6.2.6" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>6.2.6</b> SINCERA</a></li>
<li class="chapter" data-level="6.2.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-10"><i class="fa fa-check"></i><b>6.2.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#seurat-1"><i class="fa fa-check"></i><b>6.3</b> <code>Seurat</code></a><ul>
<li class="chapter" data-level="6.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-11"><i class="fa fa-check"></i><b>6.3.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>6.4</b> Feature Selection</a><ul>
<li class="chapter" data-level="6.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>6.4.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="6.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>6.4.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="6.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-methods"><i class="fa fa-check"></i><b>6.4.3</b> Comparing Methods</a></li>
<li class="chapter" data-level="6.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-12"><i class="fa fa-check"></i><b>6.4.4</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="biological-analysis.html"><a href="biological-analysis.html#pseudotime-analysis"><i class="fa fa-check"></i><b>6.5</b> Pseudotime analysis</a><ul>
<li class="chapter" data-level="6.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>6.5.1</b> TSCAN</a></li>
<li class="chapter" data-level="6.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>6.5.2</b> monocle</a></li>
<li class="chapter" data-level="6.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>6.5.3</b> Diffusion maps</a></li>
<li class="chapter" data-level="6.5.4" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>6.5.4</b> SLICER</a></li>
<li class="chapter" data-level="6.5.5" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>6.5.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="6.5.6" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>6.5.6</b> Expression of genes through time</a></li>
<li class="chapter" data-level="6.5.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-13"><i class="fa fa-check"></i><b>6.5.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>6.6</b> Imputation</a><ul>
<li class="chapter" data-level="6.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>6.6.1</b> scImpute</a></li>
<li class="chapter" data-level="6.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>6.6.2</b> MAGIC</a></li>
<li class="chapter" data-level="6.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-14"><i class="fa fa-check"></i><b>6.6.3</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="biological-analysis.html"><a href="biological-analysis.html#differential-expression-de-analysis"><i class="fa fa-check"></i><b>6.7</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="6.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>6.7.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="6.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>6.7.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="6.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>6.7.3</b> Differences in Distribution</a></li>
<li class="chapter" data-level="6.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rnaseq-data"><i class="fa fa-check"></i><b>6.7.4</b> Models of single-cell RNASeq data</a></li>
</ul></li>
<li class="chapter" data-level="6.8" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-a-real-dataset"><i class="fa fa-check"></i><b>6.8</b> DE in a real dataset</a><ul>
<li class="chapter" data-level="6.8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-7"><i class="fa fa-check"></i><b>6.8.1</b> Introduction</a></li>
<li class="chapter" data-level="6.8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>6.8.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="6.8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>6.8.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="6.8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>6.8.4</b> edgeR</a></li>
<li class="chapter" data-level="6.8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>6.8.5</b> Monocle</a></li>
<li class="chapter" data-level="6.8.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>6.8.6</b> MAST</a></li>
<li class="chapter" data-level="6.8.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-methods-1h-to-run"><i class="fa fa-check"></i><b>6.8.7</b> Slow Methods (&gt;1h to run)</a></li>
<li class="chapter" data-level="6.8.8" data-path="biological-analysis.html"><a href="biological-analysis.html#bpsc"><i class="fa fa-check"></i><b>6.8.8</b> BPSC</a></li>
<li class="chapter" data-level="6.8.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>6.8.9</b> SCDE</a></li>
<li class="chapter" data-level="6.8.10" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-15"><i class="fa fa-check"></i><b>6.8.10</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><i class="fa fa-check"></i><b>7</b> “Ideal” scRNAseq pipeline (as of Mar 2017)</a><ul>
<li class="chapter" data-level="7.1" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#experimental-design"><i class="fa fa-check"></i><b>7.1</b> Experimental Design</a></li>
<li class="chapter" data-level="7.2" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#processing-reads"><i class="fa fa-check"></i><b>7.2</b> Processing Reads</a></li>
<li class="chapter" data-level="7.3" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#preparing-expression-matrix"><i class="fa fa-check"></i><b>7.3</b> Preparing Expression Matrix</a></li>
<li class="chapter" data-level="7.4" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#biological-interpretation"><i class="fa fa-check"></i><b>7.4</b> Biological Interpretation</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="advanced-exercises.html"><a href="advanced-exercises.html"><i class="fa fa-check"></i><b>8</b> Advanced exercises</a></li>
<li class="chapter" data-level="9" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>9</b> Resources</a><ul>
<li class="chapter" data-level="9.1" data-path="resources.html"><a href="resources.html#scrna-seq-protocols"><i class="fa fa-check"></i><b>9.1</b> scRNA-seq protocols</a></li>
<li class="chapter" data-level="9.2" data-path="resources.html"><a href="resources.html#external-rna-control-consortium-ercc"><i class="fa fa-check"></i><b>9.2</b> External RNA Control Consortium (ERCC)</a></li>
<li class="chapter" data-level="9.3" data-path="resources.html"><a href="resources.html#scrna-seq-analysis-tools"><i class="fa fa-check"></i><b>9.3</b> scRNA-seq analysis tools</a></li>
<li class="chapter" data-level="9.4" data-path="resources.html"><a href="resources.html#scrna-seq-public-datasets"><i class="fa fa-check"></i><b>9.4</b> scRNA-seq public datasets</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>10</b> References</a></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab, 2017</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dealing-with-confounders-reads" class="section level1">
<h1><span class="header-section-number">5</span> Dealing with confounders (Reads)</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNA.seq.funcs)
<span class="kw">library</span>(RUVSeq)
<span class="kw">library</span>(scater)
<span class="kw">library</span>(SingleCellExperiment)
<span class="kw">library</span>(scran)
<span class="kw">library</span>(kBET)
<span class="kw">library</span>(sva) <span class="co"># Combat</span>
<span class="kw">library</span>(edgeR)
<span class="kw">set.seed</span>(<span class="dv">1234567</span>)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
reads &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;tung/reads.rds&quot;</span>)
reads.qc &lt;-<span class="st"> </span>reads[<span class="kw">rowData</span>(reads)<span class="op">$</span>use, <span class="kw">colData</span>(reads)<span class="op">$</span>use]
endog_genes &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">rowData</span>(reads.qc)<span class="op">$</span>is_feature_control
erccs &lt;-<span class="st"> </span><span class="kw">rowData</span>(reads.qc)<span class="op">$</span>is_feature_control

qclust &lt;-<span class="st"> </span><span class="kw">quickCluster</span>(reads.qc, <span class="dt">min.size =</span> <span class="dv">30</span>)
reads.qc &lt;-<span class="st"> </span><span class="kw">computeSumFactors</span>(reads.qc, <span class="dt">sizes =</span> <span class="dv">15</span>, <span class="dt">clusters =</span> qclust)
reads.qc &lt;-<span class="st"> </span><span class="kw">normalize</span>(reads.qc)</code></pre></div>
<div id="remove-unwanted-variation-1" class="section level2">
<h2><span class="header-section-number">5.1</span> Remove Unwanted Variation</h2>
<div id="ruvg-1" class="section level3">
<h3><span class="header-section-number">5.1.1</span> RUVg</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(reads.qc), erccs, <span class="dt">k =</span> <span class="dv">1</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvg1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvg &lt;-<span class="st"> </span><span class="kw">RUVg</span>(<span class="kw">counts</span>(reads.qc), erccs, <span class="dt">k =</span> <span class="dv">10</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvg10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvg<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
</div>
<div id="ruvs-1" class="section level3">
<h3><span class="header-section-number">5.1.2</span> RUVs</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scIdx &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw">max</span>(<span class="kw">table</span>(reads.qc<span class="op">$</span>individual)), <span class="dt">nrow =</span> <span class="dv">3</span>)
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>)
scIdx[<span class="dv">1</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>)
scIdx[<span class="dv">2</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
tmp &lt;-<span class="st"> </span><span class="kw">which</span>(reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>)
scIdx[<span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp)] &lt;-<span class="st"> </span>tmp
cIdx &lt;-<span class="st"> </span><span class="kw">rownames</span>(reads.qc)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(reads.qc), cIdx, <span class="dt">k =</span> <span class="dv">1</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvs1&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)
ruvs &lt;-<span class="st"> </span><span class="kw">RUVs</span>(<span class="kw">counts</span>(reads.qc), cIdx, <span class="dt">k =</span> <span class="dv">10</span>, <span class="dt">scIdx =</span> scIdx, <span class="dt">isLog =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;ruvs10&quot;</span>) &lt;-<span class="st"> </span><span class="kw">log2</span>(
    <span class="kw">t</span>(<span class="kw">t</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">/</span><span class="st"> </span><span class="kw">colSums</span>(ruvs<span class="op">$</span>normalizedCounts) <span class="op">*</span><span class="st"> </span><span class="fl">1e6</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
)</code></pre></div>
</div>
<div id="combat-1" class="section level3">
<h3><span class="header-section-number">5.1.3</span> Combat</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">combat_data &lt;-<span class="st"> </span><span class="kw">logcounts</span>(reads.qc)
mod_data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(combat_data))
<span class="co"># Basic batch removal</span>
mod0 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">data =</span> mod_data) 
<span class="co"># Preserve biological variability</span>
mod1 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>reads.qc<span class="op">$</span>individual, <span class="dt">data =</span> mod_data) 
<span class="co"># adjust for total genes detected</span>
mod2 =<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span>reads.qc<span class="op">$</span>total_features, <span class="dt">data =</span> mod_data)
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;combat&quot;</span>) &lt;-<span class="st"> </span><span class="kw">ComBat</span>(
    <span class="dt">dat =</span> <span class="kw">t</span>(mod_data), 
    <span class="dt">batch =</span> <span class="kw">factor</span>(reads.qc<span class="op">$</span>batch), 
    <span class="dt">mod =</span> mod0,
    <span class="dt">par.prior =</span> <span class="ot">TRUE</span>,
    <span class="dt">prior.plots =</span> <span class="ot">FALSE</span>
)</code></pre></div>
<pre><code>## Standardizing Data across genes</code></pre>
<p><strong>Exercise 1</strong></p>
<pre><code>## Standardizing Data across genes</code></pre>
</div>
<div id="mnncorrect-1" class="section level3">
<h3><span class="header-section-number">5.1.4</span> mnnCorrect</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">do_mnn &lt;-<span class="st"> </span><span class="cf">function</span>(data.qc) {
    batch1 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r1&quot;</span>]
    batch2 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r2&quot;</span>]
    batch3 &lt;-<span class="st"> </span>data.qc[, data.qc<span class="op">$</span>replicate <span class="op">==</span><span class="st"> &quot;r3&quot;</span>]
    
    <span class="cf">if</span> (<span class="kw">ncol</span>(batch2) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch2), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]], x<span class="op">$</span>corrected[[<span class="dv">3</span>]]))
    } <span class="cf">else</span> {
        x =<span class="st"> </span><span class="kw">mnnCorrect</span>(
          <span class="kw">logcounts</span>(batch1), <span class="kw">logcounts</span>(batch3),  
          <span class="dt">k =</span> <span class="dv">20</span>,
          <span class="dt">sigma =</span> <span class="fl">0.1</span>,
          <span class="dt">cos.norm =</span> <span class="ot">TRUE</span>,
          <span class="dt">svd.dim =</span> <span class="dv">2</span>
        )
        <span class="kw">return</span>(<span class="kw">cbind</span>(x<span class="op">$</span>corrected[[<span class="dv">1</span>]], x<span class="op">$</span>corrected[[<span class="dv">2</span>]]))
    }
}

indi1 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19098&quot;</span>])
indi2 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19101&quot;</span>])
indi3 &lt;-<span class="st"> </span><span class="kw">do_mnn</span>(reads.qc[, reads.qc<span class="op">$</span>individual <span class="op">==</span><span class="st"> &quot;NA19239&quot;</span>])

<span class="kw">assay</span>(reads.qc, <span class="st">&quot;mnn&quot;</span>) &lt;-<span class="st"> </span><span class="kw">cbind</span>(indi1, indi2, indi3);

<span class="co"># For a balanced design: </span>
<span class="co">#assay(reads.qc, &quot;mnn&quot;) &lt;- mnnCorrect(</span>
<span class="co">#    list(B1 = logcounts(batch1), B2 = logcounts(batch2), B3 = logcounts(batch3)),  </span>
<span class="co">#    k = 20,</span>
<span class="co">#    sigma = 0.1,</span>
<span class="co">#    cos.norm = TRUE,</span>
<span class="co">#    svd.dim = 2</span>
<span class="co">#)</span></code></pre></div>
</div>
<div id="glm-1" class="section level3">
<h3><span class="header-section-number">5.1.5</span> GLM</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">glm_fun &lt;-<span class="st"> </span><span class="cf">function</span>(g, batch, indi) {
  model &lt;-<span class="st"> </span><span class="kw">glm</span>(g <span class="op">~</span><span class="st"> </span>batch <span class="op">+</span><span class="st"> </span>indi)
  model<span class="op">$</span>coef[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="dv">0</span> <span class="co"># replace intercept with 0 to preserve reference batch.</span>
  <span class="kw">return</span>(model<span class="op">$</span>coef)
}
effects &lt;-<span class="st"> </span><span class="kw">apply</span>(
    <span class="kw">logcounts</span>(reads.qc), 
    <span class="dv">1</span>, 
    glm_fun, 
    <span class="dt">batch =</span> reads.qc<span class="op">$</span>batch, 
    <span class="dt">indi =</span> reads.qc<span class="op">$</span>individual
)
corrected &lt;-<span class="st"> </span><span class="kw">logcounts</span>(reads.qc) <span class="op">-</span><span class="st"> </span><span class="kw">t</span>(effects[<span class="kw">as.numeric</span>(<span class="kw">factor</span>(reads.qc<span class="op">$</span>batch)), ])
<span class="kw">assay</span>(reads.qc, <span class="st">&quot;glm&quot;</span>) &lt;-<span class="st"> </span>corrected</code></pre></div>
<p><strong>Exercise 2</strong></p>
</div>
<div id="effectiveness-1-1" class="section level3">
<h3><span class="header-section-number">5.1.6</span> Effectiveness 1</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotPCA</span>(
            reads.qc[endog_genes, ],
            <span class="dt">colour_by =</span> <span class="st">&quot;batch&quot;</span>,
            <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>,
            <span class="dt">shape_by =</span> <span class="st">&quot;individual&quot;</span>,
            <span class="dt">exprs_values =</span> n
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-3.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-4.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-5.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-6.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-7.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-8.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-9.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-10.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-11.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-10-12.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effectiveness-2-1" class="section level3">
<h3><span class="header-section-number">5.1.7</span> Effectiveness 2</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    res[[n]] &lt;-<span class="st"> </span><span class="kw">suppressWarnings</span>(<span class="kw">calc_cell_RLE</span>(<span class="kw">assay</span>(reads.qc, n), erccs))
}
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">boxplot</span>(res, <span class="dt">las=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effectiveness-3-1" class="section level3">
<h3><span class="header-section-number">5.1.8</span> Effectiveness 3</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(n <span class="cf">in</span> <span class="kw">assayNames</span>(reads.qc)) {
    <span class="kw">print</span>(
        <span class="kw">plotQC</span>(
            reads.qc[endog_genes, ],
            <span class="dt">type =</span> <span class="st">&quot;expl&quot;</span>,
            <span class="dt">exprs_values =</span> n,
            <span class="dt">variables =</span> <span class="kw">c</span>(
                <span class="st">&quot;total_features&quot;</span>,
                <span class="st">&quot;total_counts&quot;</span>,
                <span class="st">&quot;batch&quot;</span>,
                <span class="st">&quot;individual&quot;</span>,
                <span class="st">&quot;pct_counts_ERCC&quot;</span>,
                <span class="st">&quot;pct_counts_MT&quot;</span>
            )
        ) <span class="op">+</span>
<span class="st">        </span><span class="kw">ggtitle</span>(n)
    )
}</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-2.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-3.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-4.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-5.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>## Warning: Removed 6 rows containing non-finite values (stat_density).</code></pre>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-6.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>## Warning: Removed 54 rows containing non-finite values (stat_density).</code></pre>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-7.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-8.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-9.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-10.png" width="672" style="display: block; margin: auto;" /><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-11.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>## Warning in self$trans$transform(x): NaNs produced</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<pre><code>## Warning: Removed 129 rows containing non-finite values (stat_density).</code></pre>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-12-12.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effectiveness-4-1" class="section level3">
<h3><span class="header-section-number">5.1.9</span> Effectiveness 4</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">compare_kBET_results &lt;-<span class="st"> </span><span class="cf">function</span>(sce){
    indiv &lt;-<span class="st"> </span><span class="kw">unique</span>(sce<span class="op">$</span>individual)
    norms &lt;-<span class="st"> </span><span class="kw">assayNames</span>(sce) <span class="co"># Get all normalizations</span>
    results &lt;-<span class="st"> </span><span class="kw">list</span>()
    <span class="cf">for</span> (i <span class="cf">in</span> indiv){ 
        <span class="cf">for</span> (j <span class="cf">in</span> norms){
            tmp &lt;-<span class="st"> </span><span class="kw">kBET</span>(
                <span class="dt">df =</span> <span class="kw">t</span>(<span class="kw">assay</span>(sce[,sce<span class="op">$</span>individual<span class="op">==</span><span class="st"> </span>i], j)), 
                <span class="dt">batch =</span> sce<span class="op">$</span>batch[sce<span class="op">$</span>individual<span class="op">==</span>i], 
                <span class="dt">heuristic =</span> <span class="ot">TRUE</span>, 
                <span class="dt">verbose =</span> <span class="ot">FALSE</span>, 
                <span class="dt">addTest =</span> <span class="ot">FALSE</span>, 
                <span class="dt">plot =</span> <span class="ot">FALSE</span>)
            results[[i]][[j]] &lt;-<span class="st"> </span>tmp<span class="op">$</span>summary<span class="op">$</span>kBET.observed[<span class="dv">1</span>]
        }
    }
    <span class="kw">return</span>(<span class="kw">as.data.frame</span>(results))
}

eff_debatching &lt;-<span class="st"> </span><span class="kw">compare_kBET_results</span>(reads.qc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(<span class="st">&quot;reshape2&quot;</span>)
<span class="kw">require</span>(<span class="st">&quot;RColorBrewer&quot;</span>)
<span class="co"># Plot results</span>
dod &lt;-<span class="st"> </span><span class="kw">melt</span>(<span class="kw">as.matrix</span>(eff_debatching),  <span class="dt">value.name =</span> <span class="st">&quot;kBET&quot;</span>)
<span class="kw">colnames</span>(dod)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Normalisation&quot;</span>, <span class="st">&quot;Individual&quot;</span>)

colorset &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;gray&#39;</span>, <span class="kw">brewer.pal</span>(<span class="dt">n =</span> <span class="dv">9</span>, <span class="st">&quot;RdYlBu&quot;</span>))

<span class="kw">ggplot</span>(dod, <span class="kw">aes</span>(Normalisation, Individual, <span class="dt">fill=</span>kBET)) <span class="op">+</span><span class="st">  </span>
<span class="st">    </span><span class="kw">geom_tile</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradient2</span>(
        <span class="dt">na.value =</span> <span class="st">&quot;gray&quot;</span>,
        <span class="dt">low =</span> colorset[<span class="dv">2</span>],
        <span class="dt">mid=</span>colorset[<span class="dv">6</span>],
        <span class="dt">high =</span> colorset[<span class="dv">10</span>],
        <span class="dt">midpoint =</span> <span class="fl">0.5</span>, <span class="dt">limit =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_x_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_y_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(
            <span class="dt">angle =</span> <span class="dv">45</span>, 
            <span class="dt">vjust =</span> <span class="dv">1</span>, 
            <span class="dt">size =</span> <span class="dv">12</span>, 
            <span class="dt">hjust =</span> <span class="dv">1</span>
        )
    ) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&quot;Effect of batch regression methods per individual&quot;</span>)</code></pre></div>
<p><img src="16-remove-conf-reads_files/figure-html/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="sessioninfo-9" class="section level3">
<h3><span class="header-section-number">5.1.10</span> sessionInfo()</h3>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux buster/sid
## 
## Matrix products: default
## BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] RColorBrewer_1.1-2          reshape2_1.4.2             
##  [3] sva_3.24.4                  genefilter_1.58.1          
##  [5] mgcv_1.8-22                 nlme_3.1-131               
##  [7] kBET_0.99.0                 scran_1.5.14               
##  [9] scater_1.5.21               SingleCellExperiment_0.99.4
## [11] ggplot2_2.2.1               RUVSeq_1.10.0              
## [13] edgeR_3.18.1                limma_3.32.10              
## [15] EDASeq_2.10.0               ShortRead_1.34.2           
## [17] GenomicAlignments_1.12.2    SummarizedExperiment_1.6.5 
## [19] DelayedArray_0.2.7          matrixStats_0.52.2         
## [21] Rsamtools_1.28.0            GenomicRanges_1.28.6       
## [23] GenomeInfoDb_1.12.3         Biostrings_2.44.2          
## [25] XVector_0.16.0              IRanges_2.10.5             
## [27] S4Vectors_0.14.7            BiocParallel_1.10.1        
## [29] Biobase_2.36.2              BiocGenerics_0.22.1        
## [31] scRNA.seq.funcs_0.1.0       knitr_1.17                 
## 
## loaded via a namespace (and not attached):
##  [1] Rtsne_0.13              ggbeeswarm_0.6.0       
##  [3] colorspace_1.3-2        rjson_0.2.15           
##  [5] hwriter_1.3.2           dynamicTreeCut_1.63-1  
##  [7] rprojroot_1.2           DT_0.2                 
##  [9] bit64_0.9-7             AnnotationDbi_1.38.2   
## [11] splines_3.4.2           R.methodsS3_1.7.1      
## [13] tximport_1.4.0          DESeq_1.28.0           
## [15] geneplotter_1.54.0      annotate_1.54.0        
## [17] cluster_2.0.6           R.oo_1.21.0            
## [19] shinydashboard_0.6.1    shiny_1.0.5            
## [21] compiler_3.4.2          backports_1.1.1        
## [23] assertthat_0.2.0        Matrix_1.2-11          
## [25] lazyeval_0.2.0          htmltools_0.3.6        
## [27] tools_3.4.2             igraph_1.1.2           
## [29] bindrcpp_0.2            gtable_0.2.0           
## [31] glue_1.1.1              GenomeInfoDbData_0.99.0
## [33] dplyr_0.7.4             Rcpp_0.12.13           
## [35] rtracklayer_1.36.6      stringr_1.2.0          
## [37] mime_0.5                hypergeo_1.2-13        
## [39] statmod_1.4.30          XML_3.98-1.9           
## [41] zlibbioc_1.22.0         MASS_7.3-47            
## [43] zoo_1.8-0               scales_0.5.0           
## [45] aroma.light_3.6.0       rhdf5_2.20.0           
## [47] yaml_2.1.14             memoise_1.1.0          
## [49] gridExtra_2.3           biomaRt_2.32.1         
## [51] latticeExtra_0.6-28     stringi_1.1.5          
## [53] RSQLite_2.0             orthopolynom_1.0-5     
## [55] GenomicFeatures_1.28.5  contfrac_1.1-11        
## [57] rlang_0.1.2             pkgconfig_2.0.1        
## [59] moments_0.14            bitops_1.0-6           
## [61] evaluate_0.10.1         lattice_0.20-35        
## [63] bindr_0.1               labeling_0.3           
## [65] htmlwidgets_0.9         cowplot_0.8.0          
## [67] bit_1.1-12              deSolve_1.20           
## [69] plyr_1.8.4              magrittr_1.5           
## [71] bookdown_0.5            R6_2.2.2               
## [73] DBI_0.7                 survival_2.41-3        
## [75] RCurl_1.95-4.8          tibble_1.3.4           
## [77] rmarkdown_1.6           viridis_0.4.0          
## [79] locfit_1.5-9.1          grid_3.4.2             
## [81] data.table_1.10.4-2     FNN_1.1                
## [83] blob_1.1.0              digest_0.6.12          
## [85] xtable_1.8-2            httpuv_1.3.5           
## [87] elliptic_1.3-7          R.utils_2.5.0          
## [89] munsell_0.4.3           beeswarm_0.2.3         
## [91] viridisLite_0.2.0       vipor_0.4.5</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="construction-of-expression-matrix.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="biological-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/hemberg-lab/scRNA.seq.course/edit/master/16-remove-conf-reads.Rmd",
"text": "Edit"
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
